<!doctype html>
<html class="no-js" lang="en" data-content_root="../../../">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark">
        <script async src="https://www.googletagmanager.com/gtag/js?id="></script>
        <script>
            window.dataLayer = window.dataLayer || [];
            function gtag(){dataLayer.push(arguments);}
            gtag('js', new Date());
            
            gtag('config', '');
            
        </script>
    <link rel="index" title="Index" href="../../../genindex.html" /><link rel="search" title="Search" href="../../../search.html" />
        <link rel="prefetch" href="../../../_static/WorkBench.jpeg" as="image" />

    <!-- Generated with Sphinx 7.4.7 and Furo 2025.07.19 -->
        <title>multicom_ligand.utils.data_utils - MULTICOM_ligand 0.4.0</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=d111a655" />
    <link rel="stylesheet" type="text/css" href="../../../_static/styles/furo.css?v=25af2a20" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../../../_static/tabs.css?v=4c969af8" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-codeautolink.css?v=125d5c1c" />
    <link rel="stylesheet" type="text/css" href="../../../_static/styles/furo-extensions.css?v=8dab3a3b" />
    
    


<style>
  body {
    --color-code-background: #f2f2f2;
  --color-code-foreground: #1e1e1e;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #2b2b2b;
  --color-code-foreground: #f8f8f2;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #2b2b2b;
  --color-code-foreground: #f8f8f2;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-with-moon" viewBox="0 0 24 24">
    <title>Auto light/dark, in light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path style="opacity: 50%" d="M 5.411 14.504 C 5.471 14.504 5.532 14.504 5.591 14.504 C 3.639 16.319 4.383 19.569 6.931 20.352 C 7.693 20.586 8.512 20.551 9.25 20.252 C 8.023 23.207 4.056 23.725 2.11 21.184 C 0.166 18.642 1.702 14.949 4.874 14.536 C 5.051 14.512 5.231 14.5 5.411 14.5 L 5.411 14.504 Z"/>
      <line x1="14.5" y1="3.25" x2="14.5" y2="1.25"/>
      <line x1="14.5" y1="15.85" x2="14.5" y2="17.85"/>
      <line x1="10.044" y1="5.094" x2="8.63" y2="3.68"/>
      <line x1="19" y1="14.05" x2="20.414" y2="15.464"/>
      <line x1="8.2" y1="9.55" x2="6.2" y2="9.55"/>
      <line x1="20.8" y1="9.55" x2="22.8" y2="9.55"/>
      <line x1="10.044" y1="14.006" x2="8.63" y2="15.42"/>
      <line x1="19" y1="5.05" x2="20.414" y2="3.636"/>
      <circle cx="14.5" cy="9.55" r="3.6"/>
    </svg>
  </symbol>
  <symbol id="svg-moon-with-sun" viewBox="0 0 24 24">
    <title>Auto light/dark, in dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path d="M 8.282 7.007 C 8.385 7.007 8.494 7.007 8.595 7.007 C 5.18 10.184 6.481 15.869 10.942 17.24 C 12.275 17.648 13.706 17.589 15 17.066 C 12.851 22.236 5.91 23.143 2.505 18.696 C -0.897 14.249 1.791 7.786 7.342 7.063 C 7.652 7.021 7.965 7 8.282 7 L 8.282 7.007 Z"/>
      <line style="opacity: 50%" x1="18" y1="3.705" x2="18" y2="2.5"/>
      <line style="opacity: 50%" x1="18" y1="11.295" x2="18" y2="12.5"/>
      <line style="opacity: 50%" x1="15.316" y1="4.816" x2="14.464" y2="3.964"/>
      <line style="opacity: 50%" x1="20.711" y1="10.212" x2="21.563" y2="11.063"/>
      <line style="opacity: 50%" x1="14.205" y1="7.5" x2="13.001" y2="7.5"/>
      <line style="opacity: 50%" x1="21.795" y1="7.5" x2="23" y2="7.5"/>
      <line style="opacity: 50%" x1="15.316" y1="10.184" x2="14.464" y2="11.036"/>
      <line style="opacity: 50%" x1="20.711" y1="4.789" x2="21.563" y2="3.937"/>
      <circle style="opacity: 50%" cx="18" cy="7.5" r="2.169"/>
    </svg>
  </symbol>
  <symbol id="svg-pencil" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-pencil-code">
      <path d="M4 20h4l10.5 -10.5a2.828 2.828 0 1 0 -4 -4l-10.5 10.5v4" />
      <path d="M13.5 6.5l4 4" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
  <symbol id="svg-eye" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-eye-code">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M10 12a2 2 0 1 0 4 0a2 2 0 0 0 -4 0" />
      <path
        d="M11.11 17.958c-3.209 -.307 -5.91 -2.293 -8.11 -5.958c2.4 -4 5.4 -6 9 -6c3.6 0 6.6 2 9 6c-.21 .352 -.427 .688 -.647 1.008" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>

<a class="skip-to-content muted-link" href="#furo-main-content">Skip to content</a>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../../../index.html"><div class="brand">MULTICOM_ligand 0.4.0</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
          <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon no-toc" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../../../index.html">
  <div class="sidebar-logo-container">
    <img class="sidebar-logo" src="../../../_static/WorkBench.jpeg" alt="Logo"/>
  </div>
  
  <span class="sidebar-brand-text">MULTICOM_ligand 0.4.0</span>
  
</a><form class="sidebar-search-container" method="get" action="../../../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../installation.html#dynamicbind-checkpoint-0-25-gb">DynamicBind checkpoint (~0.25 GB)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials.html#rosettafold-all-atom-checkpoint-1-5-gb">RoseTTAFold-All-Atom checkpoint (~1.5 GB)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../data_preparation.html">How to prepare <cite>MULTICOM_ligand</cite> data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../available_methods.html">Available inference methods</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../method_inference.html">How to run inference with individual methods</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../ensemble_inference.html">How to run inference with a method ensemble</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../ensemble_inference.html#note-the-suffixes-for-both-output-dir-and-ensemble-benchmarking-repeat-index-should-be-modified-to-e-g-2-3">NOTE: the suffixes for both <code class="docutils literal notranslate"><span class="pre">output_dir</span></code> and <code class="docutils literal notranslate"><span class="pre">ensemble_benchmarking_repeat_index</span></code> should be modified to e.g., 2, 3, …</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../ensemble_inference.html#benchmark-using-the-posebusters-benchmark-dataset-e-g-after-generating-40-complexes-per-target-with-each-method">benchmark using the PoseBusters Benchmark dataset e.g., after generating 40 complexes per target with each method</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../ensemble_inference.html#benchmark-using-the-astex-diverse-dataset-e-g-after-generating-40-complexes-per-target-with-each-method">benchmark using the Astex Diverse dataset e.g., after generating 40 complexes per target with each method</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../ensemble_inference.html#benchmark-using-the-dockgen-dataset-e-g-after-generating-40-complexes-per-target-with-each-method">benchmark using the DockGen dataset e.g., after generating 40 complexes per target with each method</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../ensemble_inference.html#benchmark-using-the-casp15-dataset-e-g-after-generating-40-complexes-per-target-with-each-method">benchmark using the CASP15 dataset e.g., after generating 40 complexes per target with each method</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../ensemble_inference.html#analyze-benchmarking-results-for-the-posebusters-benchmark-dataset">analyze benchmarking results for the PoseBusters Benchmark dataset</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../comparative_plots.html">How to create comparative plots of inference results</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../comparative_plots.html#analyze-benchmarking-results-for-the-casp15-dataset">analyze benchmarking results for the CASP15 dataset</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../for_developers.html">For developers</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../acknowledgements.html">Acknowledgements</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../citing_this_work.html">Citing this work</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../bonus.html">Bonus</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Default Configs</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../configs/analysis.html">Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../configs/data.html">Data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../configs/model.html">Model</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API Reference</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/multicom_ligand.binding_site_crop_preparation.html">Binding site crop preparation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/multicom_ligand.complex_alignment.html">Complex alignment</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/multicom_ligand.inference_relaxation.html">Inference relaxation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/multicom_ligand.minimize_energy.html">Minimize energy</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/multicom_ligand.ensemble_generation.html">Ensemble generation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/multicom_ligand.data_utils.html">Data utilities</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/multicom_ligand.model_utils.html">Model utilities</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/multicom_ligand.utils.html">General utilities</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/multicom_ligand.resolvers.html">OmegaConf resolvers</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
              <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon no-toc" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main" id="furo-main-content">
          <h1>Source code for multicom_ligand.utils.data_utils</h1><div class="highlight"><pre>
<span></span><span class="c1"># -------------------------------------------------------------------------------------------------------------------------------------</span>
<span class="c1"># Following code curated for MULTICOM_ligand: (https://github.com/BioinfoMachineLearning/MULTICOM_ligand)</span>
<span class="c1"># -------------------------------------------------------------------------------------------------------------------------------------</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">glob</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">logging</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">shutil</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">subprocess</span>  <span class="c1"># nosec</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">warnings</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">io</span><span class="w"> </span><span class="kn">import</span> <span class="n">StringIO</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pypdb</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">rootutils</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">beartype</span><span class="w"> </span><span class="kn">import</span> <span class="n">beartype</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">beartype.typing</span><span class="w"> </span><span class="kn">import</span> <span class="n">Any</span><span class="p">,</span> <span class="n">List</span><span class="p">,</span> <span class="n">Literal</span><span class="p">,</span> <span class="n">Optional</span><span class="p">,</span> <span class="n">Set</span><span class="p">,</span> <span class="n">Tuple</span><span class="p">,</span> <span class="n">Union</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">Bio</span><span class="w"> </span><span class="kn">import</span> <span class="n">PDB</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">Bio.PDB</span><span class="w"> </span><span class="kn">import</span> <span class="n">PDBParser</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">Bio.PDB.Structure</span><span class="w"> </span><span class="kn">import</span> <span class="n">Structure</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">biopandas.pdb</span><span class="w"> </span><span class="kn">import</span> <span class="n">PandasPdb</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">prody</span><span class="w"> </span><span class="kn">import</span> <span class="n">parsePDB</span><span class="p">,</span> <span class="n">writePDB</span><span class="p">,</span> <span class="n">writePDBStream</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">rdkit</span><span class="w"> </span><span class="kn">import</span> <span class="n">Chem</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">rdkit.Chem</span><span class="w"> </span><span class="kn">import</span> <span class="n">AllChem</span>

<span class="n">rootutils</span><span class="o">.</span><span class="n">setup_root</span><span class="p">(</span><span class="vm">__file__</span><span class="p">,</span> <span class="n">indicator</span><span class="o">=</span><span class="s2">&quot;.project-root&quot;</span><span class="p">,</span> <span class="n">pythonpath</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="nb">format</span><span class="o">=</span><span class="s2">&quot;[</span><span class="si">%(asctime)s</span><span class="s2">] {</span><span class="si">%(filename)s</span><span class="s2">:</span><span class="si">%(lineno)d</span><span class="s2">} </span><span class="si">%(levelname)s</span><span class="s2"> - </span><span class="si">%(message)s</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span>


<span class="n">DNA_THREE_TO_ONE</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;DA&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span>
    <span class="s2">&quot;DC&quot;</span><span class="p">:</span> <span class="s2">&quot;C&quot;</span><span class="p">,</span>
    <span class="s2">&quot;DG&quot;</span><span class="p">:</span> <span class="s2">&quot;G&quot;</span><span class="p">,</span>
    <span class="s2">&quot;DT&quot;</span><span class="p">:</span> <span class="s2">&quot;T&quot;</span><span class="p">,</span>
<span class="p">}</span>
<span class="n">RNA_THREE_TO_ONE</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;A&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span>
    <span class="s2">&quot;C&quot;</span><span class="p">:</span> <span class="s2">&quot;C&quot;</span><span class="p">,</span>
    <span class="s2">&quot;G&quot;</span><span class="p">:</span> <span class="s2">&quot;G&quot;</span><span class="p">,</span>
    <span class="s2">&quot;U&quot;</span><span class="p">:</span> <span class="s2">&quot;U&quot;</span><span class="p">,</span>
<span class="p">}</span>
<span class="n">PROTEIN_THREE_TO_ONE</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;ALA&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ARG&quot;</span><span class="p">:</span> <span class="s2">&quot;R&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ASN&quot;</span><span class="p">:</span> <span class="s2">&quot;N&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ASP&quot;</span><span class="p">:</span> <span class="s2">&quot;D&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CYS&quot;</span><span class="p">:</span> <span class="s2">&quot;C&quot;</span><span class="p">,</span>
    <span class="s2">&quot;GLN&quot;</span><span class="p">:</span> <span class="s2">&quot;Q&quot;</span><span class="p">,</span>
    <span class="s2">&quot;GLU&quot;</span><span class="p">:</span> <span class="s2">&quot;E&quot;</span><span class="p">,</span>
    <span class="s2">&quot;GLY&quot;</span><span class="p">:</span> <span class="s2">&quot;G&quot;</span><span class="p">,</span>
    <span class="s2">&quot;HIS&quot;</span><span class="p">:</span> <span class="s2">&quot;H&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ILE&quot;</span><span class="p">:</span> <span class="s2">&quot;I&quot;</span><span class="p">,</span>
    <span class="s2">&quot;LEU&quot;</span><span class="p">:</span> <span class="s2">&quot;L&quot;</span><span class="p">,</span>
    <span class="s2">&quot;LYS&quot;</span><span class="p">:</span> <span class="s2">&quot;K&quot;</span><span class="p">,</span>
    <span class="s2">&quot;MET&quot;</span><span class="p">:</span> <span class="s2">&quot;M&quot;</span><span class="p">,</span>
    <span class="s2">&quot;MSE&quot;</span><span class="p">:</span> <span class="s2">&quot;M&quot;</span><span class="p">,</span>  <span class="c1"># NOTE: this is almost the same amino acid as `MET`; the sulfur is just replaced by `Selen`</span>
    <span class="s2">&quot;PHE&quot;</span><span class="p">:</span> <span class="s2">&quot;F&quot;</span><span class="p">,</span>
    <span class="s2">&quot;PRO&quot;</span><span class="p">:</span> <span class="s2">&quot;P&quot;</span><span class="p">,</span>
    <span class="s2">&quot;PYL&quot;</span><span class="p">:</span> <span class="s2">&quot;O&quot;</span><span class="p">,</span>
    <span class="s2">&quot;SER&quot;</span><span class="p">:</span> <span class="s2">&quot;S&quot;</span><span class="p">,</span>
    <span class="s2">&quot;SEC&quot;</span><span class="p">:</span> <span class="s2">&quot;U&quot;</span><span class="p">,</span>
    <span class="s2">&quot;THR&quot;</span><span class="p">:</span> <span class="s2">&quot;T&quot;</span><span class="p">,</span>
    <span class="s2">&quot;TRP&quot;</span><span class="p">:</span> <span class="s2">&quot;W&quot;</span><span class="p">,</span>
    <span class="s2">&quot;TYR&quot;</span><span class="p">:</span> <span class="s2">&quot;Y&quot;</span><span class="p">,</span>
    <span class="s2">&quot;VAL&quot;</span><span class="p">:</span> <span class="s2">&quot;V&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ASX&quot;</span><span class="p">:</span> <span class="s2">&quot;B&quot;</span><span class="p">,</span>
    <span class="s2">&quot;GLX&quot;</span><span class="p">:</span> <span class="s2">&quot;Z&quot;</span><span class="p">,</span>
    <span class="s2">&quot;XAA&quot;</span><span class="p">:</span> <span class="s2">&quot;X&quot;</span><span class="p">,</span>
    <span class="s2">&quot;XLE&quot;</span><span class="p">:</span> <span class="s2">&quot;J&quot;</span><span class="p">,</span>
<span class="p">}</span>


<div class="viewcode-block" id="read_molecule">
<a class="viewcode-back" href="../../../modules/multicom_ligand.data_utils.html#multicom_ligand.utils.data_utils.read_molecule">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">read_molecule</span><span class="p">(</span>
    <span class="n">molecule_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">sanitize</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">calc_charges</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">remove_hs</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Chem</span><span class="o">.</span><span class="n">Mol</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Load an RDKit molecule from a given filepath.&quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">molecule_file</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.mol2&quot;</span><span class="p">):</span>
        <span class="n">mol</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolFromMol2File</span><span class="p">(</span><span class="n">molecule_file</span><span class="p">,</span> <span class="n">sanitize</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">removeHs</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">molecule_file</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.mol&quot;</span><span class="p">):</span>
        <span class="n">mol</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolFromMolFile</span><span class="p">(</span><span class="n">molecule_file</span><span class="p">,</span> <span class="n">sanitize</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">removeHs</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">molecule_file</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.sdf&quot;</span><span class="p">):</span>
        <span class="n">supplier</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">SDMolSupplier</span><span class="p">(</span><span class="n">molecule_file</span><span class="p">,</span> <span class="n">sanitize</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">removeHs</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">mol</span> <span class="o">=</span> <span class="n">supplier</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">elif</span> <span class="n">molecule_file</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.pdbqt&quot;</span><span class="p">):</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">molecule_file</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
            <span class="n">pdbqt_data</span> <span class="o">=</span> <span class="n">file</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
        <span class="n">pdb_block</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">pdbqt_data</span><span class="p">:</span>
            <span class="n">pdb_block</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">line</span><span class="p">[:</span><span class="mi">66</span><span class="p">]</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="n">mol</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolFromPDBBlock</span><span class="p">(</span><span class="n">pdb_block</span><span class="p">,</span> <span class="n">sanitize</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">removeHs</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">molecule_file</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.pdb&quot;</span><span class="p">):</span>
        <span class="n">mol</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolFromPDBFile</span><span class="p">(</span><span class="n">molecule_file</span><span class="p">,</span> <span class="n">sanitize</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">removeHs</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="s2">&quot;Expected the format of `molecule_file` to be &quot;</span>
            <span class="sa">f</span><span class="s2">&quot;one of `.mol`, `.mol2`, `.sdf`, `.pdbqt`, or `.pdb`, yet got </span><span class="si">{</span><span class="n">molecule_file</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="p">)</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">sanitize</span> <span class="ow">or</span> <span class="n">calc_charges</span><span class="p">:</span>
            <span class="n">Chem</span><span class="o">.</span><span class="n">SanitizeMol</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">calc_charges</span><span class="p">:</span>
            <span class="c1"># compute Gasteiger charges on the molecule</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">AllChem</span><span class="o">.</span><span class="n">ComputeGasteigerCharges</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
                <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s2">&quot;Unable to compute charges for the input molecule.&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">remove_hs</span><span class="p">:</span>
            <span class="n">mol</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">RemoveHs</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="n">sanitize</span><span class="o">=</span><span class="n">sanitize</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;RDKit was unable to read the molecule due to the error: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span>

    <span class="k">return</span> <span class="n">mol</span></div>



<div class="viewcode-block" id="parse_inference_inputs_from_dir">
<a class="viewcode-back" href="../../../modules/multicom_ligand.data_utils.html#multicom_ligand.utils.data_utils.parse_inference_inputs_from_dir">[docs]</a>
<span class="nd">@beartype</span>
<span class="k">def</span><span class="w"> </span><span class="nf">parse_inference_inputs_from_dir</span><span class="p">(</span>
    <span class="n">input_data_dir</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Path</span><span class="p">],</span> <span class="n">pdb_ids</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Set</span><span class="p">[</span><span class="n">Any</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Parse a data directory containing subdirectories of protein-ligand</span>
<span class="sd">    complexes and return corresponding SMILES strings and PDB IDs.</span>

<span class="sd">    :param input_data_dir: Path to the input data directory.</span>
<span class="sd">    :param pdb_ids: Optional set of IDs by which to filter processing.</span>
<span class="sd">    :return: A list of tuples each containing a SMILES string and a PDB</span>
<span class="sd">        ID.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">smiles_and_pdb_id_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">casp_dataset_requested</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">input_data_dir</span><span class="p">)</span> <span class="o">==</span> <span class="s2">&quot;targets&quot;</span>
    <span class="k">if</span> <span class="n">casp_dataset_requested</span><span class="p">:</span>
        <span class="c1"># parse CASP inputs uniquely</span>
        <span class="n">smiles_filepaths</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">input_data_dir</span><span class="p">,</span> <span class="s2">&quot;*.smiles.txt&quot;</span><span class="p">)))</span>
        <span class="k">for</span> <span class="n">smiles_filepath</span> <span class="ow">in</span> <span class="n">smiles_filepaths</span><span class="p">:</span>
            <span class="n">pdb_id</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">smiles_filepath</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">smiles_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">smiles_filepath</span><span class="p">,</span> <span class="n">delim_whitespace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="k">assert</span> <span class="n">smiles_df</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span> <span class="o">==</span> <span class="p">[</span>
                <span class="s2">&quot;ID&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Name&quot;</span><span class="p">,</span>
                <span class="s2">&quot;SMILES&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Relevant&quot;</span><span class="p">,</span>
            <span class="p">],</span> <span class="s2">&quot;SMILES DataFrame must have columns [&#39;ID&#39;, &#39;Name&#39;, &#39;SMILES&#39;, &#39;Relevant&#39;].&quot;</span>
            <span class="n">mol_smiles</span> <span class="o">=</span> <span class="s2">&quot;|&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">smiles_df</span><span class="p">[</span><span class="s2">&quot;SMILES&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
            <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">mol_smiles</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;SMILES string for </span><span class="si">{</span><span class="n">pdb_id</span><span class="si">}</span><span class="s2"> cannot be empty.&quot;</span>
            <span class="n">smiles_and_pdb_id_list</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">mol_smiles</span><span class="p">,</span> <span class="n">pdb_id</span><span class="p">))</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">pdb_name</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">input_data_dir</span><span class="p">):</span>
            <span class="k">if</span> <span class="nb">any</span><span class="p">(</span><span class="n">substr</span> <span class="ow">in</span> <span class="n">pdb_name</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="k">for</span> <span class="n">substr</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;sequence&quot;</span><span class="p">,</span> <span class="s2">&quot;structure&quot;</span><span class="p">]):</span>
                <span class="c1"># e.g., skip ESMFold sequence files and structure directories</span>
                <span class="k">continue</span>
            <span class="k">if</span> <span class="n">pdb_ids</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">pdb_name</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">pdb_ids</span><span class="p">:</span>
                <span class="c1"># e.g., skip PoseBusters Benchmark PDBs that contain crystal contacts</span>
                <span class="c1"># reference: https://github.com/maabuu/posebusters/issues/26</span>
                <span class="k">continue</span>
            <span class="n">pdb_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">input_data_dir</span><span class="p">,</span> <span class="n">pdb_name</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="n">pdb_dir</span><span class="p">):</span>
                <span class="n">mol</span> <span class="o">=</span> <span class="kc">None</span>
                <span class="n">pdb_id</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">pdb_dir</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
                <span class="c1"># NOTE: we first try to parse `.sdf` files if they exist for the current dataset</span>
                <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">pdb_dir</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">pdb_id</span><span class="si">}</span><span class="s2">_ligand.sdf&quot;</span><span class="p">)):</span>
                    <span class="n">mol</span> <span class="o">=</span> <span class="n">read_molecule</span><span class="p">(</span>
                        <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">pdb_dir</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">pdb_id</span><span class="si">}</span><span class="s2">_ligand.sdf&quot;</span><span class="p">),</span>
                        <span class="n">remove_hs</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                        <span class="n">sanitize</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                    <span class="p">)</span>
                <span class="c1"># NOTE: DockGen uses `.pdb` files to store its ligands</span>
                <span class="k">if</span> <span class="n">mol</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">pdb_dir</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">pdb_id</span><span class="si">}</span><span class="s2">_ligand.pdb&quot;</span><span class="p">)):</span>
                    <span class="n">mol</span> <span class="o">=</span> <span class="n">read_molecule</span><span class="p">(</span>
                        <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">pdb_dir</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">pdb_id</span><span class="si">}</span><span class="s2">_ligand.pdb&quot;</span><span class="p">),</span>
                        <span class="n">remove_hs</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                        <span class="n">sanitize</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                    <span class="p">)</span>
                    <span class="k">if</span> <span class="n">mol</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                        <span class="n">mol</span> <span class="o">=</span> <span class="n">read_molecule</span><span class="p">(</span>
                            <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">pdb_dir</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">pdb_id</span><span class="si">}</span><span class="s2">_ligand.pdb&quot;</span><span class="p">),</span>
                            <span class="n">remove_hs</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                            <span class="n">sanitize</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                        <span class="p">)</span>
                <span class="k">if</span> <span class="n">mol</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No ligand file found for PDB ID </span><span class="si">{</span><span class="n">pdb_id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">mol_smiles</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolToSmiles</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">mol_smiles</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">smiles_and_pdb_id_list</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">mol_smiles</span><span class="p">,</span> <span class="n">pdb_id</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">smiles_and_pdb_id_list</span></div>



<div class="viewcode-block" id="extract_sequences_from_macromolecule_structure_file">
<a class="viewcode-back" href="../../../modules/multicom_ligand.data_utils.html#multicom_ligand.utils.data_utils.extract_sequences_from_macromolecule_structure_file">[docs]</a>
<span class="nd">@beartype</span>
<span class="k">def</span><span class="w"> </span><span class="nf">extract_sequences_from_macromolecule_structure_file</span><span class="p">(</span>
    <span class="n">pdb_filepath</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Path</span><span class="p">],</span> <span class="n">structure</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Structure</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">List</span><span class="p">[</span><span class="n">Literal</span><span class="p">[</span><span class="s2">&quot;DNA&quot;</span><span class="p">,</span> <span class="s2">&quot;RNA&quot;</span><span class="p">,</span> <span class="s2">&quot;Protein&quot;</span><span class="p">]]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Extract the chain sequences from a macromolecule structure file.</span>

<span class="sd">    :param pdb_filepath: Path to the PDB structure file.</span>
<span class="sd">    :param structure: Optional BioPython structure object to use instead.</span>
<span class="sd">    :return: A list of macromolecule sequences and their corresponding molecule chain types (e.g., `Protein`).</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">biopython_parser</span> <span class="o">=</span> <span class="n">PDBParser</span><span class="p">()</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">structure</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">structure</span> <span class="o">=</span> <span class="n">biopython_parser</span><span class="o">.</span><span class="n">get_structure</span><span class="p">(</span><span class="s2">&quot;random_id&quot;</span><span class="p">,</span> <span class="n">pdb_filepath</span><span class="p">)</span>
        <span class="n">structure</span> <span class="o">=</span> <span class="n">structure</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Due to exception </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">, could not parse macromolecule </span><span class="si">{</span><span class="n">pdb_filepath</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span>

    <span class="n">sequences</span><span class="p">,</span> <span class="n">sequence_types</span> <span class="o">=</span> <span class="p">[],</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">chain</span> <span class="ow">in</span> <span class="n">structure</span><span class="p">:</span>
        <span class="n">seq</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="k">for</span> <span class="n">residue</span> <span class="ow">in</span> <span class="n">chain</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">residue</span><span class="o">.</span><span class="n">get_resname</span><span class="p">()</span> <span class="o">==</span> <span class="s2">&quot;HOH&quot;</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="k">if</span> <span class="n">residue</span><span class="o">.</span><span class="n">get_resname</span><span class="p">()</span> <span class="ow">in</span> <span class="n">DNA_THREE_TO_ONE</span><span class="p">:</span>
                <span class="n">seq</span> <span class="o">+=</span> <span class="n">DNA_THREE_TO_ONE</span><span class="p">[</span><span class="n">residue</span><span class="o">.</span><span class="n">get_resname</span><span class="p">()]</span>
                <span class="n">sequence_types</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;DNA&quot;</span><span class="p">)</span>
                <span class="k">continue</span>
            <span class="k">if</span> <span class="n">residue</span><span class="o">.</span><span class="n">get_resname</span><span class="p">()</span> <span class="ow">in</span> <span class="n">RNA_THREE_TO_ONE</span><span class="p">:</span>
                <span class="n">seq</span> <span class="o">+=</span> <span class="n">RNA_THREE_TO_ONE</span><span class="p">[</span><span class="n">residue</span><span class="o">.</span><span class="n">get_resname</span><span class="p">()]</span>
                <span class="n">sequence_types</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;RNA&quot;</span><span class="p">)</span>
                <span class="k">continue</span>
            <span class="n">c_alpha</span><span class="p">,</span> <span class="n">n</span><span class="p">,</span> <span class="n">c</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span>
            <span class="k">for</span> <span class="n">atom</span> <span class="ow">in</span> <span class="n">residue</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">atom</span><span class="o">.</span><span class="n">name</span> <span class="o">==</span> <span class="s2">&quot;CA&quot;</span><span class="p">:</span>
                    <span class="n">c_alpha</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">atom</span><span class="o">.</span><span class="n">get_vector</span><span class="p">())</span>
                <span class="k">if</span> <span class="n">atom</span><span class="o">.</span><span class="n">name</span> <span class="o">==</span> <span class="s2">&quot;N&quot;</span><span class="p">:</span>
                    <span class="n">n</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">atom</span><span class="o">.</span><span class="n">get_vector</span><span class="p">())</span>
                <span class="k">if</span> <span class="n">atom</span><span class="o">.</span><span class="n">name</span> <span class="o">==</span> <span class="s2">&quot;C&quot;</span><span class="p">:</span>
                    <span class="n">c</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">atom</span><span class="o">.</span><span class="n">get_vector</span><span class="p">())</span>
            <span class="k">if</span> <span class="p">(</span>
                <span class="n">c_alpha</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">n</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">c</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
            <span class="p">):</span>  <span class="c1"># only append residue if it is an amino acid and not a weird residue-like entity</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">seq</span> <span class="o">+=</span> <span class="n">PROTEIN_THREE_TO_ONE</span><span class="p">[</span><span class="n">residue</span><span class="o">.</span><span class="n">get_resname</span><span class="p">()]</span>
                    <span class="n">sequence_types</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Protein&quot;</span><span class="p">)</span>
                <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                    <span class="n">seq</span> <span class="o">+=</span> <span class="s2">&quot;-&quot;</span>
                    <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;Due to exception </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">, encountered unknown amino acid </span><span class="si">{</span><span class="n">residue</span><span class="o">.</span><span class="n">get_resname</span><span class="p">()</span><span class="si">}</span><span class="s2"> in the macromolecule </span><span class="si">{</span><span class="n">pdb_filepath</span><span class="si">}</span><span class="s2">. Replacing it with a dash (i.e., `-`).&quot;</span>
                    <span class="p">)</span>
        <span class="n">sequences</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">sequences</span><span class="p">,</span> <span class="n">sequence_types</span></div>



<div class="viewcode-block" id="combine_molecules">
<a class="viewcode-back" href="../../../modules/multicom_ligand.data_utils.html#multicom_ligand.utils.data_utils.combine_molecules">[docs]</a>
<span class="nd">@beartype</span>
<span class="k">def</span><span class="w"> </span><span class="nf">combine_molecules</span><span class="p">(</span><span class="n">molecule_list</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">Chem</span><span class="o">.</span><span class="n">Mol</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">Chem</span><span class="o">.</span><span class="n">Mol</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Combine a list of RDKit molecules into a single molecule.</span>

<span class="sd">    :param molecule_list: A list of RDKit molecules.</span>
<span class="sd">    :return: A single RDKit molecule.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Initialize the combined molecule with the first molecule in the list</span>
    <span class="n">new_mol</span> <span class="o">=</span> <span class="n">molecule_list</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

    <span class="c1"># Iterate through the remaining molecules and combine them pairwise</span>
    <span class="k">for</span> <span class="n">mol</span> <span class="ow">in</span> <span class="n">molecule_list</span><span class="p">[</span><span class="mi">1</span><span class="p">:]:</span>
        <span class="n">new_mol</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">CombineMols</span><span class="p">(</span><span class="n">new_mol</span><span class="p">,</span> <span class="n">mol</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">new_mol</span></div>



<div class="viewcode-block" id="renumber_pdb_df_residues">
<a class="viewcode-back" href="../../../modules/multicom_ligand.data_utils.html#multicom_ligand.utils.data_utils.renumber_pdb_df_residues">[docs]</a>
<span class="nd">@beartype</span>
<span class="k">def</span><span class="w"> </span><span class="nf">renumber_pdb_df_residues</span><span class="p">(</span><span class="n">input_pdb_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">output_pdb_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Renumber residues in a PDB file starting from 1 for each chain.</span>

<span class="sd">    :param input_pdb_file: Path to the input PDB file.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Load the PDB file</span>
    <span class="n">pdb</span> <span class="o">=</span> <span class="n">PandasPdb</span><span class="p">()</span><span class="o">.</span><span class="n">read_pdb</span><span class="p">(</span><span class="n">input_pdb_file</span><span class="p">)</span>

    <span class="c1"># Iterate through each chain</span>
    <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">chain_df</span> <span class="ow">in</span> <span class="n">pdb</span><span class="o">.</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;ATOM&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;chain_id&quot;</span><span class="p">):</span>
        <span class="c1"># Get the minimum residue index for the current chain</span>
        <span class="n">min_residue_index</span> <span class="o">=</span> <span class="n">chain_df</span><span class="p">[</span><span class="s2">&quot;residue_number&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">()</span>

        <span class="c1"># Reindex the residues starting from 1</span>
        <span class="n">chain_df</span><span class="p">[</span><span class="s2">&quot;residue_number&quot;</span><span class="p">]</span> <span class="o">-=</span> <span class="n">min_residue_index</span> <span class="o">-</span> <span class="mi">1</span>

        <span class="c1"># Update the PDB dataframe with the reindexed chain</span>
        <span class="n">pdb</span><span class="o">.</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;ATOM&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">chain_df</span><span class="o">.</span><span class="n">index</span><span class="p">]</span> <span class="o">=</span> <span class="n">chain_df</span>

    <span class="c1"># Write the modified PDB file</span>
    <span class="n">pdb</span><span class="o">.</span><span class="n">to_pdb</span><span class="p">(</span><span class="n">output_pdb_file</span><span class="p">)</span></div>



<div class="viewcode-block" id="renumber_biopython_structure_residues">
<a class="viewcode-back" href="../../../modules/multicom_ligand.data_utils.html#multicom_ligand.utils.data_utils.renumber_biopython_structure_residues">[docs]</a>
<span class="nd">@beartype</span>
<span class="k">def</span><span class="w"> </span><span class="nf">renumber_biopython_structure_residues</span><span class="p">(</span>
    <span class="n">structure</span><span class="p">:</span> <span class="n">Structure</span><span class="p">,</span> <span class="n">gap_insertion_point</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Structure</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Renumber residues in a PDB file using BioPython starting from 1 for each</span>
<span class="sd">    chain.</span>

<span class="sd">    :param structure: BioPython structure object.</span>
<span class="sd">    :param gap_insertion_point: Optional `:`-separated string</span>
<span class="sd">        representing the chain-residue pair index of the residue at</span>
<span class="sd">        which to insert a single index gap.</span>
<span class="sd">    :return: BioPython structure object with renumbered residues.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Iterate through each model in the structure</span>
    <span class="k">if</span> <span class="n">gap_insertion_point</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">assert</span> <span class="p">(</span>
            <span class="nb">len</span><span class="p">(</span><span class="n">gap_insertion_point</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;:&quot;</span><span class="p">))</span> <span class="o">==</span> <span class="mi">2</span>
        <span class="p">),</span> <span class="s2">&quot;When provided, gap insertion point must be in the format &#39;chain_id:residue_index&#39;.&quot;</span>
    <span class="n">gap_insertion_chain_id</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">gap_insertion_point</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;:&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="k">if</span> <span class="n">gap_insertion_point</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">else</span> <span class="kc">None</span>
    <span class="p">)</span>
    <span class="n">gap_insertion_residue_index</span> <span class="o">=</span> <span class="p">(</span>
        <span class="nb">int</span><span class="p">(</span><span class="n">gap_insertion_point</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;:&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">])</span> <span class="k">if</span> <span class="n">gap_insertion_point</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">else</span> <span class="kc">None</span>
    <span class="p">)</span>
    <span class="k">for</span> <span class="n">model</span> <span class="ow">in</span> <span class="n">structure</span><span class="p">:</span>
        <span class="c1"># Iterate through each chain in the model</span>
        <span class="k">for</span> <span class="n">chain</span> <span class="ow">in</span> <span class="n">model</span><span class="p">:</span>
            <span class="c1"># Get the minimum residue index for the current chain</span>
            <span class="n">min_residue_index</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">residue</span><span class="o">.</span><span class="n">id</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">residue</span> <span class="ow">in</span> <span class="n">chain</span><span class="p">)</span>

            <span class="c1"># Reindex the residues starting from 1</span>
            <span class="n">gap_insertion_counter</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">for</span> <span class="n">residue</span> <span class="ow">in</span> <span class="n">chain</span><span class="p">:</span>
                <span class="n">new_residue_index</span> <span class="o">=</span> <span class="n">residue</span><span class="o">.</span><span class="n">id</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">-</span> <span class="n">min_residue_index</span> <span class="o">+</span> <span class="mi">1</span>
                <span class="n">gap_index_found</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="n">gap_insertion_chain_id</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
                    <span class="ow">and</span> <span class="n">gap_insertion_residue_index</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
                    <span class="ow">and</span> <span class="n">chain</span><span class="o">.</span><span class="n">id</span> <span class="o">==</span> <span class="n">gap_insertion_chain_id</span>
                    <span class="ow">and</span> <span class="n">new_residue_index</span> <span class="o">==</span> <span class="n">gap_insertion_residue_index</span>
                <span class="p">)</span>
                <span class="k">if</span> <span class="n">gap_index_found</span><span class="p">:</span>
                    <span class="n">gap_insertion_counter</span> <span class="o">=</span> <span class="mi">1</span>
                <span class="n">residue</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">,</span> <span class="n">new_residue_index</span> <span class="o">+</span> <span class="n">gap_insertion_counter</span><span class="p">,</span> <span class="n">residue</span><span class="o">.</span><span class="n">id</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
                <span class="k">for</span> <span class="n">atom</span> <span class="ow">in</span> <span class="n">residue</span><span class="p">:</span>
                    <span class="n">atom</span><span class="o">.</span><span class="n">serial_number</span> <span class="o">=</span> <span class="kc">None</span>  <span class="c1"># Reset atom serial number</span>

    <span class="k">return</span> <span class="n">structure</span></div>



<div class="viewcode-block" id="get_pdb_components_with_prody">
<a class="viewcode-back" href="../../../modules/multicom_ligand.data_utils.html#multicom_ligand.utils.data_utils.get_pdb_components_with_prody">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">get_pdb_components_with_prody</span><span class="p">(</span><span class="n">pdb_id</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Split a protein-ligand pdb into protein and ligand components using</span>
<span class="sd">    ProDy.</span>

<span class="sd">    :param pdb_id: PDB ID</span>
<span class="sd">    :return: protein structure, nucleic acid structure, and ligand</span>
<span class="sd">        residues</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">pdb</span> <span class="o">=</span> <span class="n">parsePDB</span><span class="p">(</span><span class="n">pdb_id</span><span class="p">)</span>
    <span class="n">protein</span> <span class="o">=</span> <span class="n">pdb</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;protein&quot;</span><span class="p">)</span>
    <span class="n">nucleic_acid</span> <span class="o">=</span> <span class="n">pdb</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;nucleotide&quot;</span><span class="p">)</span>
    <span class="n">ligand</span> <span class="o">=</span> <span class="n">pdb</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;not (protein or nucleotide or water)&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">protein</span><span class="p">,</span> <span class="n">nucleic_acid</span><span class="p">,</span> <span class="n">ligand</span></div>



<div class="viewcode-block" id="write_pdb_with_prody">
<a class="viewcode-back" href="../../../modules/multicom_ligand.data_utils.html#multicom_ligand.utils.data_utils.write_pdb_with_prody">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">write_pdb_with_prody</span><span class="p">(</span><span class="n">macromolecule</span><span class="p">,</span> <span class="n">pdb_name</span><span class="p">,</span> <span class="n">add_element_types</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Write a protein or nucleic acid structure to a pdb file using ProDy.</span>

<span class="sd">    :param macromolecule: protein or nucleic acid object from prody</span>
<span class="sd">    :param pdb_name: base name for the pdb file</span>
<span class="sd">    :param add_element_types: whether to add element types to the pdb</span>
<span class="sd">        file</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">writePDB</span><span class="p">(</span><span class="n">pdb_name</span><span class="p">,</span> <span class="n">macromolecule</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">add_element_types</span><span class="p">:</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">pdb_name</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;.pdb&quot;</span><span class="p">,</span> <span class="s2">&quot;_elem.pdb&quot;</span><span class="p">),</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>  <span class="c1"># nosec</span>
                <span class="sa">f</span><span class="s2">&quot;pdb_element </span><span class="si">{</span><span class="n">pdb_name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
                <span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="n">check</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="n">stdout</span><span class="o">=</span><span class="n">f</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="n">shutil</span><span class="o">.</span><span class="n">move</span><span class="p">(</span><span class="n">pdb_name</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;.pdb&quot;</span><span class="p">,</span> <span class="s2">&quot;_elem.pdb&quot;</span><span class="p">),</span> <span class="n">pdb_name</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Wrote </span><span class="si">{</span><span class="n">pdb_name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span></div>



<div class="viewcode-block" id="process_ligand_with_prody">
<a class="viewcode-back" href="../../../modules/multicom_ligand.data_utils.html#multicom_ligand.utils.data_utils.process_ligand_with_prody">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">process_ligand_with_prody</span><span class="p">(</span>
    <span class="n">ligand</span><span class="p">,</span>
    <span class="n">res_name</span><span class="p">,</span>
    <span class="n">chain</span><span class="p">,</span>
    <span class="n">resnum</span><span class="p">,</span>
    <span class="n">sanitize</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">sub_smiles</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Chem</span><span class="o">.</span><span class="n">Mol</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Add bond orders to a pdb ligand using ProDy.</span>
<span class="sd">    1. Select the ligand component with name &quot;res_name&quot;</span>
<span class="sd">    2. Get the corresponding SMILES from pypdb</span>
<span class="sd">    3. Create a template molecule from the SMILES in step 2</span>
<span class="sd">    4. Write the PDB file to a stream</span>
<span class="sd">    5. Read the stream into an RDKit molecule</span>
<span class="sd">    6. Assign the bond orders from the template from step 3</span>

<span class="sd">    :param ligand: ligand as generated by prody</span>
<span class="sd">    :param res_name: residue name of ligand to extract</span>
<span class="sd">    :param chain: chain of ligand to extract</span>
<span class="sd">    :param resnum: residue number of ligand to extract</span>
<span class="sd">    :param sanitize: whether to sanitize the molecule</span>
<span class="sd">    :param sub_smiles: optional SMILES string of the ligand molecule</span>
<span class="sd">    :return: molecule with bond orders assigned</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">sub_smiles_provided</span> <span class="o">=</span> <span class="n">sub_smiles</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>

    <span class="n">output</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">()</span>
    <span class="n">sub_mol</span> <span class="o">=</span> <span class="n">ligand</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;resname </span><span class="si">{</span><span class="n">res_name</span><span class="si">}</span><span class="s2"> and chain </span><span class="si">{</span><span class="n">chain</span><span class="si">}</span><span class="s2"> and resnum </span><span class="si">{</span><span class="n">resnum</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">chem_desc</span> <span class="o">=</span> <span class="n">pypdb</span><span class="o">.</span><span class="n">describe_chemical</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">res_name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">chem_desc</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">sub_smiles_provided</span><span class="p">:</span>
        <span class="n">sub_smiles</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">chem_desc</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;pdbx_chem_comp_descriptor&quot;</span><span class="p">,</span> <span class="p">[]):</span>
            <span class="k">if</span> <span class="n">item</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;type&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="s2">&quot;SMILES&quot;</span><span class="p">:</span>
                <span class="n">sub_smiles</span> <span class="o">=</span> <span class="n">item</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;descriptor&quot;</span><span class="p">)</span>
                <span class="k">break</span>

    <span class="k">if</span> <span class="n">sub_smiles</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">template</span> <span class="o">=</span> <span class="n">AllChem</span><span class="o">.</span><span class="n">MolFromSmiles</span><span class="p">(</span><span class="n">sub_smiles</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">template</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="n">writePDBStream</span><span class="p">(</span><span class="n">output</span><span class="p">,</span> <span class="n">sub_mol</span><span class="p">)</span>
    <span class="n">pdb_string</span> <span class="o">=</span> <span class="n">output</span><span class="o">.</span><span class="n">getvalue</span><span class="p">()</span>
    <span class="n">rd_mol</span> <span class="o">=</span> <span class="n">AllChem</span><span class="o">.</span><span class="n">MolFromPDBBlock</span><span class="p">(</span><span class="n">pdb_string</span><span class="p">,</span> <span class="n">sanitize</span><span class="o">=</span><span class="n">sanitize</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">sub_smiles_provided</span> <span class="ow">and</span> <span class="n">template</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="c1"># Ensure the input ligand perfectly matches the template ligand</span>
        <span class="k">assert</span> <span class="p">(</span>
            <span class="n">rd_mol</span><span class="o">.</span><span class="n">GetNumAtoms</span><span class="p">()</span> <span class="o">==</span> <span class="n">template</span><span class="o">.</span><span class="n">GetNumAtoms</span><span class="p">()</span>
        <span class="p">),</span> <span class="s2">&quot;Number of atoms in both molecules is different.&quot;</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">template</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">new_mol</span> <span class="o">=</span> <span class="n">AllChem</span><span class="o">.</span><span class="n">AssignBondOrdersFromTemplate</span><span class="p">(</span><span class="n">template</span><span class="p">,</span> <span class="n">rd_mol</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">new_mol</span> <span class="o">=</span> <span class="n">rd_mol</span>
    <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
        <span class="n">new_mol</span> <span class="o">=</span> <span class="n">rd_mol</span>

    <span class="k">return</span> <span class="n">new_mol</span></div>



<div class="viewcode-block" id="write_sdf">
<a class="viewcode-back" href="../../../modules/multicom_ligand.data_utils.html#multicom_ligand.utils.data_utils.write_sdf">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">write_sdf</span><span class="p">(</span><span class="n">new_mol</span><span class="p">:</span> <span class="n">Chem</span><span class="o">.</span><span class="n">Mol</span><span class="p">,</span> <span class="n">pdb_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Write an RDKit molecule to an SD file.</span>

<span class="sd">    :param new_mol: RDKit molecule</span>
<span class="sd">    :param pdb_name: name of the output file</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">writer</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">SDWriter</span><span class="p">(</span><span class="n">pdb_name</span><span class="p">)</span>
    <span class="n">writer</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">new_mol</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Wrote </span><span class="si">{</span><span class="n">pdb_name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span></div>



<div class="viewcode-block" id="extract_protein_and_ligands_with_prody">
<a class="viewcode-back" href="../../../modules/multicom_ligand.data_utils.html#multicom_ligand.utils.data_utils.extract_protein_and_ligands_with_prody">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">extract_protein_and_ligands_with_prody</span><span class="p">(</span>
    <span class="n">input_pdb_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">protein_output_pdb_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">ligands_output_sdf_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">sanitize</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">add_element_types</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">ligand_smiles</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Chem</span><span class="o">.</span><span class="n">Mol</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Using ProDy, extract protein atoms and ligand molecules from a PDB file</span>
<span class="sd">    and write them to separate files.</span>

<span class="sd">    :param input_pdb_file: The input PDB file.</span>
<span class="sd">    :param protein_output_pdb_file: The output PDB file for the protein</span>
<span class="sd">        atoms.</span>
<span class="sd">    :param ligands_output_sdf_file: The output SDF file for the ligand</span>
<span class="sd">        molecules.</span>
<span class="sd">    :param sanitize: Whether to sanitize the ligand molecules.</span>
<span class="sd">    :param add_element_types: Whether to add element types to the</span>
<span class="sd">        protein atoms.</span>
<span class="sd">    :param ligand_smiles: The SMILES string of the ligand molecule.</span>
<span class="sd">    :return: The combined final ligand molecule(s) as an RDKit molecule.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">protein</span><span class="p">,</span> <span class="n">nucleic_acid</span><span class="p">,</span> <span class="n">ligand</span> <span class="o">=</span> <span class="n">get_pdb_components_with_prody</span><span class="p">(</span><span class="n">input_pdb_file</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">protein</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">write_pdb_with_prody</span><span class="p">(</span><span class="n">protein</span><span class="p">,</span> <span class="n">protein_output_pdb_file</span><span class="p">,</span> <span class="n">add_element_types</span><span class="o">=</span><span class="n">add_element_types</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">nucleic_acid</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">write_pdb_with_prody</span><span class="p">(</span>
            <span class="n">nucleic_acid</span><span class="p">,</span>
            <span class="n">protein_output_pdb_file</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;_protein.pdb&quot;</span><span class="p">,</span> <span class="s2">&quot;_nucleic.pdb&quot;</span><span class="p">),</span>
            <span class="n">add_element_types</span><span class="o">=</span><span class="n">add_element_types</span><span class="p">,</span>
        <span class="p">)</span>

    <span class="n">ligand_resnames</span> <span class="o">=</span> <span class="n">ligand</span><span class="o">.</span><span class="n">getResnames</span><span class="p">()</span>
    <span class="n">ligand_chids</span> <span class="o">=</span> <span class="n">ligand</span><span class="o">.</span><span class="n">getChids</span><span class="p">()</span>
    <span class="n">ligand_resnums</span> <span class="o">=</span> <span class="n">ligand</span><span class="o">.</span><span class="n">getResnums</span><span class="p">()</span>

    <span class="n">seen</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="n">resname_chain_resnum_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">resname</span><span class="p">,</span> <span class="n">chid</span><span class="p">,</span> <span class="n">resnum</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">ligand_resnames</span><span class="p">,</span> <span class="n">ligand_chids</span><span class="p">,</span> <span class="n">ligand_resnums</span><span class="p">):</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">resname</span><span class="p">,</span> <span class="n">chid</span><span class="p">,</span> <span class="n">resnum</span><span class="p">)</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">seen</span><span class="p">:</span>
            <span class="n">seen</span><span class="o">.</span><span class="n">add</span><span class="p">((</span><span class="n">resname</span><span class="p">,</span> <span class="n">chid</span><span class="p">,</span> <span class="n">resnum</span><span class="p">))</span>
            <span class="n">resname_chain_resnum_list</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">resname</span><span class="p">,</span> <span class="n">chid</span><span class="p">,</span> <span class="n">resnum</span><span class="p">))</span>

    <span class="n">new_mol</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">new_mol_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">ligand_smiles_components</span> <span class="o">=</span> <span class="n">ligand_smiles</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)</span> <span class="k">if</span> <span class="n">ligand_smiles</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">else</span> <span class="kc">None</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">resname_chain_resnum</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">resname_chain_resnum_list</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
        <span class="n">resname</span><span class="p">,</span> <span class="n">chain</span><span class="p">,</span> <span class="n">resnum</span> <span class="o">=</span> <span class="n">resname_chain_resnum</span>
        <span class="n">sub_smiles</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">ligand_smiles_components</span><span class="p">[</span><span class="n">i</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">ligand_smiles_components</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
            <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">ligand_smiles_components</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">resname_chain_resnum_list</span><span class="p">)</span>
            <span class="k">else</span> <span class="kc">None</span>
        <span class="p">)</span>
        <span class="n">new_mol</span> <span class="o">=</span> <span class="n">process_ligand_with_prody</span><span class="p">(</span>
            <span class="n">ligand</span><span class="p">,</span> <span class="n">resname</span><span class="p">,</span> <span class="n">chain</span><span class="p">,</span> <span class="n">resnum</span><span class="p">,</span> <span class="n">sanitize</span><span class="o">=</span><span class="n">sanitize</span><span class="p">,</span> <span class="n">sub_smiles</span><span class="o">=</span><span class="n">sub_smiles</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="n">new_mol</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">new_mol_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">new_mol</span><span class="p">)</span>
            <span class="n">write_sdf</span><span class="p">(</span>
                <span class="n">new_mol</span><span class="p">,</span>
                <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                    <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">ligands_output_sdf_file</span><span class="p">),</span>
                    <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">Path</span><span class="p">(</span><span class="n">ligands_output_sdf_file</span><span class="p">)</span><span class="o">.</span><span class="n">stem</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">resname</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">.sdf&quot;</span><span class="p">,</span>
                <span class="p">),</span>
            <span class="p">)</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">new_mol_list</span><span class="p">):</span>
        <span class="n">new_mol</span> <span class="o">=</span> <span class="n">combine_molecules</span><span class="p">(</span><span class="n">new_mol_list</span><span class="p">)</span>
        <span class="n">write_sdf</span><span class="p">(</span><span class="n">new_mol</span><span class="p">,</span> <span class="n">ligands_output_sdf_file</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">new_mol</span></div>



<div class="viewcode-block" id="create_sdf_file_from_smiles">
<a class="viewcode-back" href="../../../modules/multicom_ligand.data_utils.html#multicom_ligand.utils.data_utils.create_sdf_file_from_smiles">[docs]</a>
<span class="nd">@beartype</span>
<span class="k">def</span><span class="w"> </span><span class="nf">create_sdf_file_from_smiles</span><span class="p">(</span><span class="n">smiles</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">output_sdf_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Create an SDF file from a SMILES string.</span>

<span class="sd">    :param smiles: SMILES string of the molecule.</span>
<span class="sd">    :param output_sdf_file: Path to the output SDF file.</span>
<span class="sd">    :return: Path to the output SDF file.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">mol</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolFromSmiles</span><span class="p">(</span><span class="n">smiles</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">mol</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">mol</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolFromSmiles</span><span class="p">(</span><span class="n">smiles</span><span class="p">,</span> <span class="n">sanitize</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">writer</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">SDWriter</span><span class="p">(</span><span class="n">output_sdf_file</span><span class="p">)</span>
    <span class="n">writer</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">output_sdf_file</span></div>



<div class="viewcode-block" id="count_num_residues_in_pdb_file">
<a class="viewcode-back" href="../../../modules/multicom_ligand.data_utils.html#multicom_ligand.utils.data_utils.count_num_residues_in_pdb_file">[docs]</a>
<span class="nd">@beartype</span>
<span class="k">def</span><span class="w"> </span><span class="nf">count_num_residues_in_pdb_file</span><span class="p">(</span><span class="n">pdb_filepath</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Count the number of Ca atoms (i.e., residues) in a PDB file.</span>

<span class="sd">    :param pdb_filepath: Path to PDB file.</span>
<span class="sd">    :return: Number of Ca atoms (i.e., residues) in the PDB file.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">PDBParser</span><span class="p">()</span>
    <span class="n">structure</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">get_structure</span><span class="p">(</span><span class="s2">&quot;protein&quot;</span><span class="p">,</span> <span class="n">pdb_filepath</span><span class="p">)</span>
    <span class="n">count</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">model</span> <span class="ow">in</span> <span class="n">structure</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">chain</span> <span class="ow">in</span> <span class="n">model</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">residue</span> <span class="ow">in</span> <span class="n">chain</span><span class="p">:</span>
                <span class="k">if</span> <span class="s2">&quot;CA&quot;</span> <span class="ow">in</span> <span class="n">residue</span><span class="p">:</span>
                    <span class="n">count</span> <span class="o">+=</span> <span class="mi">1</span>
    <span class="k">return</span> <span class="n">count</span></div>



<div class="viewcode-block" id="count_pdb_inter_residue_clashes">
<a class="viewcode-back" href="../../../modules/multicom_ligand.data_utils.html#multicom_ligand.utils.data_utils.count_pdb_inter_residue_clashes">[docs]</a>
<span class="nd">@beartype</span>
<span class="k">def</span><span class="w"> </span><span class="nf">count_pdb_inter_residue_clashes</span><span class="p">(</span><span class="n">pdb_filepath</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">clash_cutoff</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.63</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Count the number of inter-residue clashes in a protein PDB file.</span>
<span class="sd">    From: https://www.blopig.com/blog/2023/05/checking-your-pdb-file-for-clashing-atoms/</span>

<span class="sd">    :param pdb_filepath: Path to the PDB file.</span>
<span class="sd">    :param clash_cutoff: The cutoff for what is considered a clash.</span>
<span class="sd">    :return: The number of inter-residue clashes in the structure.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">PDB</span><span class="o">.</span><span class="n">PDBParser</span><span class="p">(</span><span class="n">QUIET</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">structure</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">get_structure</span><span class="p">(</span><span class="s2">&quot;protein&quot;</span><span class="p">,</span> <span class="n">pdb_filepath</span><span class="p">)</span>

    <span class="c1"># Atomic radii for various atom types.</span>
    <span class="c1"># You can comment out the ones you don&#39;t care about or add new ones</span>
    <span class="n">atom_radii</span> <span class="o">=</span> <span class="p">{</span>
        <span class="c1">#    &quot;H&quot;: 1.20,  # Who cares about hydrogen??</span>
        <span class="s2">&quot;C&quot;</span><span class="p">:</span> <span class="mf">1.70</span><span class="p">,</span>
        <span class="s2">&quot;N&quot;</span><span class="p">:</span> <span class="mf">1.55</span><span class="p">,</span>
        <span class="s2">&quot;O&quot;</span><span class="p">:</span> <span class="mf">1.52</span><span class="p">,</span>
        <span class="s2">&quot;S&quot;</span><span class="p">:</span> <span class="mf">1.80</span><span class="p">,</span>
        <span class="s2">&quot;F&quot;</span><span class="p">:</span> <span class="mf">1.47</span><span class="p">,</span>
        <span class="s2">&quot;P&quot;</span><span class="p">:</span> <span class="mf">1.80</span><span class="p">,</span>
        <span class="s2">&quot;CL&quot;</span><span class="p">:</span> <span class="mf">1.75</span><span class="p">,</span>
        <span class="s2">&quot;MG&quot;</span><span class="p">:</span> <span class="mf">1.73</span><span class="p">,</span>
    <span class="p">}</span>

    <span class="c1"># Set what we count as a clash for each pair of atoms</span>
    <span class="n">clash_cutoffs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="n">i</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="n">j</span><span class="p">:</span> <span class="p">(</span><span class="n">clash_cutoff</span> <span class="o">*</span> <span class="p">(</span><span class="n">atom_radii</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">+</span> <span class="n">atom_radii</span><span class="p">[</span><span class="n">j</span><span class="p">]))</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">atom_radii</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">atom_radii</span>
    <span class="p">}</span>

    <span class="c1"># Extract atoms for which we have a radii</span>
    <span class="n">atoms</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">structure</span><span class="o">.</span><span class="n">get_atoms</span><span class="p">()</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">element</span> <span class="ow">in</span> <span class="n">atom_radii</span><span class="p">]</span>
    <span class="n">coords</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">a</span><span class="o">.</span><span class="n">coord</span> <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">atoms</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="s2">&quot;d&quot;</span><span class="p">)</span>

    <span class="c1"># Build a KDTree (speedy!!!)</span>
    <span class="n">kdt</span> <span class="o">=</span> <span class="n">PDB</span><span class="o">.</span><span class="n">kdtrees</span><span class="o">.</span><span class="n">KDTree</span><span class="p">(</span><span class="n">coords</span><span class="p">)</span>

    <span class="c1"># Initialize a list to hold clashes</span>
    <span class="n">clashes</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="c1"># Iterate through all atoms</span>
    <span class="k">for</span> <span class="n">atom_1</span> <span class="ow">in</span> <span class="n">atoms</span><span class="p">:</span>
        <span class="c1"># Find atoms that could be clashing</span>
        <span class="n">kdt_search</span> <span class="o">=</span> <span class="n">kdt</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">atom_1</span><span class="o">.</span><span class="n">coord</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="s2">&quot;d&quot;</span><span class="p">),</span> <span class="nb">max</span><span class="p">(</span><span class="n">clash_cutoffs</span><span class="o">.</span><span class="n">values</span><span class="p">()))</span>

        <span class="c1"># Get index and distance of potential clashes</span>
        <span class="n">potential_clash</span> <span class="o">=</span> <span class="p">[(</span><span class="n">a</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">a</span><span class="o">.</span><span class="n">radius</span><span class="p">)</span> <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">kdt_search</span><span class="p">]</span>

        <span class="k">for</span> <span class="n">ix</span><span class="p">,</span> <span class="n">atom_distance</span> <span class="ow">in</span> <span class="n">potential_clash</span><span class="p">:</span>
            <span class="n">atom_2</span> <span class="o">=</span> <span class="n">atoms</span><span class="p">[</span><span class="n">ix</span><span class="p">]</span>

            <span class="c1"># Exclude clashes from atoms in the same residue</span>
            <span class="k">if</span> <span class="n">atom_1</span><span class="o">.</span><span class="n">parent</span><span class="o">.</span><span class="n">id</span> <span class="o">==</span> <span class="n">atom_2</span><span class="o">.</span><span class="n">parent</span><span class="o">.</span><span class="n">id</span><span class="p">:</span>
                <span class="k">continue</span>

            <span class="c1"># Exclude clashes from peptide bonds</span>
            <span class="k">elif</span> <span class="p">(</span><span class="n">atom_2</span><span class="o">.</span><span class="n">name</span> <span class="o">==</span> <span class="s2">&quot;C&quot;</span> <span class="ow">and</span> <span class="n">atom_1</span><span class="o">.</span><span class="n">name</span> <span class="o">==</span> <span class="s2">&quot;N&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span>
                <span class="n">atom_2</span><span class="o">.</span><span class="n">name</span> <span class="o">==</span> <span class="s2">&quot;N&quot;</span> <span class="ow">and</span> <span class="n">atom_1</span><span class="o">.</span><span class="n">name</span> <span class="o">==</span> <span class="s2">&quot;C&quot;</span>
            <span class="p">):</span>
                <span class="k">continue</span>

            <span class="c1"># Exclude clashes from disulphide bridges</span>
            <span class="k">elif</span> <span class="p">(</span><span class="n">atom_2</span><span class="o">.</span><span class="n">name</span> <span class="o">==</span> <span class="s2">&quot;SG&quot;</span> <span class="ow">and</span> <span class="n">atom_1</span><span class="o">.</span><span class="n">name</span> <span class="o">==</span> <span class="s2">&quot;SG&quot;</span><span class="p">)</span> <span class="ow">and</span> <span class="n">atom_distance</span> <span class="o">&gt;</span> <span class="mf">1.88</span><span class="p">:</span>
                <span class="k">continue</span>

            <span class="k">if</span> <span class="n">atom_distance</span> <span class="o">&lt;</span> <span class="n">clash_cutoffs</span><span class="p">[</span><span class="n">atom_2</span><span class="o">.</span><span class="n">element</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="n">atom_1</span><span class="o">.</span><span class="n">element</span><span class="p">]:</span>
                <span class="n">clashes</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">atom_1</span><span class="p">,</span> <span class="n">atom_2</span><span class="p">))</span>

    <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="n">clashes</span><span class="p">)</span> <span class="o">//</span> <span class="mi">2</span></div>

</pre></div>
        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          
          
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2025, Alex Morehead
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer no-toc">
      
      
      
    </aside>
  </div>
</div><script src="../../../_static/jquery.js?v=5d32c60e"></script>
    <script src="../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
    <script src="../../../_static/documentation_options.js?v=6c02275b"></script>
    <script src="../../../_static/doctools.js?v=9a2dae69"></script>
    <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../_static/scripts/furo.js?v=46bd48cc"></script>
    <script src="../../../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../../../_static/copybutton.js?v=f281be69"></script>
    <script src="../../../_static/tabs.js?v=3ee01567"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.1.10/require.min.js"></script>
    <script src="http://ajax.googleapis.com/ajax/libs/jquery/1.9.1/jquery.min.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    </body>
</html>