<!doctype html>
<html class="no-js" lang="en" data-content_root="../../../">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark">
        <script async src="https://www.googletagmanager.com/gtag/js?id="></script>
        <script>
            window.dataLayer = window.dataLayer || [];
            function gtag(){dataLayer.push(arguments);}
            gtag('js', new Date());
            
            gtag('config', '');
            
        </script>
    <link rel="index" title="Index" href="../../../genindex.html" /><link rel="search" title="Search" href="../../../search.html" />
        <link rel="prefetch" href="../../../_static/WorkBench.jpeg" as="image" />

    <!-- Generated with Sphinx 7.4.7 and Furo 2025.07.19 -->
        <title>multicom_ligand.models.ensemble_generation - MULTICOM_ligand 0.4.0</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=d111a655" />
    <link rel="stylesheet" type="text/css" href="../../../_static/styles/furo.css?v=25af2a20" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../../../_static/tabs.css?v=4c969af8" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-codeautolink.css?v=125d5c1c" />
    <link rel="stylesheet" type="text/css" href="../../../_static/styles/furo-extensions.css?v=8dab3a3b" />
    
    


<style>
  body {
    --color-code-background: #f2f2f2;
  --color-code-foreground: #1e1e1e;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #2b2b2b;
  --color-code-foreground: #f8f8f2;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #2b2b2b;
  --color-code-foreground: #f8f8f2;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-with-moon" viewBox="0 0 24 24">
    <title>Auto light/dark, in light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path style="opacity: 50%" d="M 5.411 14.504 C 5.471 14.504 5.532 14.504 5.591 14.504 C 3.639 16.319 4.383 19.569 6.931 20.352 C 7.693 20.586 8.512 20.551 9.25 20.252 C 8.023 23.207 4.056 23.725 2.11 21.184 C 0.166 18.642 1.702 14.949 4.874 14.536 C 5.051 14.512 5.231 14.5 5.411 14.5 L 5.411 14.504 Z"/>
      <line x1="14.5" y1="3.25" x2="14.5" y2="1.25"/>
      <line x1="14.5" y1="15.85" x2="14.5" y2="17.85"/>
      <line x1="10.044" y1="5.094" x2="8.63" y2="3.68"/>
      <line x1="19" y1="14.05" x2="20.414" y2="15.464"/>
      <line x1="8.2" y1="9.55" x2="6.2" y2="9.55"/>
      <line x1="20.8" y1="9.55" x2="22.8" y2="9.55"/>
      <line x1="10.044" y1="14.006" x2="8.63" y2="15.42"/>
      <line x1="19" y1="5.05" x2="20.414" y2="3.636"/>
      <circle cx="14.5" cy="9.55" r="3.6"/>
    </svg>
  </symbol>
  <symbol id="svg-moon-with-sun" viewBox="0 0 24 24">
    <title>Auto light/dark, in dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path d="M 8.282 7.007 C 8.385 7.007 8.494 7.007 8.595 7.007 C 5.18 10.184 6.481 15.869 10.942 17.24 C 12.275 17.648 13.706 17.589 15 17.066 C 12.851 22.236 5.91 23.143 2.505 18.696 C -0.897 14.249 1.791 7.786 7.342 7.063 C 7.652 7.021 7.965 7 8.282 7 L 8.282 7.007 Z"/>
      <line style="opacity: 50%" x1="18" y1="3.705" x2="18" y2="2.5"/>
      <line style="opacity: 50%" x1="18" y1="11.295" x2="18" y2="12.5"/>
      <line style="opacity: 50%" x1="15.316" y1="4.816" x2="14.464" y2="3.964"/>
      <line style="opacity: 50%" x1="20.711" y1="10.212" x2="21.563" y2="11.063"/>
      <line style="opacity: 50%" x1="14.205" y1="7.5" x2="13.001" y2="7.5"/>
      <line style="opacity: 50%" x1="21.795" y1="7.5" x2="23" y2="7.5"/>
      <line style="opacity: 50%" x1="15.316" y1="10.184" x2="14.464" y2="11.036"/>
      <line style="opacity: 50%" x1="20.711" y1="4.789" x2="21.563" y2="3.937"/>
      <circle style="opacity: 50%" cx="18" cy="7.5" r="2.169"/>
    </svg>
  </symbol>
  <symbol id="svg-pencil" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-pencil-code">
      <path d="M4 20h4l10.5 -10.5a2.828 2.828 0 1 0 -4 -4l-10.5 10.5v4" />
      <path d="M13.5 6.5l4 4" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
  <symbol id="svg-eye" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-eye-code">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M10 12a2 2 0 1 0 4 0a2 2 0 0 0 -4 0" />
      <path
        d="M11.11 17.958c-3.209 -.307 -5.91 -2.293 -8.11 -5.958c2.4 -4 5.4 -6 9 -6c3.6 0 6.6 2 9 6c-.21 .352 -.427 .688 -.647 1.008" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>

<a class="skip-to-content muted-link" href="#furo-main-content">Skip to content</a>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../../../index.html"><div class="brand">MULTICOM_ligand 0.4.0</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
          <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon no-toc" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../../../index.html">
  <div class="sidebar-logo-container">
    <img class="sidebar-logo" src="../../../_static/WorkBench.jpeg" alt="Logo"/>
  </div>
  
  <span class="sidebar-brand-text">MULTICOM_ligand 0.4.0</span>
  
</a><form class="sidebar-search-container" method="get" action="../../../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../installation.html#dynamicbind-checkpoint-0-25-gb">DynamicBind checkpoint (~0.25 GB)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials.html#rosettafold-all-atom-checkpoint-1-5-gb">RoseTTAFold-All-Atom checkpoint (~1.5 GB)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../data_preparation.html">How to prepare <cite>MULTICOM_ligand</cite> data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../available_methods.html">Available inference methods</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../method_inference.html">How to run inference with individual methods</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../ensemble_inference.html">How to run inference with a method ensemble</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../ensemble_inference.html#note-the-suffixes-for-both-output-dir-and-ensemble-benchmarking-repeat-index-should-be-modified-to-e-g-2-3">NOTE: the suffixes for both <code class="docutils literal notranslate"><span class="pre">output_dir</span></code> and <code class="docutils literal notranslate"><span class="pre">ensemble_benchmarking_repeat_index</span></code> should be modified to e.g., 2, 3, …</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../ensemble_inference.html#benchmark-using-the-posebusters-benchmark-dataset-e-g-after-generating-40-complexes-per-target-with-each-method">benchmark using the PoseBusters Benchmark dataset e.g., after generating 40 complexes per target with each method</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../ensemble_inference.html#benchmark-using-the-astex-diverse-dataset-e-g-after-generating-40-complexes-per-target-with-each-method">benchmark using the Astex Diverse dataset e.g., after generating 40 complexes per target with each method</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../ensemble_inference.html#benchmark-using-the-dockgen-dataset-e-g-after-generating-40-complexes-per-target-with-each-method">benchmark using the DockGen dataset e.g., after generating 40 complexes per target with each method</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../ensemble_inference.html#benchmark-using-the-casp15-dataset-e-g-after-generating-40-complexes-per-target-with-each-method">benchmark using the CASP15 dataset e.g., after generating 40 complexes per target with each method</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../ensemble_inference.html#analyze-benchmarking-results-for-the-posebusters-benchmark-dataset">analyze benchmarking results for the PoseBusters Benchmark dataset</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../comparative_plots.html">How to create comparative plots of inference results</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../comparative_plots.html#analyze-benchmarking-results-for-the-casp15-dataset">analyze benchmarking results for the CASP15 dataset</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../for_developers.html">For developers</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../acknowledgements.html">Acknowledgements</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../citing_this_work.html">Citing this work</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../bonus.html">Bonus</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Default Configs</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../configs/analysis.html">Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../configs/data.html">Data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../configs/model.html">Model</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API Reference</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/multicom_ligand.binding_site_crop_preparation.html">Binding site crop preparation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/multicom_ligand.complex_alignment.html">Complex alignment</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/multicom_ligand.inference_relaxation.html">Inference relaxation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/multicom_ligand.minimize_energy.html">Minimize energy</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/multicom_ligand.ensemble_generation.html">Ensemble generation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/multicom_ligand.data_utils.html">Data utilities</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/multicom_ligand.model_utils.html">Model utilities</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/multicom_ligand.utils.html">General utilities</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/multicom_ligand.resolvers.html">OmegaConf resolvers</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
              <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon no-toc" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main" id="furo-main-content">
          <h1>Source code for multicom_ligand.models.ensemble_generation</h1><div class="highlight"><pre>
<span></span><span class="c1"># -------------------------------------------------------------------------------------------------------------------------------------</span>
<span class="c1"># Following code curated for MULTICOM_ligand: (https://github.com/BioinfoMachineLearning/MULTICOM_ligand)</span>
<span class="c1"># -------------------------------------------------------------------------------------------------------------------------------------</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">ast</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">glob</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">logging</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">multiprocessing</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">re</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">shutil</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">subprocess</span>  <span class="c1"># nosec</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">tempfile</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">io</span><span class="w"> </span><span class="kn">import</span> <span class="n">StringIO</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">hydra</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">rootutils</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">beartype.typing</span><span class="w"> </span><span class="kn">import</span> <span class="n">Any</span><span class="p">,</span> <span class="n">Dict</span><span class="p">,</span> <span class="n">List</span><span class="p">,</span> <span class="n">Literal</span><span class="p">,</span> <span class="n">Optional</span><span class="p">,</span> <span class="n">Tuple</span><span class="p">,</span> <span class="n">Union</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">Bio</span><span class="w"> </span><span class="kn">import</span> <span class="n">PDB</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">Bio.PDB</span><span class="w"> </span><span class="kn">import</span> <span class="n">Structure</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">Bio.PDB.PDBIO</span><span class="w"> </span><span class="kn">import</span> <span class="n">PDBIO</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">Bio.PDB.PDBParser</span><span class="w"> </span><span class="kn">import</span> <span class="n">PDBParser</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">omegaconf</span><span class="w"> </span><span class="kn">import</span> <span class="n">DictConfig</span><span class="p">,</span> <span class="n">OmegaConf</span><span class="p">,</span> <span class="n">open_dict</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">posebusters</span><span class="w"> </span><span class="kn">import</span> <span class="n">PoseBusters</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">rdkit</span><span class="w"> </span><span class="kn">import</span> <span class="n">Chem</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">rdkit.Chem</span><span class="w"> </span><span class="kn">import</span> <span class="n">AllChem</span>

<span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="nb">format</span><span class="o">=</span><span class="s2">&quot;[</span><span class="si">%(asctime)s</span><span class="s2">] {</span><span class="si">%(filename)s</span><span class="s2">:</span><span class="si">%(lineno)d</span><span class="s2">} </span><span class="si">%(levelname)s</span><span class="s2"> - </span><span class="si">%(message)s</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span>

<span class="n">rootutils</span><span class="o">.</span><span class="n">setup_root</span><span class="p">(</span><span class="vm">__file__</span><span class="p">,</span> <span class="n">indicator</span><span class="o">=</span><span class="s2">&quot;.project-root&quot;</span><span class="p">,</span> <span class="n">pythonpath</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">multicom_ligand</span><span class="w"> </span><span class="kn">import</span> <span class="n">register_custom_omegaconf_resolvers</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">multicom_ligand.analysis.complex_alignment</span><span class="w"> </span><span class="kn">import</span> <span class="n">align_complex_to_protein_only</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">multicom_ligand.data.components.esmfold_apo_to_holo_alignment</span><span class="w"> </span><span class="kn">import</span> <span class="n">read_molecule</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">multicom_ligand.models.inference_relaxation</span><span class="w"> </span><span class="kn">import</span> <span class="n">relax_single_filepair</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">multicom_ligand.models.minimize_energy</span><span class="w"> </span><span class="kn">import</span> <span class="n">minimize_energy</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">multicom_ligand.utils.data_utils</span><span class="w"> </span><span class="kn">import</span> <span class="p">(</span>
    <span class="n">count_pdb_inter_residue_clashes</span><span class="p">,</span>
    <span class="n">extract_sequences_from_macromolecule_structure_file</span><span class="p">,</span>
    <span class="n">renumber_biopython_structure_residues</span><span class="p">,</span>
<span class="p">)</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">multicom_ligand.utils.model_utils</span><span class="w"> </span><span class="kn">import</span> <span class="n">calculate_rmsd</span>

<span class="n">METHODS_PREDICTING_HOLO_PROTEIN_AB_INITIO</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;neuralplexer&quot;</span><span class="p">,</span> <span class="s2">&quot;flowdock&quot;</span><span class="p">,</span> <span class="s2">&quot;rfaa&quot;</span><span class="p">}</span>
<span class="n">PREFERRED_MULTI_LIGAND_METHODS</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;neuralplexer&quot;</span><span class="p">,</span> <span class="s2">&quot;flowdock&quot;</span><span class="p">}</span>

<span class="n">ENSEMBLE_PREDICTIONS</span> <span class="o">=</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]]</span>
<span class="n">RANKED_ENSEMBLE_PREDICTIONS</span> <span class="o">=</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">,</span> <span class="nb">float</span><span class="p">]]</span>
<span class="n">PREFERRED_METHOD_FOR_RERANKING</span> <span class="o">=</span> <span class="n">Literal</span><span class="p">[</span>
    <span class="s2">&quot;diffdock&quot;</span><span class="p">,</span> <span class="s2">&quot;dynamicbind&quot;</span><span class="p">,</span> <span class="s2">&quot;neuralplexer&quot;</span><span class="p">,</span> <span class="s2">&quot;flowdock&quot;</span><span class="p">,</span> <span class="s2">&quot;rfaa&quot;</span><span class="p">,</span> <span class="s2">&quot;vina&quot;</span>
<span class="p">]</span>

<span class="c1"># NOTE: the following sequence is derived from `5S8I_2LY.pdb` of the PoseBusters Benchmark set</span>
<span class="n">LIGAND_ONLY_RECEPTOR_PLACEHOLDER_SEQUENCE</span> <span class="o">=</span> <span class="s2">&quot;DSLFAGLVGEYYGTNSQLNNISDFRALVDSKEADATFEAANISYGRGSSDVAKGTHLQEFLGSDASTLSTDPGDNTDGGIYLQGYVYLEAGTYNFKVTADDGYEITINGNPVATVDNNQSVYTVTHASFTISESGYQAIDMIWWDQGGDYVFQPTLSADGGSTYFVLDSAILSSTGETPY&quot;</span>

<span class="n">AFFINITY_UNIT_CONVERSION_DICT</span> <span class="o">=</span> <span class="p">{</span>
    <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span>  <span class="c1"># NaN to NaN</span>
    <span class="s2">&quot;uM&quot;</span><span class="p">:</span> <span class="mf">1e-6</span><span class="p">,</span>  <span class="c1"># micromolar to M</span>
    <span class="s2">&quot;nM&quot;</span><span class="p">:</span> <span class="mf">1e-9</span><span class="p">,</span>  <span class="c1"># nanomolar to M</span>
    <span class="s2">&quot;mM&quot;</span><span class="p">:</span> <span class="mf">1e-3</span><span class="p">,</span>  <span class="c1"># millimolar to M</span>
    <span class="s2">&quot;pM&quot;</span><span class="p">:</span> <span class="mf">1e-12</span><span class="p">,</span>  <span class="c1"># picomolar to M</span>
    <span class="s2">&quot;M&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span>  <span class="c1"># Molar to M</span>
    <span class="s2">&quot;M^-1&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span>  <span class="c1"># Reciprocal Molar to M</span>
    <span class="s2">&quot;fM&quot;</span><span class="p">:</span> <span class="mf">1e-15</span><span class="p">,</span>  <span class="c1"># femtomolar to M</span>
<span class="p">}</span>


<div class="viewcode-block" id="create_temporary_fasta_file">
<a class="viewcode-back" href="../../../modules/multicom_ligand.ensemble_generation.html#multicom_ligand.models.ensemble_generation.create_temporary_fasta_file">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">create_temporary_fasta_file</span><span class="p">(</span><span class="n">protein_sequence</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">name</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Create a temporary FASTA file for the input protein sequence.</span>

<span class="sd">    :param protein_sequence: Amino acid sequence of the protein.</span>
<span class="sd">    :param name: Optional name of the temporary FASTA file.</span>
<span class="sd">    :return: Path to the temporary FASTA file.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">name</span> <span class="o">=</span> <span class="n">name</span> <span class="k">if</span> <span class="n">name</span> <span class="k">else</span> <span class="s2">&quot;temp&quot;</span>
    <span class="k">with</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">NamedTemporaryFile</span><span class="p">(</span><span class="n">mode</span><span class="o">=</span><span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">delete</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">suffix</span><span class="o">=</span><span class="s2">&quot;.fasta&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fasta_file</span><span class="p">:</span>
        <span class="n">fasta_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&gt;</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="se">\n</span><span class="si">{</span><span class="n">protein_sequence</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">temp_fasta_file</span> <span class="o">=</span> <span class="n">fasta_file</span><span class="o">.</span><span class="n">name</span>
    <span class="k">return</span> <span class="n">temp_fasta_file</span></div>



<div class="viewcode-block" id="predict_protein_structure_from_sequence">
<a class="viewcode-back" href="../../../modules/multicom_ligand.ensemble_generation.html#multicom_ligand.models.ensemble_generation.predict_protein_structure_from_sequence">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">predict_protein_structure_from_sequence</span><span class="p">(</span>
    <span class="n">python_exec_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">structure_prediction_script_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">fasta_filepath</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">output_pdb_dir</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">chunk_size</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">cpu_only</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">cpu_offload</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">cuda_device_index</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Predict protein structure from amino acid sequence.</span>

<span class="sd">    :param python_exec_path: Path to the Python executable with which to</span>
<span class="sd">        run Python scripts.</span>
<span class="sd">    :param structure_prediction_script_path: Path to the ESMFold</span>
<span class="sd">        structure prediction script to run.</span>
<span class="sd">    :param fasta_filepath: Path to the input FASTA file.</span>
<span class="sd">    :param output_pdb_dir: Path to the output PDB directory.</span>
<span class="sd">    :param chunk_size: Optional chunk size for structure prediction.</span>
<span class="sd">    :param cpu_only: Whether to use CPU only for structure prediction.</span>
<span class="sd">    :param cpu_offload: Whether to use CPU offloading for structure</span>
<span class="sd">        prediction.</span>
<span class="sd">    :param cuda_device_index: The optional index of the CUDA device to</span>
<span class="sd">        use for structure prediction.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">cmd_inputs</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">python_exec_path</span><span class="p">,</span>
        <span class="n">structure_prediction_script_path</span><span class="p">,</span>
        <span class="s2">&quot;--fasta&quot;</span><span class="p">,</span>
        <span class="n">fasta_filepath</span><span class="p">,</span>
        <span class="s2">&quot;--pdb&quot;</span><span class="p">,</span>
        <span class="n">output_pdb_dir</span><span class="p">,</span>
        <span class="s2">&quot;--cuda-device-index&quot;</span><span class="p">,</span>
        <span class="nb">str</span><span class="p">(</span><span class="n">cuda_device_index</span><span class="p">),</span>
    <span class="p">]</span>
    <span class="k">if</span> <span class="n">chunk_size</span><span class="p">:</span>
        <span class="n">cmd_inputs</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="s2">&quot;--chunk-size&quot;</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">chunk_size</span><span class="p">)])</span>
    <span class="k">if</span> <span class="n">cpu_only</span><span class="p">:</span>
        <span class="n">cmd_inputs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;--cpu-only&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">cpu_offload</span><span class="p">:</span>
        <span class="n">cmd_inputs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;--cpu-offload&quot;</span><span class="p">)</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">cmd_inputs</span><span class="p">,</span> <span class="n">check</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>  <span class="c1"># nosec</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">e</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;ESMFold structure prediction for input protein sequence has been saved to </span><span class="si">{</span><span class="n">output_pdb_dir</span><span class="si">}</span><span class="s2">.&quot;</span>
    <span class="p">)</span></div>



<div class="viewcode-block" id="insert_hpc_headers">
<a class="viewcode-back" href="../../../modules/multicom_ligand.ensemble_generation.html#multicom_ligand.models.ensemble_generation.insert_hpc_headers">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">insert_hpc_headers</span><span class="p">(</span>
    <span class="n">method</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">gpu_partition</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;chengji-lab-gpu&quot;</span><span class="p">,</span>
    <span class="n">gpu_account</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;chengji-lab&quot;</span><span class="p">,</span>
    <span class="n">gpu_type</span><span class="p">:</span> <span class="n">Literal</span><span class="p">[</span><span class="s2">&quot;A100&quot;</span><span class="p">,</span> <span class="s2">&quot;H100&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;H100&quot;</span><span class="p">,</span>
    <span class="n">cpu_memory_in_gb</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">59</span><span class="p">,</span>
    <span class="n">time_limit</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;7-00:00:00&quot;</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Insert batch headers for SLURM job scheduling.</span>

<span class="sd">    :param method: Name of the method for which to generate a prediction</span>
<span class="sd">        script.</span>
<span class="sd">    :param gpu_partition: Name of the GPU partition to use.</span>
<span class="sd">    :param gpu_account: Name of the GPU account to use.</span>
<span class="sd">    :param cpu_memory_in_gb: Amount of CPU memory to request in GB.</span>
<span class="sd">    :param time_limit: Time limit for the job as a SLURM-compatible</span>
<span class="sd">        string.</span>
<span class="sd">    :return: Batch headers string for SLURM job scheduling.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;&quot;&quot;######################### Batch Headers #########################</span>
<span class="s2">#SBATCH --partition </span><span class="si">{</span><span class="n">gpu_partition</span><span class="si">}</span><span class="s2"> # use reserved partition `chengji-lab-gpu`</span>
<span class="s2">#SBATCH --account </span><span class="si">{</span><span class="n">gpu_account</span><span class="si">}</span><span class="s2">  # NOTE: this must be specified to use the reserved partition above</span>
<span class="s2">#SBATCH --nodes=1              # NOTE: this needs to match Lightning&#39;s `Trainer(num_nodes=...)`</span>
<span class="s2">#SBATCH --gres gpu:</span><span class="si">{</span><span class="n">gpu_type</span><span class="si">}</span><span class="s2">:1      # request </span><span class="si">{</span><span class="n">gpu_type</span><span class="si">}</span><span class="s2"> GPU resource(s)</span>
<span class="s2">#SBATCH --ntasks-per-node=1    # NOTE: this needs to be `1` on SLURM clusters when using Lightning&#39;s `ddp_spawn` strategy`; otherwise, set to match Lightning&#39;s quantity of `Trainer(devices=...)`</span>
<span class="s2">#SBATCH --mem=</span><span class="si">{</span><span class="n">cpu_memory_in_gb</span><span class="si">}</span><span class="s2">G              # NOTE: use `--mem=0` to request all memory &quot;available&quot; on the assigned node</span>
<span class="s2">#SBATCH -t </span><span class="si">{</span><span class="n">time_limit</span><span class="si">}</span><span class="s2">          # time limit for the job (up to two days: `2-00:00:00`)</span>
<span class="s2">#SBATCH -J multicom_ligand_</span><span class="si">{</span><span class="n">method</span><span class="si">}</span><span class="s2">_ensembling # job name</span>
<span class="s2">#SBATCH --output=R-%x.%j.out   # output log file</span>
<span class="s2">#SBATCH --error=R-%x.%j.err    # error log file</span>

<span class="s2">module purge</span>
<span class="s2">module load cuda/11.8.0_gcc_9.5.0</span>

<span class="s2"># determine location of the project directory</span>
<span class="s2">use_private_project_dir=false # NOTE: customize as needed</span>
<span class="s2">if [ &quot;$use_private_project_dir&quot; = true ]; then</span>
<span class="s2">    project_dir=&quot;/home/$USER/data/Repositories/Lab_Repositories/MULTICOM_ligand&quot;</span>
<span class="s2">else</span>
<span class="s2">    project_dir=&quot;/cluster/pixstor/chengji-lab/$USER/Repositories/Lab_Repositories/MULTICOM_ligand&quot;</span>
<span class="s2">fi</span>

<span class="s2"># shellcheck source=/dev/null</span>
<span class="s2">source /home/$USER/mambaforge/etc/profile.d/conda.sh</span>

<span class="s2">cd &quot;$project_dir&quot; || exit&quot;&quot;&quot;</span></div>



<div class="viewcode-block" id="create_diffdock_bash_script">
<a class="viewcode-back" href="../../../modules/multicom_ligand.ensemble_generation.html#multicom_ligand.models.ensemble_generation.create_diffdock_bash_script">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">create_diffdock_bash_script</span><span class="p">(</span>
    <span class="n">protein_filepath</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">ligand_smiles</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">input_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">output_filepath</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">cfg</span><span class="p">:</span> <span class="n">DictConfig</span><span class="p">,</span>
    <span class="n">generate_hpc_scripts</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Create a bash script to run DiffDock protein-ligand complex prediction.</span>

<span class="sd">    :param protein_filepath: Path to the input protein structure PDB</span>
<span class="sd">        file.</span>
<span class="sd">    :param ligand_smiles: SMILES string of the input ligand.</span>
<span class="sd">    :param input_id: Input ID.</span>
<span class="sd">    :param output_filepath: Path to the output bash script file.</span>
<span class="sd">    :param cfg: Configuration dictionary for runtime arguments.</span>
<span class="sd">    :param generate_hpc_scripts: Whether to generate HPC scripts for</span>
<span class="sd">        DiffDock.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">bash_script_content</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;&quot;&quot;#!/bin/bash -l</span>
<span class="si">{</span><span class="n">insert_hpc_headers</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s1">&#39;diffdock&#39;</span><span class="p">)</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">generate_hpc_scripts</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;source /home/$USER/mambaforge/etc/profile.d/conda.sh&#39;</span><span class="si">}</span>
<span class="s2">conda activate </span><span class="si">{</span><span class="s2">&quot;$project_dir/MULTICOM_ligand/&quot;</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">generate_hpc_scripts</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;MULTICOM_ligand&#39;</span><span class="si">}</span>

<span class="s2"># command to run diffdock_input_preparation.py</span>
<span class="s2">python multicom_ligand/data/diffdock_input_preparation.py </span><span class="se">\\</span>
<span class="s2">    output_csv_path=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">diffdock_input_csv_path</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    protein_filepath=&quot;</span><span class="si">{</span><span class="n">protein_filepath</span><span class="si">}</span><span class="s2">&quot; </span><span class="se">\\</span>
<span class="s2">    ligand_smiles=&#39;&quot;</span><span class="si">{</span><span class="n">ligand_smiles</span><span class="si">}</span><span class="s2">&quot;&#39; </span><span class="se">\\</span>
<span class="s2">    input_id=&quot;</span><span class="si">{</span><span class="n">input_id</span><span class="si">}</span><span class="s2">&quot;</span>

<span class="s2"># command to run diffdock_inference.py</span>
<span class="s2">echo &quot;Calling diffdock_inference.py!&quot;</span>
<span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">diffdock_python_exec_path</span><span class="si">}</span><span class="s2"> multicom_ligand/models/diffdock_inference.py </span><span class="se">\\</span>
<span class="s2">    cuda_device_index=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">cuda_device_index</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    python_exec_path=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">diffdock_python_exec_path</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    diffdock_exec_dir=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">diffdock_exec_dir</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    input_csv_path=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">diffdock_input_csv_path</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    output_dir=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">diffdock_output_dir</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    model_dir=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">diffdock_model_dir</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    confidence_model_dir=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">diffdock_confidence_model_dir</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    inference_steps=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">diffdock_inference_steps</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    samples_per_complex=</span><span class="si">{</span><span class="nb">min</span><span class="p">(</span><span class="n">cfg</span><span class="o">.</span><span class="n">diffdock_samples_per_complex</span><span class="p">,</span><span class="w"> </span><span class="n">cfg</span><span class="o">.</span><span class="n">max_method_predictions</span><span class="p">)</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    batch_size=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">diffdock_batch_size</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    actual_steps=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">diffdock_actual_steps</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    no_final_step_noise=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">diffdock_no_final_step_noise</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    skip_existing=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">diffdock_skip_existing</span><span class="si">}</span>

<span class="s2">echo &quot;Finished calling diffdock_inference.py!&quot;</span>
<span class="s2">    &quot;&quot;&quot;</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">output_filepath</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
        <span class="n">file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">bash_script_content</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Bash script &#39;</span><span class="si">{</span><span class="n">output_filepath</span><span class="si">}</span><span class="s2">&#39; created successfully.&quot;</span><span class="p">)</span></div>



<div class="viewcode-block" id="create_dynamicbind_bash_script">
<a class="viewcode-back" href="../../../modules/multicom_ligand.ensemble_generation.html#multicom_ligand.models.ensemble_generation.create_dynamicbind_bash_script">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">create_dynamicbind_bash_script</span><span class="p">(</span>
    <span class="n">protein_filepath</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">ligand_smiles</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">output_filepath</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">cfg</span><span class="p">:</span> <span class="n">DictConfig</span><span class="p">,</span>
    <span class="n">generate_hpc_scripts</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Create a bash script to run DynamicBind protein-ligand complex</span>
<span class="sd">    prediction.</span>

<span class="sd">    :param protein_filepath: Path to the input protein structure PDB</span>
<span class="sd">        file.</span>
<span class="sd">    :param ligand_smiles: SMILES string of the input ligand.</span>
<span class="sd">    :param output_filepath: Path to the output bash script file.</span>
<span class="sd">    :param cfg: Configuration dictionary for runtime arguments.</span>
<span class="sd">    :param generate_hpc_scripts: Whether to generate HPC scripts for</span>
<span class="sd">        DynamicBind.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">bash_script_content</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;&quot;&quot;#!/bin/bash</span>
<span class="si">{</span><span class="n">insert_hpc_headers</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s1">&#39;dynamicbind&#39;</span><span class="p">)</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">generate_hpc_scripts</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;source /home/$USER/mambaforge/etc/profile.d/conda.sh&#39;</span><span class="si">}</span>
<span class="s2">conda activate </span><span class="si">{</span><span class="s2">&quot;$project_dir/MULTICOM_ligand/&quot;</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">generate_hpc_scripts</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;MULTICOM_ligand&#39;</span><span class="si">}</span>

<span class="s2"># command to run dynamicbind_input_preparation.py</span>
<span class="s2">python multicom_ligand/data/dynamicbind_input_preparation.py </span><span class="se">\\</span>
<span class="s2">    output_csv_dir=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">dynamicbind_input_ligand_csv_dir</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    input_protein_data_dir=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">dynamicbind_input_protein_data_dir</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    protein_filepath=&quot;</span><span class="si">{</span><span class="n">protein_filepath</span><span class="si">}</span><span class="s2">&quot; </span><span class="se">\\</span>
<span class="s2">    ligand_smiles=&#39;&quot;</span><span class="si">{</span><span class="n">ligand_smiles</span><span class="si">}</span><span class="s2">&quot;&#39;</span>

<span class="s2"># command to run dynamicbind_inference.py</span>
<span class="s2">echo &quot;Calling dynamicbind_inference.py!&quot;</span>
<span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">dynamicbind_python_exec_path</span><span class="si">}</span><span class="s2"> multicom_ligand/models/dynamicbind_inference.py </span><span class="se">\\</span>
<span class="s2">    cuda_device_index=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">cuda_device_index</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    python_exec_path=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">dynamicbind_python_exec_path</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    dynamicbind_exec_dir=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">dynamicbind_exec_dir</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    dataset=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">dynamicbind_dataset</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    input_data_dir=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">dynamicbind_input_protein_data_dir</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    input_ligand_csv_dir=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">dynamicbind_input_ligand_csv_dir</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    samples_per_complex=</span><span class="si">{</span><span class="nb">min</span><span class="p">(</span><span class="n">cfg</span><span class="o">.</span><span class="n">dynamicbind_samples_per_complex</span><span class="p">,</span><span class="w"> </span><span class="n">cfg</span><span class="o">.</span><span class="n">max_method_predictions</span><span class="p">)</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    savings_per_complex=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">dynamicbind_savings_per_complex</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    inference_steps=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">dynamicbind_inference_steps</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    batch_size=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">dynamicbind_batch_size</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    header=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">dynamicbind_header</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    num_workers=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">dynamicbind_num_workers</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    skip_existing=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">dynamicbind_skip_existing</span><span class="si">}</span>

<span class="s2">echo &quot;Finished calling dynamicbind_inference.py!&quot;</span>
<span class="s2">    &quot;&quot;&quot;</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">output_filepath</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
        <span class="n">file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">bash_script_content</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Bash script &#39;</span><span class="si">{</span><span class="n">output_filepath</span><span class="si">}</span><span class="s2">&#39; created successfully.&quot;</span><span class="p">)</span></div>



<div class="viewcode-block" id="create_neuralplexer_bash_script">
<a class="viewcode-back" href="../../../modules/multicom_ligand.ensemble_generation.html#multicom_ligand.models.ensemble_generation.create_neuralplexer_bash_script">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">create_neuralplexer_bash_script</span><span class="p">(</span>
    <span class="n">protein_filepath</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">ligand_smiles</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">input_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">output_filepath</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">cfg</span><span class="p">:</span> <span class="n">DictConfig</span><span class="p">,</span>
    <span class="n">generate_hpc_scripts</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Create a bash script to run NeuralPLexer protein-ligand complex</span>
<span class="sd">    prediction.</span>

<span class="sd">    :param protein_filepath: Path to the input protein structure PDB</span>
<span class="sd">        file.</span>
<span class="sd">    :param ligand_smiles: SMILES string of the input ligand.</span>
<span class="sd">    :param input_id: Input ID.</span>
<span class="sd">    :param output_filepath: Path to the output bash script file.</span>
<span class="sd">    :param cfg: Configuration dictionary for runtime arguments.</span>
<span class="sd">    :param generate_hpc_scripts: Whether to generate HPC scripts for</span>
<span class="sd">        NeuralPLexer.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">bash_script_content</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;&quot;&quot;#!/bin/bash</span>
<span class="si">{</span><span class="n">insert_hpc_headers</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s1">&#39;neuralplexer&#39;</span><span class="p">)</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">generate_hpc_scripts</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;source /home/$USER/mambaforge/etc/profile.d/conda.sh&#39;</span><span class="si">}</span>
<span class="s2">conda activate </span><span class="si">{</span><span class="s2">&quot;$project_dir/MULTICOM_ligand/&quot;</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">generate_hpc_scripts</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;MULTICOM_ligand&#39;</span><span class="si">}</span>

<span class="s2"># command to run neuralplexer_input_preparation.py</span>
<span class="s2">python multicom_ligand/data/neuralplexer_input_preparation.py </span><span class="se">\\</span>
<span class="s2">    output_csv_path=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">neuralplexer_input_csv_path</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    input_receptor=&#39;</span><span class="si">{</span><span class="n">protein_filepath</span><span class="si">}</span><span class="s2">&#39; </span><span class="se">\\</span>
<span class="s2">    input_ligand=&#39;&quot;</span><span class="si">{</span><span class="n">ligand_smiles</span><span class="si">}</span><span class="s2">&quot;&#39; </span><span class="se">\\</span>
<span class="s2">    input_template=&#39;</span><span class="si">{</span><span class="n">protein_filepath</span><span class="si">}</span><span class="s2">&#39; </span><span class="se">\\</span>
<span class="s2">    input_id=&#39;</span><span class="si">{</span><span class="n">input_id</span><span class="si">}</span><span class="s2">&#39;</span>

<span class="s2"># command to run neuralplexer_inference.py</span>
<span class="s2">echo &quot;Calling neuralplexer_inference.py!&quot;</span>
<span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">neuralplexer_python_exec_path</span><span class="si">}</span><span class="s2"> multicom_ligand/models/neuralplexer_inference.py </span><span class="se">\\</span>
<span class="s2">    python_exec_path=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">neuralplexer_python_exec_path</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    neuralplexer_exec_dir=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">neuralplexer_exec_dir</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    input_csv_path=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">neuralplexer_input_csv_path</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    skip_existing=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">neuralplexer_skip_existing</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    task=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">neuralplexer_task</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    sample_id=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">neuralplexer_sample_id</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    template_id=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">neuralplexer_template_id</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    cuda_device_index=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">cuda_device_index</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    model_checkpoint=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">neuralplexer_model_checkpoint</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    out_path=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">neuralplexer_out_path</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    n_samples=</span><span class="si">{</span><span class="nb">min</span><span class="p">(</span><span class="n">cfg</span><span class="o">.</span><span class="n">neuralplexer_n_samples</span><span class="p">,</span><span class="w"> </span><span class="n">cfg</span><span class="o">.</span><span class="n">max_method_predictions</span><span class="p">)</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    chunk_size=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">neuralplexer_chunk_size</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    num_steps=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">neuralplexer_num_steps</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    sampler=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">neuralplexer_sampler</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    start_time=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">neuralplexer_start_time</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    max_chain_encoding_k=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">neuralplexer_max_chain_encoding_k</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    exact_prior=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">neuralplexer_exact_prior</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    discard_ligand=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">neuralplexer_discard_ligand</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    discard_sdf_coords=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">neuralplexer_discard_sdf_coords</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    detect_covalent=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">neuralplexer_detect_covalent</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    use_template=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">neuralplexer_use_template</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    separate_pdb=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">neuralplexer_separate_pdb</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    rank_outputs_by_confidence=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">neuralplexer_rank_outputs_by_confidence</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    plddt_ranking_type=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">neuralplexer_plddt_ranking_type</span><span class="si">}</span>

<span class="s2">echo &quot;Finished calling neuralplexer_inference.py!&quot;</span>
<span class="s2">    &quot;&quot;&quot;</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">output_filepath</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
        <span class="n">file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">bash_script_content</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Bash script &#39;</span><span class="si">{</span><span class="n">output_filepath</span><span class="si">}</span><span class="s2">&#39; created successfully.&quot;</span><span class="p">)</span></div>



<div class="viewcode-block" id="create_flowdock_bash_script">
<a class="viewcode-back" href="../../../modules/multicom_ligand.ensemble_generation.html#multicom_ligand.models.ensemble_generation.create_flowdock_bash_script">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">create_flowdock_bash_script</span><span class="p">(</span>
    <span class="n">protein_filepath</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">ligand_smiles</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">input_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">output_filepath</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">cfg</span><span class="p">:</span> <span class="n">DictConfig</span><span class="p">,</span>
    <span class="n">generate_hpc_scripts</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">auxiliary_estimation_input_dir</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Create a bash script to run FlowDock protein-ligand complex prediction.</span>

<span class="sd">    :param protein_filepath: Path to the input protein structure PDB</span>
<span class="sd">        file.</span>
<span class="sd">    :param ligand_smiles: SMILES string of the input ligand.</span>
<span class="sd">    :param input_id: Input ID.</span>
<span class="sd">    :param output_filepath: Path to the output bash script file.</span>
<span class="sd">    :param cfg: Configuration dictionary for runtime arguments.</span>
<span class="sd">    :param generate_hpc_scripts: Whether to generate HPC scripts for</span>
<span class="sd">        FlowDock.</span>
<span class="sd">    :param auxiliary_estimation_input_dir: Optional path to the</span>
<span class="sd">        directory containing auxiliary estimation input files for</span>
<span class="sd">        FlowDock.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">bash_script_content</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;&quot;&quot;#!/bin/bash</span>
<span class="si">{</span><span class="n">insert_hpc_headers</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s1">&#39;flowdock&#39;</span><span class="p">)</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">generate_hpc_scripts</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;source /home/$USER/mambaforge/etc/profile.d/conda.sh&#39;</span><span class="si">}</span>
<span class="s2">conda activate </span><span class="si">{</span><span class="s2">&quot;$project_dir/MULTICOM_ligand/&quot;</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">generate_hpc_scripts</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;MULTICOM_ligand&#39;</span><span class="si">}</span>

<span class="s2"># command to run flowdock_input_preparation.py</span>
<span class="s2">python multicom_ligand/data/flowdock_input_preparation.py </span><span class="se">\\</span>
<span class="s2">    output_csv_path=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">flowdock_input_csv_path</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    input_receptor=&#39;</span><span class="si">{</span><span class="n">protein_filepath</span><span class="si">}</span><span class="s2">&#39; </span><span class="se">\\</span>
<span class="s2">    input_ligand=&#39;&quot;</span><span class="si">{</span><span class="n">ligand_smiles</span><span class="si">}</span><span class="s2">&quot;&#39; </span><span class="se">\\</span>
<span class="s2">    input_template=&#39;</span><span class="si">{</span><span class="n">protein_filepath</span><span class="si">}</span><span class="s2">&#39; </span><span class="se">\\</span>
<span class="s2">    input_id=&#39;</span><span class="si">{</span><span class="n">input_id</span><span class="si">}</span><span class="s2">&#39;</span>

<span class="s2"># command to run flowdock_inference.py</span>
<span class="s2">echo &quot;Calling flowdock_inference.py!&quot;</span>
<span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">flowdock_python_exec_path</span><span class="si">}</span><span class="s2"> multicom_ligand/models/flowdock_inference.py </span><span class="se">\\</span>
<span class="s2">    python_exec_path=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">flowdock_python_exec_path</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    flowdock_exec_dir=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">flowdock_exec_dir</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    input_csv_path=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">flowdock_input_csv_path</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    skip_existing=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">flowdock_skip_existing</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    sampling_task=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">flowdock_sampling_task</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    sample_id=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">flowdock_sample_id</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    cuda_device_index=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">cuda_device_index</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    model_checkpoint=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">flowdock_model_checkpoint</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    out_path=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">flowdock_out_path</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    n_samples=</span><span class="si">{</span><span class="nb">min</span><span class="p">(</span><span class="n">cfg</span><span class="o">.</span><span class="n">flowdock_n_samples</span><span class="p">,</span><span class="w"> </span><span class="n">cfg</span><span class="o">.</span><span class="n">max_method_predictions</span><span class="p">)</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    chunk_size=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">flowdock_chunk_size</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    num_steps=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">flowdock_num_steps</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    latent_model=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">flowdock_latent_model</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">cfg</span><span class="o">.</span><span class="n">flowdock_latent_model</span><span class="w"> </span><span class="ow">is</span><span class="w"> </span><span class="ow">not</span><span class="w"> </span><span class="kc">None</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;null&#39;</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    sampler=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">flowdock_sampler</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    sampler_eta=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">flowdock_sampler_eta</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    start_time=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">flowdock_start_time</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    max_chain_encoding_k=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">flowdock_max_chain_encoding_k</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    exact_prior=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">flowdock_exact_prior</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    discard_ligand=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">flowdock_discard_ligand</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    discard_sdf_coords=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">flowdock_discard_sdf_coords</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    detect_covalent=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">flowdock_detect_covalent</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    use_template=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">flowdock_use_template</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    separate_pdb=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">flowdock_separate_pdb</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    rank_outputs_by_confidence=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">flowdock_rank_outputs_by_confidence</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    plddt_ranking_type=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">flowdock_plddt_ranking_type</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    visualize_sample_trajectories=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">flowdock_visualize_sample_trajectories</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    auxiliary_estimation_only=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">flowdock_auxiliary_estimation_only</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    auxiliary_estimation_input_dir=</span><span class="si">{</span><span class="n">auxiliary_estimation_input_dir</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">auxiliary_estimation_input_dir</span><span class="w"> </span><span class="ow">is</span><span class="w"> </span><span class="ow">not</span><span class="w"> </span><span class="kc">None</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;null&#39;</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    esmfold_chunk_size=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">flowdock_esmfold_chunk_size</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">cfg</span><span class="o">.</span><span class="n">flowdock_esmfold_chunk_size</span><span class="w"> </span><span class="ow">is</span><span class="w"> </span><span class="ow">not</span><span class="w"> </span><span class="kc">None</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;null&#39;</span><span class="si">}</span>

<span class="s2">echo &quot;Finished calling flowdock_inference.py!&quot;</span>
<span class="s2">    &quot;&quot;&quot;</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">output_filepath</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
        <span class="n">file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">bash_script_content</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Bash script &#39;</span><span class="si">{</span><span class="n">output_filepath</span><span class="si">}</span><span class="s2">&#39; created successfully.&quot;</span><span class="p">)</span></div>



<div class="viewcode-block" id="rfaa_get_chain_letter">
<a class="viewcode-back" href="../../../modules/multicom_ligand.ensemble_generation.html#multicom_ligand.models.ensemble_generation.rfaa_get_chain_letter">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">rfaa_get_chain_letter</span><span class="p">(</span><span class="n">index</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Get the RFAA chain letter based on index.&quot;&quot;&quot;</span>
    <span class="n">alphabet</span> <span class="o">=</span> <span class="s2">&quot;abcdefghijklmnopqrstuvwxyz&quot;</span>
    <span class="n">num_chars</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">alphabet</span><span class="p">)</span>
    <span class="n">num_single_chars</span> <span class="o">=</span> <span class="n">num_chars</span> <span class="o">-</span> <span class="mi">1</span>  <span class="c1"># Exclude &#39;z&#39;</span>
    <span class="k">if</span> <span class="n">index</span> <span class="o">&lt;=</span> <span class="n">num_single_chars</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">alphabet</span><span class="p">[</span><span class="n">index</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># Use multiple characters for chain IDs beyond &#39;z&#39;, such as &#39;aa&#39;, &#39;ab&#39;, &#39;ac&#39;, etc.</span>
        <span class="n">num_multiple_chars</span> <span class="o">=</span> <span class="n">index</span> <span class="o">-</span> <span class="n">num_single_chars</span>
        <span class="n">num_full_sets</span> <span class="o">=</span> <span class="n">num_multiple_chars</span> <span class="o">//</span> <span class="n">num_chars</span>
        <span class="n">remainder</span> <span class="o">=</span> <span class="n">num_multiple_chars</span> <span class="o">%</span> <span class="n">num_chars</span>
        <span class="k">if</span> <span class="n">remainder</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">alphabet</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="n">num_full_sets</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">alphabet</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="n">num_full_sets</span> <span class="o">+</span> <span class="n">alphabet</span><span class="p">[</span><span class="n">remainder</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span></div>



<div class="viewcode-block" id="dynamically_build_rfaa_input_config">
<a class="viewcode-back" href="../../../modules/multicom_ligand.ensemble_generation.html#multicom_ligand.models.ensemble_generation.dynamically_build_rfaa_input_config">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">dynamically_build_rfaa_input_config</span><span class="p">(</span>
    <span class="n">fasta_filepaths</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
    <span class="n">fasta_types</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
    <span class="n">sdf_filepaths</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]],</span>
    <span class="n">input_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">cfg</span><span class="p">:</span> <span class="n">DictConfig</span><span class="p">,</span>
    <span class="n">smiles_strings</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Dynamically build the RoseTTAFold-All-Atom inference configuration file</span>
<span class="sd">    for input proteins, RNA, DNA, and ligands.</span>

<span class="sd">    :param fasta_filepaths: List of FASTA filepaths.</span>
<span class="sd">    :param fasta_types: List of FASTA chain molecule types.</span>
<span class="sd">    :param sdf_filepaths: List of optional SDF filepaths.</span>
<span class="sd">    :param input_id: Input ID.</span>
<span class="sd">    :param cfg: Configuration dictionary for runtime arguments.</span>
<span class="sd">    :param smiles_strings: Optional list of SMILES strings of the input</span>
<span class="sd">        ligands to use directly.</span>
<span class="sd">    :return: Path to the dynamically built configuration file.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Build {chain_molecule_type}_inputs section dynamically</span>
    <span class="n">molecule_inputs_content</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
    <span class="k">assert</span> <span class="nb">all</span><span class="p">(</span><span class="n">fasta_type</span> <span class="o">==</span> <span class="s2">&quot;Protein&quot;</span> <span class="k">for</span> <span class="n">fasta_type</span> <span class="ow">in</span> <span class="n">fasta_types</span><span class="p">)</span> <span class="ow">or</span> <span class="nb">all</span><span class="p">(</span>
        <span class="n">fasta_type</span> <span class="ow">in</span> <span class="p">{</span><span class="s2">&quot;RNA&quot;</span><span class="p">,</span> <span class="s2">&quot;DNA&quot;</span><span class="p">}</span> <span class="k">for</span> <span class="n">fasta_type</span> <span class="ow">in</span> <span class="n">fasta_types</span>
    <span class="p">),</span> <span class="s2">&quot;RFAA only supports protein, RNA, or DNA inputs, where DNA or RNA inputs must be grouped together.&quot;</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">fasta_filepath</span><span class="p">,</span> <span class="n">fasta_type</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">fasta_filepaths</span><span class="p">,</span> <span class="n">fasta_types</span><span class="p">),</span> <span class="n">start</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
        <span class="n">chain_letter</span> <span class="o">=</span> <span class="n">rfaa_get_chain_letter</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
        <span class="n">molecule_inputs_content</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;&quot;&quot;  </span><span class="si">{</span><span class="n">chain_letter</span><span class="si">}</span><span class="s2">:</span>
<span class="s2">    </span><span class="si">{</span><span class="s1">&#39;fasta&#39;</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">fasta_type</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="p">{</span><span class="s1">&#39;DNA&#39;</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;RNA&#39;</span><span class="p">}</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;fasta_file&#39;</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">fasta_filepath</span><span class="si">}</span>
<span class="s2">&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">fasta_type</span> <span class="o">==</span> <span class="s2">&quot;DNA&quot;</span><span class="p">:</span>
            <span class="n">molecule_inputs_content</span> <span class="o">+=</span> <span class="s2">&quot;&quot;&quot;    input_type: &quot;dna&quot;</span>
<span class="s2">&quot;&quot;&quot;</span>
        <span class="k">elif</span> <span class="n">fasta_type</span> <span class="o">==</span> <span class="s2">&quot;RNA&quot;</span><span class="p">:</span>
            <span class="n">molecule_inputs_content</span> <span class="o">+=</span> <span class="s2">&quot;&quot;&quot;    input_type: &quot;rna&quot;</span>
<span class="s2">&quot;&quot;&quot;</span>

    <span class="c1"># Build sm_inputs section dynamically</span>
    <span class="n">sm_inputs_content</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
    <span class="k">if</span> <span class="n">smiles_strings</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">assert</span> <span class="p">(</span>
            <span class="n">sdf_filepaths</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
        <span class="p">),</span> <span class="s2">&quot;No SMILES strings or SDF filepaths found for input ligands.&quot;</span>
        <span class="n">smiles_strings</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">Chem</span><span class="o">.</span><span class="n">MolToSmiles</span><span class="p">(</span><span class="n">Chem</span><span class="o">.</span><span class="n">MolFromMolFile</span><span class="p">(</span><span class="n">sdf_filepath</span><span class="p">))</span> <span class="k">for</span> <span class="n">sdf_filepath</span> <span class="ow">in</span> <span class="n">sdf_filepaths</span>
        <span class="p">]</span>
    <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">smiles_strings</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;No SMILES strings found for input ligands.&quot;</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">smiles</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span>
        <span class="n">smiles_strings</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">fasta_filepaths</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span>
    <span class="p">):</span>  <span class="c1"># Starting index from len(fasta_filepaths)+1</span>
        <span class="n">chain_letter</span> <span class="o">=</span> <span class="n">rfaa_get_chain_letter</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
        <span class="n">sm_inputs_content</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;&quot;&quot;  </span><span class="si">{</span><span class="n">chain_letter</span><span class="si">}</span><span class="s2">:</span>
<span class="s2">    input: &#39;</span><span class="si">{</span><span class="n">smiles</span><span class="si">}</span><span class="s2">&#39;</span>
<span class="s2">    input_type: &quot;smiles&quot;</span>
<span class="s2">&quot;&quot;&quot;</span>

    <span class="n">config_file_content</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;&quot;&quot;# RoseTTAFold-All-Atom inference configuration file for input &#39;</span><span class="si">{</span><span class="n">input_id</span><span class="si">}</span><span class="s2">&#39;</span>
<span class="s2">defaults:</span>
<span class="s2">  - base</span>
<span class="s2">job_name: &quot;</span><span class="si">{</span><span class="n">input_id</span><span class="si">}</span><span class="s2">&quot;</span>

<span class="si">{</span><span class="s1">&#39;na&#39;</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="nb">any</span><span class="p">(</span><span class="n">fasta_type</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="p">{</span><span class="s1">&#39;DNA&#39;</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;RNA&#39;</span><span class="p">}</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="n">fasta_type</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">fasta_types</span><span class="p">)</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;protein&#39;</span><span class="si">}</span><span class="s2">_inputs:</span>
<span class="s2">  </span><span class="si">{</span><span class="n">molecule_inputs_content</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="si">}</span>

<span class="s2">sm_inputs:</span>
<span class="s2">  </span><span class="si">{</span><span class="n">sm_inputs_content</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="si">}</span>
<span class="s2">&quot;&quot;&quot;</span>

    <span class="n">config_filepath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
        <span class="p">(</span><span class="n">cfg</span><span class="o">.</span><span class="n">rfaa_config_dir</span> <span class="k">if</span> <span class="n">cfg</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;rfaa_config_dir&quot;</span><span class="p">)</span> <span class="k">else</span> <span class="n">cfg</span><span class="o">.</span><span class="n">config_dir</span><span class="p">),</span>
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">input_id</span><span class="si">}</span><span class="s2">_rfaa_inference.yaml&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">config_filepath</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
        <span class="n">file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">config_file_content</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">config_filepath</span></div>



<div class="viewcode-block" id="create_rfaa_bash_script">
<a class="viewcode-back" href="../../../modules/multicom_ligand.ensemble_generation.html#multicom_ligand.models.ensemble_generation.create_rfaa_bash_script">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">create_rfaa_bash_script</span><span class="p">(</span>
    <span class="n">fasta_filepaths</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
    <span class="n">fasta_types</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
    <span class="n">sdf_filepaths</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]],</span>
    <span class="n">input_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">cfg</span><span class="p">:</span> <span class="n">DictConfig</span><span class="p">,</span>
    <span class="n">output_filepath</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">smiles_strings</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">generate_hpc_scripts</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Create a bash script to run RoseTTAFold-All-Atom protein-ligand complex</span>
<span class="sd">    prediction.</span>

<span class="sd">    :param fasta_filepaths: List of FASTA filepaths.</span>
<span class="sd">    :param fasta_types: List of FASTA chain molecule types.</span>
<span class="sd">    :param sdf_filepaths: List of optional SDF filepaths.</span>
<span class="sd">    :param input_id: Input ID.</span>
<span class="sd">    :param cfg: Configuration dictionary for runtime arguments.</span>
<span class="sd">    :param output_filepath: Optional path to the output bash script</span>
<span class="sd">        file.</span>
<span class="sd">    :param smiles_strings: Optional list of SMILES strings of the input</span>
<span class="sd">        ligands to use directly.</span>
<span class="sd">    :param generate_hpc_scripts: Whether to generate HPC scripts for</span>
<span class="sd">        RoseTTAFold-All-Atom.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">if</span> <span class="n">output_filepath</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">output_filepath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cfg</span><span class="o">.</span><span class="n">output_dir</span><span class="p">,</span> <span class="n">input_id</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">input_id</span><span class="si">}</span><span class="s2">_rfaa_inference.sh&quot;</span><span class="p">)</span>
        <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">output_filepath</span><span class="p">),</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="c1"># Dynamically build the RoseTTAFold-All-Atom inference configuration file</span>
    <span class="n">config_filepath</span> <span class="o">=</span> <span class="n">dynamically_build_rfaa_input_config</span><span class="p">(</span>
        <span class="n">fasta_filepaths</span><span class="o">=</span><span class="n">fasta_filepaths</span><span class="p">,</span>
        <span class="n">fasta_types</span><span class="o">=</span><span class="n">fasta_types</span><span class="p">,</span>
        <span class="n">sdf_filepaths</span><span class="o">=</span><span class="n">sdf_filepaths</span><span class="p">,</span>
        <span class="n">input_id</span><span class="o">=</span><span class="n">input_id</span><span class="p">,</span>
        <span class="n">cfg</span><span class="o">=</span><span class="n">cfg</span><span class="p">,</span>
        <span class="n">smiles_strings</span><span class="o">=</span><span class="n">smiles_strings</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">bash_script_content</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;&quot;&quot;#!/bin/bash -l</span>
<span class="si">{</span><span class="n">insert_hpc_headers</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s1">&#39;rfaa&#39;</span><span class="p">,</span><span class="w"> </span><span class="n">time_limit</span><span class="o">=</span><span class="s1">&#39;0-12:00:00&#39;</span><span class="p">)</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">generate_hpc_scripts</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;source /home/$USER/mambaforge/etc/profile.d/conda.sh&#39;</span><span class="si">}</span>
<span class="s2">conda activate </span><span class="si">{</span><span class="s2">&quot;$project_dir/forks/RoseTTAFold-All-Atom/RFAA/&quot;</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">generate_hpc_scripts</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;forks/RoseTTAFold-All-Atom/RFAA/&#39;</span><span class="si">}</span>
<span class="s2">echo &quot;Beginning RoseTTAFold-All-Atom inference for input &#39;</span><span class="si">{</span><span class="n">input_id</span><span class="si">}</span><span class="s2">&#39;!&quot;</span>

<span class="s2"># command to run RoseTTAFold-All-Atom inference</span>
<span class="s2">cd </span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">rfaa_exec_dir</span><span class="si">}</span>
<span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">rfaa_python_exec_path</span><span class="si">}</span><span class="s2"> -m rf2aa.run_inference --config-name </span><span class="si">{</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">config_filepath</span><span class="p">)</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    loader_params.MAXCYCLE=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">rfaa_max_cycles</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    output_path=</span><span class="si">{</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">output_filepath</span><span class="p">)</span><span class="si">}</span>

<span class="s2"># clean up temporary config file</span>
<span class="s2">rm </span><span class="si">{</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cfg</span><span class="o">.</span><span class="n">rfaa_config_dir</span><span class="p">,</span><span class="w"> </span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">input_id</span><span class="si">}</span><span class="s2">_rfaa_inference.yaml&quot;</span><span class="p">)</span><span class="si">}</span>
<span class="s2">rm -r </span><span class="si">{</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cfg</span><span class="o">.</span><span class="n">rfaa_config_dir</span><span class="p">,</span><span class="w"> </span><span class="n">input_id</span><span class="p">)</span><span class="si">}</span>
<span class="s2">echo &quot;Finished RoseTTAFold-All-Atom inference for input &#39;</span><span class="si">{</span><span class="n">input_id</span><span class="si">}</span><span class="s2">&#39;!&quot;</span>
<span class="s2">&quot;&quot;&quot;</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">output_filepath</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
        <span class="n">file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">bash_script_content</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Bash script &#39;</span><span class="si">{</span><span class="n">output_filepath</span><span class="si">}</span><span class="s2">&#39; created successfully.&quot;</span><span class="p">)</span></div>



<div class="viewcode-block" id="create_vina_bash_script">
<a class="viewcode-back" href="../../../modules/multicom_ligand.ensemble_generation.html#multicom_ligand.models.ensemble_generation.create_vina_bash_script">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">create_vina_bash_script</span><span class="p">(</span>
    <span class="n">binding_site_method</span><span class="p">:</span> <span class="n">Literal</span><span class="p">[</span><span class="s2">&quot;diffdock&quot;</span><span class="p">,</span> <span class="s2">&quot;fabind&quot;</span><span class="p">,</span> <span class="s2">&quot;dynamicbind&quot;</span><span class="p">,</span> <span class="s2">&quot;neuralplexer&quot;</span><span class="p">,</span> <span class="s2">&quot;rfaa&quot;</span><span class="p">],</span>
    <span class="n">protein_filepath</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">ligand_filepath</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">apo_protein_filepath</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">input_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">output_filepath</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">cfg</span><span class="p">:</span> <span class="n">DictConfig</span><span class="p">,</span>
    <span class="n">generate_hpc_scripts</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Create a bash script to run Vina-based protein-ligand complex</span>
<span class="sd">    prediction.</span>

<span class="sd">    :param binding_site_method: Name of the method used to predict the</span>
<span class="sd">        binding site.</span>
<span class="sd">    :param protein_filepath: Path to the input protein structure PDB</span>
<span class="sd">        file.</span>
<span class="sd">    :param ligand_filepath: Path to the input ligand structure SDF file.</span>
<span class="sd">    :param apo_protein_filepath: Path to the predicted apo protein</span>
<span class="sd">        structure PDB file.</span>
<span class="sd">    :param input_id: Input ID.</span>
<span class="sd">    :param output_filepath: Path to the output bash script file.</span>
<span class="sd">    :param cfg: Configuration dictionary for runtime arguments.</span>
<span class="sd">    :param generate_hpc_scripts: Whether to generate HPC scripts for</span>
<span class="sd">        Vina.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">bash_script_content</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;&quot;&quot;#!/bin/bash -l</span>
<span class="si">{</span><span class="n">insert_hpc_headers</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s1">&#39;vina&#39;</span><span class="p">)</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">generate_hpc_scripts</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;source /home/$USER/mambaforge/etc/profile.d/conda.sh&#39;</span><span class="si">}</span>
<span class="s2">conda activate </span><span class="si">{</span><span class="s2">&quot;$project_dir/MULTICOM_ligand/&quot;</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">generate_hpc_scripts</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;MULTICOM_ligand&#39;</span><span class="si">}</span>

<span class="s2"># command to run vina_inference.py</span>
<span class="s2">echo &quot;Calling vina_inference.py!&quot;</span>
<span class="s2">python3 multicom_ligand/models/vina_inference.py </span><span class="se">\\</span>
<span class="s2">    method=</span><span class="si">{</span><span class="n">binding_site_method</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    python2_exec_path=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">vina_python2_exec_path</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    prepare_receptor_script_path=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">vina_prepare_receptor_script_path</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    output_dir=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">vina_output_dir</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    cpu=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">vina_cpu</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    seed=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">vina_seed</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    exhaustiveness=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">vina_exhaustiveness</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    ligand_ligand_distance_threshold=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">vina_ligand_ligand_distance_threshold</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    protein_ligand_distance_threshold=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">vina_protein_ligand_distance_threshold</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    binding_site_size_x=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">vina_binding_site_size_x</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    binding_site_size_y=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">vina_binding_site_size_y</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    binding_site_size_z=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">vina_binding_site_size_z</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    binding_site_spacing=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">vina_binding_site_spacing</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    num_modes=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">vina_num_modes</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    skip_existing=</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">vina_skip_existing</span><span class="si">}</span><span class="s2"> </span><span class="se">\\</span>
<span class="s2">    protein_filepath=&quot;</span><span class="si">{</span><span class="n">protein_filepath</span><span class="si">}</span><span class="s2">&quot; </span><span class="se">\\</span>
<span class="s2">    ligand_filepath=&quot;</span><span class="si">{</span><span class="n">ligand_filepath</span><span class="si">}</span><span class="s2">&quot; </span><span class="se">\\</span>
<span class="s2">    apo_protein_filepath=&quot;</span><span class="si">{</span><span class="n">apo_protein_filepath</span><span class="si">}</span><span class="s2">&quot; </span><span class="se">\\</span>
<span class="s2">    input_id=&quot;</span><span class="si">{</span><span class="n">input_id</span><span class="si">}</span><span class="s2">&quot; </span><span class="se">\\</span>

<span class="s2">echo &quot;Finished calling vina_inference.py!&quot;</span>
<span class="s2">    &quot;&quot;&quot;</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">output_filepath</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
        <span class="n">file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">bash_script_content</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Bash script &#39;</span><span class="si">{</span><span class="n">output_filepath</span><span class="si">}</span><span class="s2">&#39; created successfully.&quot;</span><span class="p">)</span></div>



<div class="viewcode-block" id="generate_method_prediction_script">
<a class="viewcode-back" href="../../../modules/multicom_ligand.ensemble_generation.html#multicom_ligand.models.ensemble_generation.generate_method_prediction_script">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">generate_method_prediction_script</span><span class="p">(</span>
    <span class="n">method</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">protein_filepath</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">ligand_smiles</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">input_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">output_filepath</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">cfg</span><span class="p">:</span> <span class="n">DictConfig</span><span class="p">,</span>
    <span class="n">generate_hpc_scripts</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
    <span class="n">method_filepaths_mapping</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">auxiliary_estimation_input_dir</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Generate a script to run the method&#39;s protein-ligand complex prediction.</span>

<span class="sd">    :param method: Name of the method to generate a prediction script</span>
<span class="sd">        for.</span>
<span class="sd">    :param protein_filepath: Path to the input protein structure PDB</span>
<span class="sd">        file.</span>
<span class="sd">    :param ligand_smiles: SMILES string of the input ligand.</span>
<span class="sd">    :param input_id: Input ID.</span>
<span class="sd">    :param output_filepath: Path to the output Bash script file.</span>
<span class="sd">    :param cfg: Configuration dictionary for runtime arguments.</span>
<span class="sd">    :param generate_hpc_scripts: Whether to generate HPC scripts for the</span>
<span class="sd">        method.</span>
<span class="sd">    :param method_filepaths_mapping: Optional mapping of method names to</span>
<span class="sd">        a list of tuples of protein and ligand filepaths.</span>
<span class="sd">    :param auxiliary_estimation_input_dir: Optional path to the</span>
<span class="sd">        directory containing auxiliary estimation input files e.g., for</span>
<span class="sd">        FlowDock.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">extract_chains_to_fasta_files</span><span class="p">(</span><span class="n">pdb_filepath</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Extract individual chains from a macromolecule file and save them as</span>
<span class="sd">        separate FASTA files.</span>

<span class="sd">        :param pdb_filepath: Path to the PDB file.</span>
<span class="sd">        :return: List of paths to the extracted FASTA files and</span>
<span class="sd">            corresponding chain molecule types.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Parse the macromolecule file using Bio.PDB</span>
        <span class="n">parser</span> <span class="o">=</span> <span class="n">PDBParser</span><span class="p">()</span>
        <span class="n">structure</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">get_structure</span><span class="p">(</span><span class="s2">&quot;structure&quot;</span><span class="p">,</span> <span class="n">pdb_filepath</span><span class="p">)</span>
        <span class="n">sequences</span><span class="p">,</span> <span class="n">sequence_types</span> <span class="o">=</span> <span class="n">extract_sequences_from_macromolecule_structure_file</span><span class="p">(</span>
            <span class="n">pdb_filepath</span><span class="p">,</span> <span class="n">structure</span><span class="o">=</span><span class="n">structure</span>
        <span class="p">)</span>

        <span class="c1"># Create a temporary directory to store the FASTA files</span>
        <span class="n">temp_dir</span> <span class="o">=</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">mkdtemp</span><span class="p">()</span>

        <span class="c1"># Iterate over each chain and save it as a separate FASTA file</span>
        <span class="n">fasta_filepaths</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">model</span> <span class="ow">in</span> <span class="n">structure</span><span class="p">:</span>
            <span class="n">model_chains</span> <span class="o">=</span> <span class="p">[</span><span class="n">chain</span> <span class="k">for</span> <span class="n">chain</span> <span class="ow">in</span> <span class="n">model</span><span class="p">]</span>
            <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">model_chains</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span>
                <span class="n">sequences</span>
            <span class="p">),</span> <span class="s2">&quot;For RFAA, numbers of BioPython chains and parsed sequences do not match.&quot;</span>
            <span class="k">for</span> <span class="n">chain_index</span><span class="p">,</span> <span class="n">chain</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">model_chains</span><span class="p">):</span>
                <span class="n">fasta_filename</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">Path</span><span class="p">(</span><span class="n">pdb_filepath</span><span class="p">)</span><span class="o">.</span><span class="n">stem</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">chain</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">.fasta&quot;</span>
                <span class="n">fasta_filepath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">temp_dir</span><span class="p">,</span> <span class="n">fasta_filename</span><span class="p">)</span>
                <span class="n">chain_sequence</span> <span class="o">=</span> <span class="n">sequences</span><span class="p">[</span><span class="n">chain_index</span><span class="p">]</span>
                <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">fasta_filepath</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&gt;</span><span class="si">{</span><span class="n">chain</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="se">\n</span><span class="si">{</span><span class="n">chain_sequence</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">fasta_filepaths</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fasta_filepath</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">fasta_filepaths</span><span class="p">,</span> <span class="n">sequence_types</span>

    <span class="k">if</span> <span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;diffdock&quot;</span><span class="p">:</span>
        <span class="n">create_diffdock_bash_script</span><span class="p">(</span>
            <span class="n">protein_filepath</span><span class="p">,</span>
            <span class="n">ligand_smiles</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span>
                <span class="s2">&quot;:&quot;</span><span class="p">,</span> <span class="s2">&quot;|&quot;</span>
            <span class="p">),</span>  <span class="c1"># NOTE: DiffDock supports multi-ligands using the separator &quot;|&quot;</span>
            <span class="n">input_id</span><span class="p">,</span>
            <span class="n">output_filepath</span><span class="p">,</span>
            <span class="n">cfg</span><span class="p">,</span>
            <span class="n">generate_hpc_scripts</span><span class="o">=</span><span class="n">generate_hpc_scripts</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="k">elif</span> <span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;dynamicbind&quot;</span><span class="p">:</span>
        <span class="n">create_dynamicbind_bash_script</span><span class="p">(</span>
            <span class="n">protein_filepath</span><span class="p">,</span>
            <span class="n">ligand_smiles</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span>
                <span class="s2">&quot;:&quot;</span><span class="p">,</span> <span class="s2">&quot;|&quot;</span>
            <span class="p">),</span>  <span class="c1"># NOTE: DynamicBind supports multi-ligands using the separator &quot;|&quot;</span>
            <span class="n">output_filepath</span><span class="p">,</span>
            <span class="n">cfg</span><span class="p">,</span>
            <span class="n">generate_hpc_scripts</span><span class="o">=</span><span class="n">generate_hpc_scripts</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="k">elif</span> <span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;neuralplexer&quot;</span><span class="p">:</span>
        <span class="n">create_neuralplexer_bash_script</span><span class="p">(</span>
            <span class="n">protein_filepath</span><span class="p">,</span>
            <span class="n">ligand_smiles</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span>
                <span class="s2">&quot;:&quot;</span><span class="p">,</span> <span class="s2">&quot;|&quot;</span>
            <span class="p">),</span>  <span class="c1"># NOTE: NeuralPLexer supports multi-ligands using the separator &quot;|&quot;</span>
            <span class="n">input_id</span><span class="p">,</span>
            <span class="n">output_filepath</span><span class="p">,</span>
            <span class="n">cfg</span><span class="p">,</span>
            <span class="n">generate_hpc_scripts</span><span class="o">=</span><span class="n">generate_hpc_scripts</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="k">elif</span> <span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;flowdock&quot;</span><span class="p">:</span>
        <span class="n">create_flowdock_bash_script</span><span class="p">(</span>
            <span class="n">protein_filepath</span><span class="p">,</span>
            <span class="n">ligand_smiles</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span>
                <span class="s2">&quot;:&quot;</span><span class="p">,</span> <span class="s2">&quot;|&quot;</span>
            <span class="p">),</span>  <span class="c1"># NOTE: FlowDock supports multi-ligands using the separator &quot;|&quot;</span>
            <span class="n">input_id</span><span class="p">,</span>
            <span class="n">output_filepath</span><span class="p">,</span>
            <span class="n">cfg</span><span class="p">,</span>
            <span class="n">generate_hpc_scripts</span><span class="o">=</span><span class="n">generate_hpc_scripts</span><span class="p">,</span>
            <span class="n">auxiliary_estimation_input_dir</span><span class="o">=</span><span class="n">auxiliary_estimation_input_dir</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="k">elif</span> <span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;rfaa&quot;</span><span class="p">:</span>
        <span class="n">fasta_filepaths</span><span class="p">,</span> <span class="n">fasta_types</span> <span class="o">=</span> <span class="n">extract_chains_to_fasta_files</span><span class="p">(</span><span class="n">protein_filepath</span><span class="p">)</span>
        <span class="n">smiles_strings</span> <span class="o">=</span> <span class="n">ligand_smiles</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;:&quot;</span><span class="p">)</span> <span class="k">if</span> <span class="s2">&quot;:&quot;</span> <span class="ow">in</span> <span class="n">ligand_smiles</span> <span class="k">else</span> <span class="p">[</span><span class="n">ligand_smiles</span><span class="p">]</span>
        <span class="n">create_rfaa_bash_script</span><span class="p">(</span>
            <span class="n">fasta_filepaths</span><span class="o">=</span><span class="n">fasta_filepaths</span><span class="p">,</span>
            <span class="n">fasta_types</span><span class="o">=</span><span class="n">fasta_types</span><span class="p">,</span>
            <span class="n">sdf_filepaths</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="n">input_id</span><span class="o">=</span><span class="n">input_id</span><span class="p">,</span>
            <span class="n">cfg</span><span class="o">=</span><span class="n">cfg</span><span class="p">,</span>
            <span class="n">output_filepath</span><span class="o">=</span><span class="n">output_filepath</span><span class="p">,</span>
            <span class="n">smiles_strings</span><span class="o">=</span><span class="n">smiles_strings</span><span class="p">,</span>
            <span class="n">generate_hpc_scripts</span><span class="o">=</span><span class="n">generate_hpc_scripts</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="k">elif</span> <span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;vina&quot;</span><span class="p">:</span>
        <span class="k">assert</span> <span class="p">(</span>
            <span class="n">cfg</span><span class="o">.</span><span class="n">generate_vina_scripts</span> <span class="ow">and</span> <span class="n">cfg</span><span class="o">.</span><span class="n">resume</span>
        <span class="p">),</span> <span class="s2">&quot;Vina predictions must be resumed from prior method predictions.&quot;</span>
        <span class="k">for</span> <span class="n">binding_site_method</span> <span class="ow">in</span> <span class="n">cfg</span><span class="o">.</span><span class="n">vina_binding_site_methods</span><span class="p">:</span>
            <span class="c1"># NOTE: these correspond to selecting the top-ranked method predictions (since the method predictions are already rank-ordered)</span>
            <span class="n">method_protein_filepath</span> <span class="o">=</span> <span class="n">method_filepaths_mapping</span><span class="p">[</span><span class="n">binding_site_method</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">method_ligand_filepath</span> <span class="o">=</span> <span class="n">method_filepaths_mapping</span><span class="p">[</span><span class="n">binding_site_method</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>
            <span class="n">vina_output_filepath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                <span class="n">cfg</span><span class="o">.</span><span class="n">output_bash_file_dir</span><span class="p">,</span>
                <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">output_filepath</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;vina_&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;vina_</span><span class="si">{</span><span class="n">binding_site_method</span><span class="si">}</span><span class="s2">_&quot;</span><span class="p">),</span>
            <span class="p">)</span>
            <span class="n">create_vina_bash_script</span><span class="p">(</span>
                <span class="n">binding_site_method</span><span class="p">,</span>
                <span class="n">method_protein_filepath</span><span class="p">,</span>
                <span class="n">method_ligand_filepath</span><span class="p">,</span>
                <span class="n">protein_filepath</span><span class="p">,</span>
                <span class="n">input_id</span><span class="p">,</span>
                <span class="n">vina_output_filepath</span><span class="p">,</span>
                <span class="n">cfg</span><span class="p">,</span>
                <span class="n">generate_hpc_scripts</span><span class="o">=</span><span class="n">generate_hpc_scripts</span><span class="p">,</span>
            <span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Method </span><span class="si">{</span><span class="n">method</span><span class="si">}</span><span class="s2"> is not supported.&quot;</span><span class="p">)</span></div>



<div class="viewcode-block" id="rank_key">
<a class="viewcode-back" href="../../../modules/multicom_ligand.ensemble_generation.html#multicom_ligand.models.ensemble_generation.rank_key">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">rank_key</span><span class="p">(</span><span class="n">file_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Define a custom key for ranking the predictions.</span>

<span class="sd">    :param file_path: Path to the file to rank.</span>
<span class="sd">    :return: The rank key for the file.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">filename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span>
    <span class="n">match</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;rank(\d+)&quot;</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">match</span><span class="p">:</span>
        <span class="k">return</span> <span class="nb">int</span><span class="p">(</span><span class="n">match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">))</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="nb">float</span><span class="p">(</span><span class="s2">&quot;inf&quot;</span><span class="p">)</span>  <span class="c1"># NOTE: if no match found, put it at the end</span></div>



<div class="viewcode-block" id="get_method_predictions">
<a class="viewcode-back" href="../../../modules/multicom_ligand.ensemble_generation.html#multicom_ligand.models.ensemble_generation.get_method_predictions">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">get_method_predictions</span><span class="p">(</span>
    <span class="n">method</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">target</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">cfg</span><span class="p">:</span> <span class="n">DictConfig</span><span class="p">,</span>
    <span class="n">binding_site_method</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">input_protein_filepath</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Get the predictions generated by the method.</span>

<span class="sd">    :param method: Name of the method to get predictions for.</span>
<span class="sd">    :param target: Name of the target protein-ligand pair.</span>
<span class="sd">    :param cfg: Configuration dictionary for runtime arguments.</span>
<span class="sd">    :param binding_site_method: Optional name of the method used to</span>
<span class="sd">        predict AutoDock Vina&#39;s binding sites.</span>
<span class="sd">    :param input_protein_filepath: Optional path to the input protein</span>
<span class="sd">        structure PDB file.</span>
<span class="sd">    :return: List of method predictions, each as a tuple of the output</span>
<span class="sd">        protein filepath and the output ligand filepath.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;diffdock&quot;</span><span class="p">:</span>
        <span class="n">ensemble_benchmarking_output_dir</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">Path</span><span class="p">(</span><span class="n">cfg</span><span class="o">.</span><span class="n">diffdock_output_dir</span><span class="p">)</span><span class="o">.</span><span class="n">parent</span>
            <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;diffdock_</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_benchmarking_dataset</span><span class="si">}</span><span class="s2">_output_</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_benchmarking_repeat_index</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="k">if</span> <span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_benchmarking</span>
            <span class="k">else</span> <span class="n">cfg</span><span class="o">.</span><span class="n">diffdock_output_dir</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_benchmarking</span><span class="p">:</span>
            <span class="n">protein_output_files</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span>
                <span class="n">Path</span><span class="p">(</span><span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_benchmarking_apo_protein_dir</span><span class="p">)</span><span class="o">.</span><span class="n">rglob</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;*</span><span class="si">{</span><span class="n">target</span><span class="si">}</span><span class="s2">*.pdb&quot;</span><span class="p">)</span>
            <span class="p">)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">protein_output_files</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;No apo protein structure found for target </span><span class="si">{</span><span class="n">target</span><span class="si">}</span><span class="s2"> in directory </span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_benchmarking_apo_protein_dir</span><span class="si">}</span><span class="s2">. Skipping this target...&quot;</span>
                <span class="p">)</span>
                <span class="k">return</span> <span class="p">[]</span>
            <span class="k">assert</span> <span class="p">(</span>
                <span class="nb">len</span><span class="p">(</span><span class="n">protein_output_files</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span>
            <span class="p">),</span> <span class="s2">&quot;The DiffDock ensemble benchmarking dataset must contain one apo protein input structure per target.&quot;</span>
            <span class="n">protein_output_files</span> <span class="o">=</span> <span class="p">[</span>
                <span class="nb">str</span><span class="p">(</span><span class="n">protein_output_files</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">cfg</span><span class="o">.</span><span class="n">method_top_n_to_select</span><span class="p">)</span>
            <span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">cfg</span><span class="o">.</span><span class="n">diffdock_input_csv_path</span><span class="p">):</span>
                <span class="n">diffdock_input_csv</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">cfg</span><span class="o">.</span><span class="n">diffdock_input_csv_path</span><span class="p">)</span>
                <span class="n">protein_output_files</span> <span class="o">=</span> <span class="p">[</span>
                    <span class="n">diffdock_input_csv</span><span class="p">[</span>
                        <span class="n">diffdock_input_csv</span><span class="o">.</span><span class="n">complex_name</span> <span class="o">==</span> <span class="n">target</span>
                    <span class="p">]</span><span class="o">.</span><span class="n">protein_path</span><span class="o">.</span><span class="n">item</span><span class="p">()</span>
                    <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">cfg</span><span class="o">.</span><span class="n">method_top_n_to_select</span><span class="p">)</span>
                <span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">assert</span> <span class="p">(</span>
                    <span class="n">input_protein_filepath</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
                <span class="p">),</span> <span class="s2">&quot;Input protein filepath must be provided if `diffdock_input_csv_path` does not exist.&quot;</span>
                <span class="n">protein_output_files</span> <span class="o">=</span> <span class="p">[</span>
                    <span class="n">input_protein_filepath</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">cfg</span><span class="o">.</span><span class="n">method_top_n_to_select</span><span class="p">)</span>
                <span class="p">]</span>
        <span class="n">diffdock_target</span> <span class="o">=</span> <span class="p">(</span>
            <span class="s2">&quot;_&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">target</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[:</span><span class="mi">3</span><span class="p">])</span>
            <span class="k">if</span> <span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_benchmarking</span> <span class="ow">and</span> <span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_benchmarking_dataset</span> <span class="o">==</span> <span class="s2">&quot;dockgen&quot;</span>
            <span class="k">else</span> <span class="n">target</span>
        <span class="p">)</span>
        <span class="n">ligand_output_files</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
            <span class="p">[</span>
                <span class="n">file</span>
                <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="nb">map</span><span class="p">(</span>
                    <span class="nb">str</span><span class="p">,</span>
                    <span class="n">Path</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ensemble_benchmarking_output_dir</span><span class="p">,</span> <span class="n">diffdock_target</span><span class="p">))</span><span class="o">.</span><span class="n">rglob</span><span class="p">(</span>
                        <span class="s2">&quot;*.sdf&quot;</span>
                    <span class="p">),</span>
                <span class="p">)</span>
                <span class="k">if</span> <span class="s2">&quot;rank&quot;</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
                <span class="ow">and</span> <span class="s2">&quot;confidence&quot;</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
                <span class="ow">and</span> <span class="s2">&quot;relaxed&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
            <span class="p">],</span>
            <span class="n">key</span><span class="o">=</span><span class="n">rank_key</span><span class="p">,</span>
        <span class="p">)[:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">method_top_n_to_select</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">cfg</span><span class="o">.</span><span class="n">diffdock_validate_target_names</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">protein_output_file</span><span class="p">,</span> <span class="n">ligand_output_file</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span>
                <span class="n">protein_output_files</span><span class="p">,</span> <span class="n">ligand_output_files</span>
            <span class="p">):</span>
                <span class="k">assert</span> <span class="p">(</span>
                    <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">splitext</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">protein_output_file</span><span class="p">))[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_holo&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
                    <span class="o">==</span> <span class="n">target</span>
                <span class="p">),</span> <span class="s2">&quot;Protein files must be for the designated target.&quot;</span>
                <span class="k">assert</span> <span class="p">(</span>
                    <span class="n">Path</span><span class="p">(</span><span class="n">ligand_output_file</span><span class="p">)</span><span class="o">.</span><span class="n">parent</span><span class="o">.</span><span class="n">stem</span> <span class="o">==</span> <span class="n">diffdock_target</span>
                <span class="p">),</span> <span class="s2">&quot;Ligand files must be for the designated target.&quot;</span>
    <span class="k">elif</span> <span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;dynamicbind&quot;</span><span class="p">:</span>
        <span class="n">target_dir_name</span> <span class="o">=</span> <span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_benchmarking_dataset</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">target</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_benchmarking_repeat_index</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="k">if</span> <span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_benchmarking</span>
            <span class="k">else</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">dynamicbind_header</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">target</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="p">)</span>
        <span class="n">protein_output_files</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span>
            <span class="nb">map</span><span class="p">(</span>
                <span class="nb">str</span><span class="p">,</span>
                <span class="n">Path</span><span class="p">(</span><span class="n">cfg</span><span class="o">.</span><span class="n">dynamicbind_input_ligand_csv_dir</span><span class="p">)</span><span class="o">.</span><span class="n">parent</span><span class="o">.</span><span class="n">rglob</span><span class="p">(</span>
                    <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                        <span class="s2">&quot;outputs&quot;</span><span class="p">,</span>
                        <span class="s2">&quot;results&quot;</span><span class="p">,</span>
                        <span class="n">target_dir_name</span><span class="p">,</span>
                        <span class="s2">&quot;index0_idx_0&quot;</span><span class="p">,</span>
                        <span class="s2">&quot;rank*_receptor_*.pdb&quot;</span><span class="p">,</span>
                    <span class="p">)</span>
                <span class="p">),</span>
            <span class="p">)</span>
        <span class="p">)</span>
        <span class="n">ligand_output_files</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span>
            <span class="nb">map</span><span class="p">(</span>
                <span class="nb">str</span><span class="p">,</span>
                <span class="n">Path</span><span class="p">(</span><span class="n">cfg</span><span class="o">.</span><span class="n">dynamicbind_input_ligand_csv_dir</span><span class="p">)</span><span class="o">.</span><span class="n">parent</span><span class="o">.</span><span class="n">rglob</span><span class="p">(</span>
                    <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                        <span class="s2">&quot;outputs&quot;</span><span class="p">,</span>
                        <span class="s2">&quot;results&quot;</span><span class="p">,</span>
                        <span class="n">target_dir_name</span><span class="p">,</span>
                        <span class="s2">&quot;index0_idx_0&quot;</span><span class="p">,</span>
                        <span class="s2">&quot;rank*_ligand_*.sdf&quot;</span><span class="p">,</span>
                    <span class="p">)</span>
                <span class="p">),</span>
            <span class="p">)</span>
        <span class="p">)</span>
        <span class="n">protein_output_files</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
            <span class="p">[</span><span class="n">file</span> <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">protein_output_files</span> <span class="k">if</span> <span class="s2">&quot;relaxed&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">file</span><span class="p">)],</span>
            <span class="n">key</span><span class="o">=</span><span class="n">rank_key</span><span class="p">,</span>
        <span class="p">)[:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">method_top_n_to_select</span><span class="p">]</span>
        <span class="n">ligand_output_files</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
            <span class="p">[</span><span class="n">file</span> <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">ligand_output_files</span> <span class="k">if</span> <span class="s2">&quot;relaxed&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">file</span><span class="p">)],</span>
            <span class="n">key</span><span class="o">=</span><span class="n">rank_key</span><span class="p">,</span>
        <span class="p">)[:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">method_top_n_to_select</span><span class="p">]</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">protein_output_files</span><span class="p">)</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">ligand_output_files</span><span class="p">):</span>
            <span class="n">ligand_output_files</span> <span class="o">=</span> <span class="p">[</span>
                <span class="n">file</span>
                <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">ligand_output_files</span>
                <span class="k">if</span> <span class="n">Path</span><span class="p">(</span><span class="n">file</span><span class="p">)</span><span class="o">.</span><span class="n">stem</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;_ligand_&quot;</span><span class="p">,</span> <span class="s2">&quot;_receptor_&quot;</span><span class="p">)</span>
                <span class="ow">in</span> <span class="p">{</span><span class="n">Path</span><span class="p">(</span><span class="n">protein_file</span><span class="p">)</span><span class="o">.</span><span class="n">stem</span> <span class="k">for</span> <span class="n">protein_file</span> <span class="ow">in</span> <span class="n">protein_output_files</span><span class="p">}</span>
            <span class="p">]</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">ligand_output_files</span><span class="p">)</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">protein_output_files</span><span class="p">):</span>
            <span class="n">protein_output_files</span> <span class="o">=</span> <span class="p">[</span>
                <span class="n">file</span>
                <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">protein_output_files</span>
                <span class="k">if</span> <span class="n">Path</span><span class="p">(</span><span class="n">file</span><span class="p">)</span><span class="o">.</span><span class="n">stem</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;_receptor_&quot;</span><span class="p">,</span> <span class="s2">&quot;_ligand_&quot;</span><span class="p">)</span>
                <span class="ow">in</span> <span class="p">{</span><span class="n">Path</span><span class="p">(</span><span class="n">ligand_file</span><span class="p">)</span><span class="o">.</span><span class="n">stem</span> <span class="k">for</span> <span class="n">ligand_file</span> <span class="ow">in</span> <span class="n">ligand_output_files</span><span class="p">}</span>
            <span class="p">]</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">protein_output_files</span><span class="p">)</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">ligand_output_files</span><span class="p">):</span>
            <span class="n">ligand_output_files</span> <span class="o">=</span> <span class="p">[</span>
                <span class="n">file</span>
                <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">ligand_output_files</span>
                <span class="k">if</span> <span class="n">Path</span><span class="p">(</span><span class="n">file</span><span class="p">)</span><span class="o">.</span><span class="n">stem</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;_ligand_&quot;</span><span class="p">,</span> <span class="s2">&quot;_receptor_&quot;</span><span class="p">)</span>
                <span class="ow">in</span> <span class="p">{</span><span class="n">Path</span><span class="p">(</span><span class="n">protein_file</span><span class="p">)</span><span class="o">.</span><span class="n">stem</span> <span class="k">for</span> <span class="n">protein_file</span> <span class="ow">in</span> <span class="n">protein_output_files</span><span class="p">}</span>
            <span class="p">]</span>
        <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">protein_output_files</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span>
            <span class="n">ligand_output_files</span>
        <span class="p">),</span> <span class="s2">&quot;The number of DynamicBind protein and ligand files must match.&quot;</span>
    <span class="k">elif</span> <span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;neuralplexer&quot;</span><span class="p">:</span>
        <span class="n">ensemble_benchmarking_output_dir</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">Path</span><span class="p">(</span><span class="n">cfg</span><span class="o">.</span><span class="n">input_dir</span> <span class="k">if</span> <span class="n">cfg</span><span class="o">.</span><span class="n">input_dir</span> <span class="k">else</span> <span class="n">cfg</span><span class="o">.</span><span class="n">neuralplexer_out_path</span><span class="p">)</span><span class="o">.</span><span class="n">parent</span>
            <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;neuralplexer_</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_benchmarking_dataset</span><span class="si">}</span><span class="s2">_outputs_</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_benchmarking_repeat_index</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="k">if</span> <span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_benchmarking</span>
            <span class="k">else</span> <span class="p">(</span><span class="n">cfg</span><span class="o">.</span><span class="n">input_dir</span> <span class="k">if</span> <span class="n">cfg</span><span class="o">.</span><span class="n">input_dir</span> <span class="k">else</span> <span class="n">cfg</span><span class="o">.</span><span class="n">neuralplexer_out_path</span><span class="p">)</span>
        <span class="p">)</span>
        <span class="n">protein_output_files</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
            <span class="p">[</span>
                <span class="n">file</span>
                <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="nb">map</span><span class="p">(</span>
                    <span class="nb">str</span><span class="p">,</span>
                    <span class="n">Path</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ensemble_benchmarking_output_dir</span><span class="p">,</span> <span class="n">target</span><span class="p">))</span><span class="o">.</span><span class="n">rglob</span><span class="p">(</span><span class="s2">&quot;*.pdb&quot;</span><span class="p">),</span>
                <span class="p">)</span>
                <span class="k">if</span> <span class="s2">&quot;rank&quot;</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
                <span class="ow">and</span> <span class="s2">&quot;relaxed&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
                <span class="ow">and</span> <span class="s2">&quot;aligned&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
            <span class="p">],</span>
            <span class="n">key</span><span class="o">=</span><span class="n">rank_key</span><span class="p">,</span>
        <span class="p">)[:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">method_top_n_to_select</span><span class="p">]</span>
        <span class="n">ligand_output_files</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
            <span class="p">[</span>
                <span class="n">file</span>
                <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="nb">map</span><span class="p">(</span>
                    <span class="nb">str</span><span class="p">,</span>
                    <span class="n">Path</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ensemble_benchmarking_output_dir</span><span class="p">,</span> <span class="n">target</span><span class="p">))</span><span class="o">.</span><span class="n">rglob</span><span class="p">(</span><span class="s2">&quot;*.sdf&quot;</span><span class="p">),</span>
                <span class="p">)</span>
                <span class="k">if</span> <span class="s2">&quot;rank&quot;</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
                <span class="ow">and</span> <span class="s2">&quot;relaxed&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
                <span class="ow">and</span> <span class="s2">&quot;aligned&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
            <span class="p">],</span>
            <span class="n">key</span><span class="o">=</span><span class="n">rank_key</span><span class="p">,</span>
        <span class="p">)[:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">method_top_n_to_select</span><span class="p">]</span>
    <span class="k">elif</span> <span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;flowdock&quot;</span><span class="p">:</span>
        <span class="n">ensemble_benchmarking_output_dir</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">Path</span><span class="p">(</span><span class="n">cfg</span><span class="o">.</span><span class="n">input_dir</span> <span class="k">if</span> <span class="n">cfg</span><span class="o">.</span><span class="n">input_dir</span> <span class="k">else</span> <span class="n">cfg</span><span class="o">.</span><span class="n">flowdock_out_path</span><span class="p">)</span><span class="o">.</span><span class="n">parent</span>
            <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;flowdock_</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_benchmarking_dataset</span><span class="si">}</span><span class="s2">_outputs_</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_benchmarking_repeat_index</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="k">if</span> <span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_benchmarking</span>
            <span class="k">else</span> <span class="p">(</span><span class="n">cfg</span><span class="o">.</span><span class="n">input_dir</span> <span class="k">if</span> <span class="n">cfg</span><span class="o">.</span><span class="n">input_dir</span> <span class="k">else</span> <span class="n">cfg</span><span class="o">.</span><span class="n">flowdock_out_path</span><span class="p">)</span>
        <span class="p">)</span>
        <span class="n">protein_output_files</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
            <span class="p">[</span>
                <span class="n">file</span>
                <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="nb">map</span><span class="p">(</span>
                    <span class="nb">str</span><span class="p">,</span>
                    <span class="n">Path</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ensemble_benchmarking_output_dir</span><span class="p">,</span> <span class="n">target</span><span class="p">))</span><span class="o">.</span><span class="n">rglob</span><span class="p">(</span><span class="s2">&quot;*.pdb&quot;</span><span class="p">),</span>
                <span class="p">)</span>
                <span class="k">if</span> <span class="s2">&quot;rank&quot;</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
                <span class="ow">and</span> <span class="s2">&quot;relaxed&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
                <span class="ow">and</span> <span class="s2">&quot;aligned&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
            <span class="p">],</span>
            <span class="n">key</span><span class="o">=</span><span class="n">rank_key</span><span class="p">,</span>
        <span class="p">)[:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">method_top_n_to_select</span><span class="p">]</span>
        <span class="n">ligand_output_files</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
            <span class="p">[</span>
                <span class="n">file</span>
                <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="nb">map</span><span class="p">(</span>
                    <span class="nb">str</span><span class="p">,</span>
                    <span class="n">Path</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ensemble_benchmarking_output_dir</span><span class="p">,</span> <span class="n">target</span><span class="p">))</span><span class="o">.</span><span class="n">rglob</span><span class="p">(</span><span class="s2">&quot;*.sdf&quot;</span><span class="p">),</span>
                <span class="p">)</span>
                <span class="k">if</span> <span class="s2">&quot;rank&quot;</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
                <span class="ow">and</span> <span class="s2">&quot;relaxed&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
                <span class="ow">and</span> <span class="s2">&quot;aligned&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
            <span class="p">],</span>
            <span class="n">key</span><span class="o">=</span><span class="n">rank_key</span><span class="p">,</span>
        <span class="p">)[:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">method_top_n_to_select</span><span class="p">]</span>
    <span class="k">elif</span> <span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;rfaa&quot;</span><span class="p">:</span>
        <span class="n">ensemble_benchmarking_output_dir</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">Path</span><span class="p">(</span><span class="n">cfg</span><span class="o">.</span><span class="n">rfaa_output_dir</span><span class="p">)</span><span class="o">.</span><span class="n">parent</span>
            <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;rfaa_</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_benchmarking_dataset</span><span class="si">}</span><span class="s2">_outputs_</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_benchmarking_repeat_index</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="k">if</span> <span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_benchmarking</span>
            <span class="k">else</span> <span class="n">cfg</span><span class="o">.</span><span class="n">rfaa_output_dir</span>
        <span class="p">)</span>
        <span class="n">protein_output_files</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
            <span class="p">[</span>
                <span class="n">file</span>
                <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="nb">map</span><span class="p">(</span>
                    <span class="nb">str</span><span class="p">,</span>
                    <span class="n">Path</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ensemble_benchmarking_output_dir</span><span class="p">,</span> <span class="n">target</span><span class="p">))</span><span class="o">.</span><span class="n">rglob</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">target</span><span class="si">}</span><span class="s2">_protein.pdb&quot;</span>
                    <span class="p">),</span>
                <span class="p">)</span>
            <span class="p">],</span>
            <span class="n">key</span><span class="o">=</span><span class="n">rank_key</span><span class="p">,</span>
        <span class="p">)[:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">method_top_n_to_select</span><span class="p">]</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">protein_output_files</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">protein_output_files</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
                <span class="p">[</span>
                    <span class="n">file</span>
                    <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="nb">map</span><span class="p">(</span>
                        <span class="nb">str</span><span class="p">,</span>
                        <span class="n">Path</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ensemble_benchmarking_output_dir</span><span class="p">,</span> <span class="n">target</span><span class="p">))</span><span class="o">.</span><span class="n">rglob</span><span class="p">(</span>
                            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">target</span><span class="si">}</span><span class="s2">_nucleic.pdb&quot;</span>
                        <span class="p">),</span>
                    <span class="p">)</span>
                <span class="p">],</span>
                <span class="n">key</span><span class="o">=</span><span class="n">rank_key</span><span class="p">,</span>
            <span class="p">)[:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">method_top_n_to_select</span><span class="p">]</span>
        <span class="n">ligand_output_files</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
            <span class="p">[</span>
                <span class="n">file</span>
                <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="nb">map</span><span class="p">(</span>
                    <span class="nb">str</span><span class="p">,</span>
                    <span class="n">Path</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ensemble_benchmarking_output_dir</span><span class="p">,</span> <span class="n">target</span><span class="p">))</span><span class="o">.</span><span class="n">rglob</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">target</span><span class="si">}</span><span class="s2">_ligand.sdf&quot;</span>
                    <span class="p">),</span>
                <span class="p">)</span>
            <span class="p">],</span>
            <span class="n">key</span><span class="o">=</span><span class="n">rank_key</span><span class="p">,</span>
        <span class="p">)[:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">method_top_n_to_select</span><span class="p">]</span>
    <span class="k">elif</span> <span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;vina&quot;</span><span class="p">:</span>
        <span class="k">assert</span> <span class="n">binding_site_method</span><span class="p">,</span> <span class="s2">&quot;Binding site method must be provided for Vina predictions.&quot;</span>
        <span class="n">ensemble_benchmarking_output_dir</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">Path</span><span class="p">(</span><span class="n">cfg</span><span class="o">.</span><span class="n">vina_output_dir</span><span class="p">)</span><span class="o">.</span><span class="n">parent</span>
            <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;vina_</span><span class="si">{</span><span class="n">binding_site_method</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_benchmarking_dataset</span><span class="si">}</span><span class="s2">_outputs_</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_benchmarking_repeat_index</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="k">if</span> <span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_benchmarking</span>
            <span class="k">else</span> <span class="n">cfg</span><span class="o">.</span><span class="n">vina_output_dir</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;vina_&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;vina_</span><span class="si">{</span><span class="n">binding_site_method</span><span class="si">}</span><span class="s2">_&quot;</span><span class="p">)</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_benchmarking</span><span class="p">:</span>
            <span class="n">protein_output_files</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span>
                <span class="n">Path</span><span class="p">(</span><span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_benchmarking_apo_protein_dir</span><span class="p">)</span><span class="o">.</span><span class="n">rglob</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;*</span><span class="si">{</span><span class="n">target</span><span class="si">}</span><span class="s2">*.pdb&quot;</span><span class="p">)</span>
            <span class="p">)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">protein_output_files</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;No apo protein structure found for target </span><span class="si">{</span><span class="n">target</span><span class="si">}</span><span class="s2"> in directory </span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_benchmarking_apo_protein_dir</span><span class="si">}</span><span class="s2">. Skipping this target...&quot;</span>
                <span class="p">)</span>
                <span class="k">return</span> <span class="p">[]</span>
            <span class="k">assert</span> <span class="p">(</span>
                <span class="nb">len</span><span class="p">(</span><span class="n">protein_output_files</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span>
            <span class="p">),</span> <span class="s2">&quot;The DiffDock ensemble benchmarking dataset must contain one apo protein input structure per target.&quot;</span>
            <span class="n">protein_output_files</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">protein_output_files</span><span class="p">[</span><span class="mi">0</span><span class="p">])]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">assert</span> <span class="p">(</span>
                <span class="n">input_protein_filepath</span>
            <span class="p">),</span> <span class="s2">&quot;Input protein filepath must be provided for Vina predictions when not ensemble-benchmarking.&quot;</span>
            <span class="n">protein_output_files</span> <span class="o">=</span> <span class="p">[</span><span class="n">input_protein_filepath</span><span class="p">]</span>
        <span class="c1"># NOTE: Vina saves only the top-ranked ligand prediction per target</span>
        <span class="n">vina_target</span> <span class="o">=</span> <span class="p">(</span>
            <span class="s2">&quot;_&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">target</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[:</span><span class="mi">3</span><span class="p">])</span>
            <span class="k">if</span> <span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_benchmarking</span> <span class="ow">and</span> <span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_benchmarking_dataset</span> <span class="o">==</span> <span class="s2">&quot;dockgen&quot;</span>
            <span class="k">else</span> <span class="n">target</span>
        <span class="p">)</span>
        <span class="n">ligand_output_files</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
            <span class="p">[</span>
                <span class="n">file</span>
                <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="nb">map</span><span class="p">(</span>
                    <span class="nb">str</span><span class="p">,</span>
                    <span class="n">Path</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ensemble_benchmarking_output_dir</span><span class="p">,</span> <span class="n">vina_target</span><span class="p">))</span><span class="o">.</span><span class="n">rglob</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">vina_target</span><span class="si">}</span><span class="s2">*.sdf&quot;</span>
                    <span class="p">),</span>
                <span class="p">)</span>
                <span class="k">if</span> <span class="s2">&quot;relaxed&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
                <span class="ow">and</span> <span class="s2">&quot;group&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
            <span class="p">],</span>
            <span class="n">key</span><span class="o">=</span><span class="n">rank_key</span><span class="p">,</span>
        <span class="p">)[:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">method_top_n_to_select</span><span class="p">]</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">len</span><span class="p">(</span><span class="n">ligand_output_files</span><span class="p">):</span>
            <span class="c1"># NOTE: when predicting e.g., with DiffDock-Vina, if no ligand predictions are found, skip the protein prediction as well</span>
            <span class="n">protein_output_files</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">protein_output_files</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span>
            <span class="n">ligand_output_files</span>
        <span class="p">),</span> <span class="s2">&quot;Number of Vina protein and ligand files must match.&quot;</span>
    <span class="k">elif</span> <span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;tulip&quot;</span><span class="p">:</span>
        <span class="k">assert</span> <span class="n">input_protein_filepath</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span>
            <span class="n">input_protein_filepath</span>
        <span class="p">),</span> <span class="s2">&quot;A valid input protein filepath must be provided to analyze TULIP&#39;s results.&quot;</span>
        <span class="n">ensemble_benchmarking_output_dir</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">Path</span><span class="p">(</span><span class="n">cfg</span><span class="o">.</span><span class="n">tulip_output_dir</span><span class="p">)</span><span class="o">.</span><span class="n">parent</span>
            <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;tulip_</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_benchmarking_dataset</span><span class="si">}</span><span class="s2">_outputs_</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_benchmarking_repeat_index</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="k">if</span> <span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_benchmarking</span>
            <span class="k">else</span> <span class="n">cfg</span><span class="o">.</span><span class="n">tulip_output_dir</span>
        <span class="p">)</span>
        <span class="n">ligand_output_files</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
            <span class="p">[</span>
                <span class="n">file</span>
                <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="nb">map</span><span class="p">(</span>
                    <span class="nb">str</span><span class="p">,</span>
                    <span class="n">Path</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ensemble_benchmarking_output_dir</span><span class="p">,</span> <span class="n">target</span><span class="p">))</span><span class="o">.</span><span class="n">rglob</span><span class="p">(</span>
                        <span class="s2">&quot;rank*.sdf&quot;</span>
                    <span class="p">),</span>
                <span class="p">)</span>
                <span class="k">if</span> <span class="s2">&quot;relaxed&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
            <span class="p">],</span>
            <span class="n">key</span><span class="o">=</span><span class="n">rank_key</span><span class="p">,</span>
        <span class="p">)[:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">method_top_n_to_select</span><span class="p">]</span>
        <span class="n">protein_output_files</span> <span class="o">=</span> <span class="p">[</span><span class="n">input_protein_filepath</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">ligand_output_files</span><span class="p">))]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Method </span><span class="si">{</span><span class="n">method</span><span class="si">}</span><span class="s2"> is not supported.&quot;</span><span class="p">)</span>

    <span class="n">protein_output_files</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">protein_output_files</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="n">rank_key</span><span class="p">)</span>
    <span class="n">ligand_output_files</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">ligand_output_files</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="n">rank_key</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_method_ligands_pre_ranking</span><span class="p">:</span>
        <span class="c1"># relax ligand structures prior to ranking as desired;</span>
        <span class="c1"># NOTE: relaxation is practically necessary when making</span>
        <span class="c1"># multi-ligand predictions with DiffDock and DynamicBind,</span>
        <span class="c1"># and is optional (yet highly recommended) for all other predictions</span>
        <span class="n">pool</span> <span class="o">=</span> <span class="n">multiprocessing</span><span class="o">.</span><span class="n">Pool</span><span class="p">(</span><span class="n">processes</span><span class="o">=</span><span class="n">cfg</span><span class="o">.</span><span class="n">relax_num_processes</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">protein_output_file</span><span class="p">,</span> <span class="n">ligand_output_file</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span>
            <span class="n">protein_output_files</span><span class="p">,</span> <span class="n">ligand_output_files</span>
        <span class="p">):</span>
            <span class="k">if</span> <span class="p">(</span>
                <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">ligand_output_file</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;.sdf&quot;</span><span class="p">,</span> <span class="s2">&quot;_ensemble_relaxed.sdf&quot;</span><span class="p">))</span>
                <span class="ow">or</span> <span class="ow">not</span> <span class="n">cfg</span><span class="o">.</span><span class="n">skip_existing</span>
            <span class="p">):</span>
                <span class="k">with</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">TemporaryDirectory</span><span class="p">(</span>
                    <span class="n">suffix</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;_</span><span class="si">{</span><span class="n">method</span><span class="si">}</span><span class="s2">_ensemble_cache_dir&quot;</span>
                <span class="p">)</span> <span class="k">as</span> <span class="n">temp_directory</span><span class="p">:</span>
                    <span class="n">pool</span><span class="o">.</span><span class="n">apply_async</span><span class="p">(</span>
                        <span class="n">relax_single_filepair</span><span class="p">,</span>
                        <span class="n">args</span><span class="o">=</span><span class="p">(</span>
                            <span class="n">Path</span><span class="p">(</span><span class="n">protein_output_file</span><span class="p">),</span>
                            <span class="n">Path</span><span class="p">(</span><span class="n">ligand_output_file</span><span class="p">),</span>
                            <span class="n">Path</span><span class="p">(</span><span class="n">ligand_output_file</span><span class="p">)</span><span class="o">.</span><span class="n">parent</span><span class="p">,</span>
                            <span class="n">temp_directory</span><span class="p">,</span>
                            <span class="n">OmegaConf</span><span class="o">.</span><span class="n">create</span><span class="p">(</span>
                                <span class="p">{</span>
                                    <span class="s2">&quot;method&quot;</span><span class="p">:</span> <span class="s2">&quot;ensemble&quot;</span><span class="p">,</span>
                                    <span class="s2">&quot;add_solvent&quot;</span><span class="p">:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_add_solvent</span><span class="p">,</span>
                                    <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="n">target</span><span class="p">,</span>
                                    <span class="s2">&quot;prep_only&quot;</span><span class="p">:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_prep_only</span><span class="p">,</span>
                                    <span class="s2">&quot;platform&quot;</span><span class="p">:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_platform</span><span class="p">,</span>
                                    <span class="s2">&quot;cuda_device_index&quot;</span><span class="p">:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">cuda_device_index</span><span class="p">,</span>
                                    <span class="s2">&quot;log_level&quot;</span><span class="p">:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_log_level</span><span class="p">,</span>
                                    <span class="s2">&quot;relax_protein&quot;</span><span class="p">:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_protein</span><span class="p">,</span>
                                    <span class="s2">&quot;remove_initial_protein_hydrogens&quot;</span><span class="p">:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_remove_initial_protein_hydrogens</span>
                                    <span class="ow">or</span> <span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;rfaa&quot;</span><span class="p">,</span>
                                    <span class="s2">&quot;assign_each_ligand_unique_force&quot;</span><span class="p">:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_assign_each_ligand_unique_force</span><span class="p">,</span>
                                    <span class="s2">&quot;model_ions&quot;</span><span class="p">:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_model_ions</span><span class="p">,</span>
                                    <span class="s2">&quot;cache_files&quot;</span><span class="p">:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_cache_files</span><span class="p">,</span>
                                    <span class="s2">&quot;assign_partial_charges_manually&quot;</span><span class="p">:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_assign_partial_charges_manually</span><span class="p">,</span>
                                    <span class="s2">&quot;report_initial_energy_only&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                    <span class="s2">&quot;max_final_e_value&quot;</span><span class="p">:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_max_final_e_value</span><span class="p">,</span>
                                    <span class="s2">&quot;max_num_attempts&quot;</span><span class="p">:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_max_num_attempts</span><span class="p">,</span>
                                    <span class="s2">&quot;skip_existing&quot;</span><span class="p">:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_skip_existing</span><span class="p">,</span>
                                <span class="p">}</span>
                            <span class="p">),</span>
                        <span class="p">),</span>
                    <span class="p">)</span>
        <span class="n">pool</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
        <span class="n">pool</span><span class="o">.</span><span class="n">join</span><span class="p">()</span>
        <span class="n">relaxed_ligand_output_files</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">file</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;.sdf&quot;</span><span class="p">,</span> <span class="s2">&quot;_ensemble_relaxed.sdf&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">ligand_output_files</span>
        <span class="p">]</span>
        <span class="k">assert</span> <span class="nb">all</span><span class="p">(</span>
            <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">file</span><span class="p">)</span> <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">relaxed_ligand_output_files</span>
        <span class="p">),</span> <span class="sa">f</span><span class="s2">&quot;Ligand relaxation failed for file(s): </span><span class="si">{</span><span class="p">[</span><span class="n">file</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="n">file</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">relaxed_ligand_output_files</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="ow">not</span><span class="w"> </span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">file</span><span class="p">)]</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="k">if</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_protein</span><span class="p">:</span>
            <span class="n">relaxed_protein_output_files</span> <span class="o">=</span> <span class="p">[</span>
                <span class="n">file</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;.pdb&quot;</span><span class="p">,</span> <span class="s2">&quot;_protein_ensemble_relaxed.pdb&quot;</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">protein_output_files</span>
            <span class="p">]</span>
            <span class="k">assert</span> <span class="nb">all</span><span class="p">(</span>
                <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">file</span><span class="p">)</span> <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">relaxed_protein_output_files</span>
            <span class="p">),</span> <span class="sa">f</span><span class="s2">&quot;Protein relaxation failed for file(s): </span><span class="si">{</span><span class="p">[</span><span class="n">file</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="n">file</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">relaxed_protein_output_files</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="ow">not</span><span class="w"> </span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">file</span><span class="p">)]</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="n">protein_output_files</span> <span class="o">=</span> <span class="n">relaxed_protein_output_files</span>
        <span class="n">ligand_output_files</span> <span class="o">=</span> <span class="n">relaxed_ligand_output_files</span>
    <span class="k">return</span> <span class="nb">list</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">protein_output_files</span><span class="p">,</span> <span class="n">ligand_output_files</span><span class="p">))</span></div>



<div class="viewcode-block" id="generate_ensemble_predictions">
<a class="viewcode-back" href="../../../modules/multicom_ligand.ensemble_generation.html#multicom_ligand.models.ensemble_generation.generate_ensemble_predictions">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">generate_ensemble_predictions</span><span class="p">(</span>
    <span class="n">protein_filepath</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">ligand_smiles</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">input_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">cfg</span><span class="p">:</span> <span class="n">DictConfig</span><span class="p">,</span>
    <span class="n">generate_hpc_scripts</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">method_filepaths_mapping</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">auxiliary_estimation_input_dir</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">Optional</span><span class="p">[</span><span class="n">ENSEMBLE_PREDICTIONS</span><span class="p">],</span> <span class="nb">bool</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Generate bound complex predictions using an ensemble of methods.</span>

<span class="sd">    :param protein_filepath: Path to the input protein structure PDB</span>
<span class="sd">        file.</span>
<span class="sd">    :param ligand_input: Path to the input ligand SMILES string.</span>
<span class="sd">    :param input_id: Input ID.</span>
<span class="sd">    :param target: Name of the target protein-ligand pair.</span>
<span class="sd">    :param cfg: Configuration dictionary for runtime arguments.</span>
<span class="sd">    :param generate_hpc_scripts: Whether to generate HPC scripts for the</span>
<span class="sd">        ensemble predictions.</span>
<span class="sd">    :param method_filepaths_mapping: Optional mapping of method names to</span>
<span class="sd">        a list of tuples of protein and ligand filepaths.</span>
<span class="sd">    :param auxiliary_estimation_input_dir: Optional path to the</span>
<span class="sd">        directory containing auxiliary estimation input files e.g., for</span>
<span class="sd">        FlowDock.</span>
<span class="sd">    :return: Dictionary of method names and their corresponding</span>
<span class="sd">        predictions as well as whether the prediction scripts were</span>
<span class="sd">        generated and now need to be run.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">cfg</span><span class="o">.</span><span class="n">output_bash_file_dir</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="n">generating_script</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="k">for</span> <span class="n">method</span> <span class="ow">in</span> <span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_methods</span><span class="p">:</span>
        <span class="n">output_filepath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
            <span class="n">cfg</span><span class="o">.</span><span class="n">output_bash_file_dir</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">method</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">input_id</span><span class="si">}</span><span class="s2">_inference.sh&quot;</span>
        <span class="p">)</span>
        <span class="n">generating_script</span> <span class="o">=</span> <span class="ow">not</span> <span class="n">cfg</span><span class="o">.</span><span class="n">resume</span> <span class="ow">or</span> <span class="p">(</span>
            <span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;vina&quot;</span> <span class="ow">and</span> <span class="n">cfg</span><span class="o">.</span><span class="n">generate_vina_scripts</span> <span class="ow">and</span> <span class="n">cfg</span><span class="o">.</span><span class="n">resume</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="n">generating_script</span><span class="p">:</span>
            <span class="n">generate_method_prediction_script</span><span class="p">(</span>
                <span class="n">method</span><span class="p">,</span>
                <span class="n">protein_filepath</span><span class="p">,</span>
                <span class="n">ligand_smiles</span><span class="p">,</span>
                <span class="n">input_id</span><span class="p">,</span>
                <span class="n">output_filepath</span><span class="p">,</span>
                <span class="n">cfg</span><span class="p">,</span>
                <span class="n">generate_hpc_scripts</span><span class="p">,</span>
                <span class="n">method_filepaths_mapping</span><span class="o">=</span><span class="n">method_filepaths_mapping</span><span class="p">,</span>
                <span class="n">auxiliary_estimation_input_dir</span><span class="o">=</span><span class="n">auxiliary_estimation_input_dir</span><span class="p">,</span>
            <span class="p">)</span>
    <span class="k">if</span> <span class="n">generating_script</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span><span class="p">,</span> <span class="n">generating_script</span>

    <span class="k">if</span> <span class="n">cfg</span><span class="o">.</span><span class="n">resume</span><span class="p">:</span>
        <span class="n">ensemble_predictions_dict</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">method</span> <span class="ow">in</span> <span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_methods</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;vina&quot;</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">binding_site_method</span> <span class="ow">in</span> <span class="n">cfg</span><span class="o">.</span><span class="n">vina_binding_site_methods</span><span class="p">:</span>
                    <span class="n">method_predictions</span> <span class="o">=</span> <span class="n">get_method_predictions</span><span class="p">(</span>
                        <span class="n">method</span><span class="p">,</span>
                        <span class="n">input_id</span><span class="p">,</span>
                        <span class="n">cfg</span><span class="p">,</span>
                        <span class="n">binding_site_method</span><span class="o">=</span><span class="n">binding_site_method</span><span class="p">,</span>
                        <span class="n">input_protein_filepath</span><span class="o">=</span><span class="n">protein_filepath</span><span class="p">,</span>
                    <span class="p">)</span>
                    <span class="n">ensemble_predictions_dict</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;vina_</span><span class="si">{</span><span class="n">binding_site_method</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">method_predictions</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">method_predictions</span> <span class="o">=</span> <span class="n">get_method_predictions</span><span class="p">(</span>
                    <span class="n">method</span><span class="p">,</span> <span class="n">input_id</span><span class="p">,</span> <span class="n">cfg</span><span class="p">,</span> <span class="n">input_protein_filepath</span><span class="o">=</span><span class="n">protein_filepath</span>
                <span class="p">)</span>
                <span class="n">ensemble_predictions_dict</span><span class="p">[</span><span class="n">method</span><span class="p">]</span> <span class="o">=</span> <span class="n">method_predictions</span>

    <span class="k">return</span> <span class="n">ensemble_predictions_dict</span><span class="p">,</span> <span class="n">generating_script</span></div>



<div class="viewcode-block" id="consensus_rank_ensemble_predictions">
<a class="viewcode-back" href="../../../modules/multicom_ligand.ensemble_generation.html#multicom_ligand.models.ensemble_generation.consensus_rank_ensemble_predictions">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">consensus_rank_ensemble_predictions</span><span class="p">(</span>
    <span class="n">cfg</span><span class="p">:</span> <span class="n">DictConfig</span><span class="p">,</span>
    <span class="n">method_ligand_positions</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">],</span>
    <span class="n">ensemble_predictions_list</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]],</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">RANKED_ENSEMBLE_PREDICTIONS</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Consensus-rank the predictions to select the top prediction(s).</span>

<span class="sd">    :param cfg: Configuration dictionary for runtime arguments.</span>
<span class="sd">    :param method_ligand_positions: List of ligand positions from each</span>
<span class="sd">        method&#39;s predictions.</span>
<span class="sd">    :param ensemble_predictions_list: List of tuples of method name,</span>
<span class="sd">        output protein filepath, and output ligand filepath.</span>
<span class="sd">    :return: Dictionary of consensus-ranked predictions indexed by each</span>
<span class="sd">        prediction&#39;s consensus ranking and valued as its method name,</span>
<span class="sd">        output protein filepath, output ligand filepath, and average</span>
<span class="sd">        pairwise RMSD.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_methods</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span> <span class="ow">or</span> <span class="p">(</span>
        <span class="nb">len</span><span class="p">(</span><span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_methods</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">cfg</span><span class="o">.</span><span class="n">rank_single_method_intrinsically</span>
    <span class="p">):</span>
        <span class="c1"># calculate RMSD values between each pair of method predictions</span>
        <span class="n">rmsd_values_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">positions1</span> <span class="ow">in</span> <span class="n">method_ligand_positions</span><span class="p">:</span>
            <span class="n">rmsd_values</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">positions2</span> <span class="ow">in</span> <span class="n">method_ligand_positions</span><span class="p">:</span>
                <span class="n">rmsd</span> <span class="o">=</span> <span class="n">calculate_rmsd</span><span class="p">(</span><span class="n">positions1</span><span class="p">,</span> <span class="n">positions2</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">rmsd</span> <span class="o">&gt;</span> <span class="mf">0.0</span><span class="p">:</span>  <span class="c1"># avoid self-comparison</span>
                    <span class="n">rmsd_values</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">rmsd</span><span class="p">)</span>
            <span class="n">rmsd_values_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">rmsd_values</span><span class="p">)</span>
        <span class="n">avg_rmsd_values_array</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">rmsd_values_list</span><span class="p">)</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span>

        <span class="c1"># rank the predictions by their average RMSD values</span>
        <span class="n">sorted_indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">avg_rmsd_values_array</span><span class="p">)</span>
        <span class="n">ranked_predictions</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">rank</span> <span class="o">+</span> <span class="mi">1</span><span class="p">:</span> <span class="p">(</span><span class="o">*</span><span class="n">ensemble_predictions_list</span><span class="p">[</span><span class="n">index</span><span class="p">],</span> <span class="n">avg_rmsd_values_array</span><span class="p">[</span><span class="n">index</span><span class="p">])</span>
            <span class="k">for</span> <span class="n">rank</span><span class="p">,</span> <span class="n">index</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">sorted_indices</span><span class="p">)</span>
        <span class="p">}</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># rank the predictions by their explicit rank assignments</span>
        <span class="n">ranked_predictions</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">rank</span> <span class="o">+</span> <span class="mi">1</span><span class="p">:</span> <span class="p">(</span><span class="o">*</span><span class="n">ensemble_predictions_list</span><span class="p">[</span><span class="n">index</span><span class="p">],</span> <span class="mf">0.0</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">rank</span><span class="p">,</span> <span class="n">index</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">ensemble_predictions_list</span><span class="p">)))</span>
        <span class="p">}</span>

    <span class="k">return</span> <span class="n">ranked_predictions</span></div>



<div class="viewcode-block" id="ff_rank_ensemble_predictions">
<a class="viewcode-back" href="../../../modules/multicom_ligand.ensemble_generation.html#multicom_ligand.models.ensemble_generation.ff_rank_ensemble_predictions">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">ff_rank_ensemble_predictions</span><span class="p">(</span>
    <span class="n">cfg</span><span class="p">:</span> <span class="n">DictConfig</span><span class="p">,</span>
    <span class="n">ensemble_predictions_list</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]],</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">RANKED_ENSEMBLE_PREDICTIONS</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Rank the predictions using an OpenMM force field (FF) to select the top</span>
<span class="sd">    prediction(s) according to the criterion of minimum energy.</span>

<span class="sd">    :param cfg: Configuration dictionary for runtime arguments.</span>
<span class="sd">    :param ensemble_predictions_list: List of tuples of method name,</span>
<span class="sd">        output protein filepath, and output ligand filepath.</span>
<span class="sd">    :return: Dictionary of Vina-ranked predictions indexed by each</span>
<span class="sd">        prediction&#39;s consensus ranking and valued as its method name,</span>
<span class="sd">        output protein filepath, output ligand filepath, and Vina energy</span>
<span class="sd">        score.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_methods</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span> <span class="ow">or</span> <span class="p">(</span>
        <span class="nb">len</span><span class="p">(</span><span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_methods</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">cfg</span><span class="o">.</span><span class="n">rank_single_method_intrinsically</span>
    <span class="p">):</span>
        <span class="c1"># calculate Vina energy scores of the (intrinsically) top-ranked method predictions</span>
        <span class="n">vina_energy_scores_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">new_ensemble_predictions_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">prediction</span> <span class="ow">in</span> <span class="n">ensemble_predictions_list</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">method</span><span class="p">,</span> <span class="n">protein_filepath</span><span class="p">,</span> <span class="n">ligand_filepath</span> <span class="o">=</span> <span class="n">prediction</span>
                <span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;_&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="n">protein_filepath</span><span class="p">)</span><span class="o">.</span><span class="n">stem</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[:</span><span class="mi">2</span><span class="p">])</span>
                <span class="k">with</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">TemporaryDirectory</span><span class="p">(</span>
                    <span class="n">suffix</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;_</span><span class="si">{</span><span class="n">method</span><span class="si">}</span><span class="s2">_ensemble_cache_dir&quot;</span>
                <span class="p">)</span> <span class="k">as</span> <span class="n">temp_directory</span><span class="p">:</span>
                    <span class="n">result</span> <span class="o">=</span> <span class="n">minimize_energy</span><span class="p">(</span>
                        <span class="n">OmegaConf</span><span class="o">.</span><span class="n">create</span><span class="p">(</span>
                            <span class="p">{</span>
                                <span class="s2">&quot;protein_file&quot;</span><span class="p">:</span> <span class="n">protein_filepath</span><span class="p">,</span>
                                <span class="s2">&quot;ligand_file&quot;</span><span class="p">:</span> <span class="n">ligand_filepath</span><span class="p">,</span>
                                <span class="s2">&quot;output_file&quot;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span>
                                    <span class="n">Path</span><span class="p">(</span>
                                        <span class="n">Path</span><span class="p">(</span><span class="n">ligand_filepath</span><span class="p">)</span><span class="o">.</span><span class="n">parent</span><span class="p">,</span>
                                        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">Path</span><span class="p">(</span><span class="n">ligand_filepath</span><span class="p">)</span><span class="o">.</span><span class="n">stem</span><span class="si">}</span><span class="s2">_ensemble_relaxed.sdf&quot;</span><span class="p">,</span>
                                    <span class="p">)</span>
                                <span class="p">),</span>
                                <span class="s2">&quot;protein_output_file&quot;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span>
                                    <span class="n">Path</span><span class="p">(</span>
                                        <span class="n">Path</span><span class="p">(</span><span class="n">ligand_filepath</span><span class="p">)</span><span class="o">.</span><span class="n">parent</span><span class="p">,</span>
                                        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">Path</span><span class="p">(</span><span class="n">protein_filepath</span><span class="p">)</span><span class="o">.</span><span class="n">stem</span><span class="si">}</span><span class="s2">_protein_ensemble_relaxed.pdb&quot;</span><span class="p">,</span>
                                    <span class="p">)</span>
                                <span class="p">),</span>
                                <span class="s2">&quot;complex_output_file&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
                                <span class="s2">&quot;temp_dir&quot;</span><span class="p">:</span> <span class="n">temp_directory</span><span class="p">,</span>
                                <span class="s2">&quot;add_solvent&quot;</span><span class="p">:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_add_solvent</span><span class="p">,</span>
                                <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="n">name</span><span class="p">,</span>
                                <span class="s2">&quot;prep_only&quot;</span><span class="p">:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_prep_only</span><span class="p">,</span>
                                <span class="s2">&quot;platform&quot;</span><span class="p">:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_platform</span><span class="p">,</span>
                                <span class="s2">&quot;cuda_device_index&quot;</span><span class="p">:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">cuda_device_index</span><span class="p">,</span>
                                <span class="s2">&quot;log_level&quot;</span><span class="p">:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_log_level</span><span class="p">,</span>
                                <span class="s2">&quot;relax_protein&quot;</span><span class="p">:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_protein</span><span class="p">,</span>
                                <span class="s2">&quot;remove_initial_protein_hydrogens&quot;</span><span class="p">:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_remove_initial_protein_hydrogens</span>
                                <span class="ow">or</span> <span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;rfaa&quot;</span><span class="p">,</span>
                                <span class="s2">&quot;assign_each_ligand_unique_force&quot;</span><span class="p">:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_assign_each_ligand_unique_force</span><span class="p">,</span>
                                <span class="s2">&quot;model_ions&quot;</span><span class="p">:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_model_ions</span><span class="p">,</span>
                                <span class="s2">&quot;cache_files&quot;</span><span class="p">:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_cache_files</span><span class="p">,</span>
                                <span class="s2">&quot;assign_partial_charges_manually&quot;</span><span class="p">:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_assign_partial_charges_manually</span><span class="p">,</span>
                                <span class="s2">&quot;report_initial_energy_only&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                <span class="s2">&quot;max_final_e_value&quot;</span><span class="p">:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_max_final_e_value</span><span class="p">,</span>
                                <span class="s2">&quot;max_num_attempts&quot;</span><span class="p">:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_max_num_attempts</span><span class="p">,</span>
                            <span class="p">}</span>
                        <span class="p">)</span>
                    <span class="p">)</span>
                    <span class="n">new_ensemble_predictions_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">prediction</span><span class="p">)</span>
                    <span class="n">vina_energy_scores_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="s2">&quot;e_init&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">_value</span><span class="p">)</span>

            <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Failed to calculate Vina energy score for prediction </span><span class="si">{</span><span class="n">prediction</span><span class="si">}</span><span class="s2">. Skipping due to: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="p">)</span>
                <span class="k">continue</span>
        <span class="n">ensemble_predictions_list</span> <span class="o">=</span> <span class="n">new_ensemble_predictions_list</span>
        <span class="n">vina_energy_scores_array</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">vina_energy_scores_list</span><span class="p">)</span>

        <span class="c1"># rank the predictions by their Vina energy scores (NOTE: lower is better)</span>
        <span class="n">sorted_indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">vina_energy_scores_array</span><span class="p">)</span>
        <span class="n">ranked_predictions</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">rank</span> <span class="o">+</span> <span class="mi">1</span><span class="p">:</span> <span class="p">(</span><span class="o">*</span><span class="n">ensemble_predictions_list</span><span class="p">[</span><span class="n">index</span><span class="p">],</span> <span class="n">vina_energy_scores_array</span><span class="p">[</span><span class="n">index</span><span class="p">])</span>
            <span class="k">for</span> <span class="n">rank</span><span class="p">,</span> <span class="n">index</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">sorted_indices</span><span class="p">)</span>
        <span class="p">}</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># rank the predictions by their intrinsic (explicit) rank assignments</span>
        <span class="n">ranked_predictions</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">rank</span> <span class="o">+</span> <span class="mi">1</span><span class="p">:</span> <span class="p">(</span><span class="o">*</span><span class="n">ensemble_predictions_list</span><span class="p">[</span><span class="n">index</span><span class="p">],</span> <span class="mf">0.0</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">rank</span><span class="p">,</span> <span class="n">index</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">ensemble_predictions_list</span><span class="p">)))</span>
        <span class="p">}</span>

    <span class="k">return</span> <span class="n">ranked_predictions</span></div>



<div class="viewcode-block" id="rerank_ligand_predictions_according_to_method">
<a class="viewcode-back" href="../../../modules/multicom_ligand.ensemble_generation.html#multicom_ligand.models.ensemble_generation.rerank_ligand_predictions_according_to_method">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">rerank_ligand_predictions_according_to_method</span><span class="p">(</span>
    <span class="n">ranked_ensemble_predictions</span><span class="p">:</span> <span class="n">RANKED_ENSEMBLE_PREDICTIONS</span><span class="p">,</span>
    <span class="n">preferred_method_for_reranking</span><span class="p">:</span> <span class="n">PREFERRED_METHOD_FOR_RERANKING</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">RANKED_ENSEMBLE_PREDICTIONS</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Re-rank predictions to favor structures produced by a preferred</span>
<span class="sd">    structure generation method.</span>

<span class="sd">    :param ranked_ensemble_predictions: Dictionary of ranked ensemble</span>
<span class="sd">        predictions.</span>
<span class="sd">    :param preferred_multi_ligand_method: Name of the preferred</span>
<span class="sd">        structure generation method.</span>
<span class="sd">    :return: Dictionary of re-ranked ensemble predictions.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">preferred_method_counts</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">rank</span><span class="p">,</span> <span class="n">prediction</span> <span class="ow">in</span> <span class="n">ranked_ensemble_predictions</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="c1"># NOTE: lower is better (i.e., higher-ranked) in this case</span>
        <span class="n">method</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">prediction</span>
        <span class="n">preferred_method_counts</span><span class="p">[</span><span class="n">rank</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span> <span class="k">if</span> <span class="n">method</span> <span class="o">==</span> <span class="n">preferred_method_for_reranking</span> <span class="k">else</span> <span class="mi">1</span>

    <span class="n">reranked_predictions</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
        <span class="n">ranked_ensemble_predictions</span><span class="o">.</span><span class="n">items</span><span class="p">(),</span>
        <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">item</span><span class="p">:</span> <span class="p">(</span><span class="n">preferred_method_counts</span><span class="p">[</span><span class="n">item</span><span class="p">[</span><span class="mi">0</span><span class="p">]],</span> <span class="n">item</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span>
    <span class="p">)</span>

    <span class="n">new_ranked_ensemble_predictions</span> <span class="o">=</span> <span class="p">{</span>
        <span class="n">new_rank</span><span class="p">:</span> <span class="n">prediction</span>
        <span class="k">for</span> <span class="n">new_rank</span><span class="p">,</span> <span class="p">(</span><span class="n">_</span><span class="p">,</span> <span class="n">prediction</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">reranked_predictions</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="p">}</span>

    <span class="k">return</span> <span class="n">new_ranked_ensemble_predictions</span></div>



<div class="viewcode-block" id="rerank_clashing_predictions">
<a class="viewcode-back" href="../../../modules/multicom_ligand.ensemble_generation.html#multicom_ligand.models.ensemble_generation.rerank_clashing_predictions">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">rerank_clashing_predictions</span><span class="p">(</span>
    <span class="n">ranked_ensemble_predictions</span><span class="p">:</span> <span class="n">RANKED_ENSEMBLE_PREDICTIONS</span><span class="p">,</span> <span class="n">clash_cutoff</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.2</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">RANKED_ENSEMBLE_PREDICTIONS</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Re-rank predictions to penalize protein residue-residue steric clashes.</span>

<span class="sd">    :param ranked_ensemble_predictions: Dictionary of ranked ensemble</span>
<span class="sd">        predictions.</span>
<span class="sd">    :param clash_cutoff: Cutoff distance for counting (severe) steric</span>
<span class="sd">        clashes.</span>
<span class="sd">    :return: Dictionary of re-ranked ensemble predictions.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">clash_counts</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">rank</span><span class="p">,</span> <span class="n">prediction</span> <span class="ow">in</span> <span class="n">ranked_ensemble_predictions</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">_</span><span class="p">,</span> <span class="n">protein_filepath</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">prediction</span>
        <span class="n">clash_counts</span><span class="p">[</span><span class="n">rank</span><span class="p">]</span> <span class="o">=</span> <span class="n">count_pdb_inter_residue_clashes</span><span class="p">(</span>
            <span class="n">protein_filepath</span><span class="p">,</span> <span class="n">clash_cutoff</span><span class="o">=</span><span class="n">clash_cutoff</span>
        <span class="p">)</span>

    <span class="n">reranked_predictions</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
        <span class="n">ranked_ensemble_predictions</span><span class="o">.</span><span class="n">items</span><span class="p">(),</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">item</span><span class="p">:</span> <span class="p">(</span><span class="n">clash_counts</span><span class="p">[</span><span class="n">item</span><span class="p">[</span><span class="mi">0</span><span class="p">]],</span> <span class="n">item</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="p">)</span>

    <span class="n">new_ranked_ensemble_predictions</span> <span class="o">=</span> <span class="p">{</span>
        <span class="n">new_rank</span><span class="p">:</span> <span class="n">prediction</span>
        <span class="k">for</span> <span class="n">new_rank</span><span class="p">,</span> <span class="p">(</span><span class="n">_</span><span class="p">,</span> <span class="n">prediction</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">reranked_predictions</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="p">}</span>

    <span class="k">return</span> <span class="n">new_ranked_ensemble_predictions</span></div>



<div class="viewcode-block" id="rank_ensemble_predictions">
<a class="viewcode-back" href="../../../modules/multicom_ligand.ensemble_generation.html#multicom_ligand.models.ensemble_generation.rank_ensemble_predictions">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">rank_ensemble_predictions</span><span class="p">(</span>
    <span class="n">ensemble_predictions_dict</span><span class="p">:</span> <span class="n">ENSEMBLE_PREDICTIONS</span><span class="p">,</span>
    <span class="n">name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">is_multi_ligand_target</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
    <span class="n">cfg</span><span class="p">:</span> <span class="n">DictConfig</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">RANKED_ENSEMBLE_PREDICTIONS</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Rank the predictions to select the top prediction(s).</span>

<span class="sd">    :param ensemble_predictions_dict: Dictionary of method names and</span>
<span class="sd">        their corresponding predictions.</span>
<span class="sd">    :param name: Name of the target protein-ligand pair.</span>
<span class="sd">    :param is_multi_ligand_target: Whether the target protein-ligand</span>
<span class="sd">        pair contains multiple ligands.</span>
<span class="sd">    :param cfg: Configuration dictionary for runtime arguments.</span>
<span class="sd">    :return: Dictionary of consensus-ranked predictions indexed by each</span>
<span class="sd">        prediction&#39;s consensus ranking and valued as its method name,</span>
<span class="sd">        output protein filepath, output ligand filepath, and average</span>
<span class="sd">        pairwise RMSD or Vina energy score.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># cache filepath to predicted apo protein structure e.g., from ESMFold</span>
    <span class="k">if</span> <span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_benchmarking</span><span class="p">:</span>
        <span class="n">apo_reference_protein_filepaths</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span>
            <span class="n">Path</span><span class="p">(</span><span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_benchmarking_apo_protein_dir</span><span class="p">)</span><span class="o">.</span><span class="n">rglob</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;*</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">*.pdb&quot;</span><span class="p">)</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">apo_reference_protein_filepaths</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;No apo protein structure found for target </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2"> in directory </span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_benchmarking_apo_protein_dir</span><span class="si">}</span><span class="s2">. Skipping this target...&quot;</span>
            <span class="p">)</span>
            <span class="k">return</span> <span class="p">{}</span>
        <span class="k">assert</span> <span class="p">(</span>
            <span class="nb">len</span><span class="p">(</span><span class="n">apo_reference_protein_filepaths</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span>
        <span class="p">),</span> <span class="sa">f</span><span class="s2">&quot;Expected one apo protein structure for target </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">.&quot;</span>
        <span class="n">apo_reference_protein_filepath</span> <span class="o">=</span> <span class="n">apo_reference_protein_filepaths</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">elif</span> <span class="n">ensemble_predictions_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;diffdock&quot;</span><span class="p">):</span>
        <span class="n">predictions</span> <span class="o">=</span> <span class="n">ensemble_predictions_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;diffdock&quot;</span><span class="p">)</span>
        <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">predictions</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;No predictions found for DiffDock.&quot;</span>
        <span class="k">assert</span> <span class="p">(</span>
            <span class="nb">len</span><span class="p">(</span><span class="n">predictions</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">==</span> <span class="mi">2</span>
        <span class="p">),</span> <span class="s2">&quot;Expected both a protein and ligand prediction filepath for DiffDock.&quot;</span>
        <span class="n">apo_reference_protein_filepath</span> <span class="o">=</span> <span class="n">predictions</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">apo_reference_protein_filepath</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="c1"># determine reference ligand molecule for prediction sanity checking</span>
    <span class="k">if</span> <span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_benchmarking</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">ensemble_predictions_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;neuralplexer&quot;</span><span class="p">):</span>
            <span class="n">reference_ligand_filepath</span> <span class="o">=</span> <span class="n">ensemble_predictions_dict</span><span class="p">[</span><span class="s2">&quot;neuralplexer&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>
        <span class="k">elif</span> <span class="n">ensemble_predictions_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;diffdock&quot;</span><span class="p">):</span>
            <span class="n">reference_ligand_filepath</span> <span class="o">=</span> <span class="n">ensemble_predictions_dict</span><span class="p">[</span><span class="s2">&quot;diffdock&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">reference_ligand_filepath</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">elif</span> <span class="n">ensemble_predictions_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;neuralplexer&quot;</span><span class="p">):</span>
        <span class="n">reference_ligand_filepath</span> <span class="o">=</span> <span class="n">ensemble_predictions_dict</span><span class="p">[</span><span class="s2">&quot;neuralplexer&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">reference_ligand_filepath</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="k">if</span> <span class="n">reference_ligand_filepath</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">assert</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span>
            <span class="n">reference_ligand_filepath</span>
        <span class="p">),</span> <span class="sa">f</span><span class="s2">&quot;Reference ligand not found at </span><span class="si">{</span><span class="n">reference_ligand_filepath</span><span class="si">}</span><span class="s2">.&quot;</span>
        <span class="n">reference_ligand</span> <span class="o">=</span> <span class="n">read_molecule</span><span class="p">(</span><span class="n">reference_ligand_filepath</span><span class="p">)</span>
        <span class="k">assert</span> <span class="p">(</span>
            <span class="n">reference_ligand</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
        <span class="p">),</span> <span class="sa">f</span><span class="s2">&quot;Failed to read reference ligand structure from </span><span class="si">{</span><span class="n">reference_ligand_filepath</span><span class="si">}</span><span class="s2">.&quot;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">reference_ligand</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="c1"># collect ligand positions from each method&#39;s predictions</span>
    <span class="n">unique_ligand_positions</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">method_ligand_positions</span><span class="p">,</span> <span class="n">ensemble_predictions_list</span> <span class="o">=</span> <span class="p">[],</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">method</span> <span class="ow">in</span> <span class="n">ensemble_predictions_dict</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">protein_filepath</span><span class="p">,</span> <span class="n">ligand_filepath</span> <span class="ow">in</span> <span class="n">ensemble_predictions_dict</span><span class="p">[</span><span class="n">method</span><span class="p">]:</span>
            <span class="k">if</span> <span class="p">(</span>
                <span class="n">method</span> <span class="ow">in</span> <span class="n">METHODS_PREDICTING_HOLO_PROTEIN_AB_INITIO</span>
                <span class="ow">and</span> <span class="n">apo_reference_protein_filepath</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
            <span class="p">):</span>
                <span class="n">align_complex_to_protein_only</span><span class="p">(</span>
                    <span class="n">protein_filepath</span><span class="p">,</span> <span class="n">ligand_filepath</span><span class="p">,</span> <span class="n">apo_reference_protein_filepath</span>
                <span class="p">)</span>
                <span class="n">protein_filepath</span> <span class="o">=</span> <span class="n">protein_filepath</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;.pdb&quot;</span><span class="p">,</span> <span class="s2">&quot;_aligned.pdb&quot;</span><span class="p">)</span>
                <span class="n">ligand_filepath</span> <span class="o">=</span> <span class="n">ligand_filepath</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;.sdf&quot;</span><span class="p">,</span> <span class="s2">&quot;_aligned.sdf&quot;</span><span class="p">)</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">ligand</span> <span class="o">=</span> <span class="n">read_molecule</span><span class="p">(</span><span class="n">ligand_filepath</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Failed to read predicted ligand positions of file </span><span class="si">{</span><span class="n">ligand_filepath</span><span class="si">}</span><span class="s2"> from method </span><span class="si">{</span><span class="n">method</span><span class="si">}</span><span class="s2">. Skipping due to: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="p">)</span>
                <span class="k">continue</span>
            <span class="k">if</span> <span class="n">ligand</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Failed to read predicted ligand positions of file </span><span class="si">{</span><span class="n">ligand_filepath</span><span class="si">}</span><span class="s2"> from method </span><span class="si">{</span><span class="n">method</span><span class="si">}</span><span class="s2">. Skipping...&quot;</span>
                <span class="p">)</span>
                <span class="k">continue</span>
            <span class="c1"># check whether predicted structure matches the expected number of atoms;</span>
            <span class="c1"># NOTE: occasionally, for the same SMILES input string, e.g., DynamicBind</span>
            <span class="c1"># may predict (multi-)ligand structures with a different number of atoms</span>
            <span class="k">if</span> <span class="p">(</span>
                <span class="n">reference_ligand</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
                <span class="ow">and</span> <span class="n">ligand</span><span class="o">.</span><span class="n">GetNumAtoms</span><span class="p">()</span> <span class="o">!=</span> <span class="n">reference_ligand</span><span class="o">.</span><span class="n">GetNumAtoms</span><span class="p">()</span>
            <span class="p">):</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Number of atoms in predicted ligand structure </span><span class="si">{</span><span class="n">ligand_filepath</span><span class="si">}</span><span class="s2"> from method </span><span class="si">{</span><span class="n">method</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="n">ligand</span><span class="o">.</span><span class="n">GetNumAtoms</span><span class="p">()</span><span class="si">}</span><span class="s2">) does not match that of the reference ligand (</span><span class="si">{</span><span class="n">reference_ligand</span><span class="o">.</span><span class="n">GetNumAtoms</span><span class="p">()</span><span class="si">}</span><span class="s2">). Skipping </span><span class="si">{</span><span class="n">method</span><span class="si">}</span><span class="s2">&#39;s prediction...&quot;</span>
                <span class="p">)</span>
                <span class="k">continue</span>
            <span class="c1"># ensure only unique ligand position predictions are considered (e.g., as DiffDock may predict the same ligand structure multiple times due to numerical instability)</span>
            <span class="n">ligand_positions</span> <span class="o">=</span> <span class="n">ligand</span><span class="o">.</span><span class="n">GetConformer</span><span class="p">()</span><span class="o">.</span><span class="n">GetPositions</span><span class="p">()</span>
            <span class="n">ligand_positions_tuple</span> <span class="o">=</span> <span class="nb">tuple</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">tuple</span><span class="p">,</span> <span class="n">ligand_positions</span><span class="p">))</span>
            <span class="k">if</span> <span class="n">ligand_positions_tuple</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">unique_ligand_positions</span><span class="p">:</span>
                <span class="n">unique_ligand_positions</span><span class="p">[</span><span class="n">ligand_positions_tuple</span><span class="p">]</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">unique_ligand_positions</span><span class="p">)</span>
                <span class="n">method_ligand_positions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ligand_positions</span><span class="p">)</span>
                <span class="n">ensemble_predictions_list</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">method</span><span class="p">,</span> <span class="n">protein_filepath</span><span class="p">,</span> <span class="n">ligand_filepath</span><span class="p">))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Skipping duplicate ligand position prediction from method </span><span class="si">{</span><span class="n">method</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>

    <span class="c1"># rank the predictions using the requested ranking method</span>
    <span class="k">if</span> <span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_ranking_method</span> <span class="o">==</span> <span class="s2">&quot;consensus&quot;</span><span class="p">:</span>
        <span class="n">ranked_predictions</span> <span class="o">=</span> <span class="n">consensus_rank_ensemble_predictions</span><span class="p">(</span>
            <span class="n">cfg</span><span class="p">,</span>
            <span class="n">method_ligand_positions</span><span class="p">,</span>
            <span class="n">ensemble_predictions_list</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="k">elif</span> <span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_ranking_method</span> <span class="o">==</span> <span class="s2">&quot;ff&quot;</span><span class="p">:</span>
        <span class="n">ranked_predictions</span> <span class="o">=</span> <span class="n">ff_rank_ensemble_predictions</span><span class="p">(</span>
            <span class="n">cfg</span><span class="p">,</span>
            <span class="n">ensemble_predictions_list</span><span class="p">,</span>
        <span class="p">)</span>

    <span class="c1"># for CASP exports, re-rank predictions that contain inter-residue steric clashes</span>
    <span class="k">if</span> <span class="n">cfg</span><span class="o">.</span><span class="n">export_file_format</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;casp16&quot;</span><span class="p">]</span> <span class="ow">and</span> <span class="n">cfg</span><span class="o">.</span><span class="n">rerank_clashing_predictions</span><span class="p">:</span>
        <span class="n">ranked_predictions</span> <span class="o">=</span> <span class="n">rerank_clashing_predictions</span><span class="p">(</span><span class="n">ranked_predictions</span><span class="p">)</span>
    <span class="k">if</span> <span class="p">(</span>
        <span class="n">cfg</span><span class="o">.</span><span class="n">export_file_format</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;casp16&quot;</span><span class="p">]</span>
        <span class="ow">and</span> <span class="n">is_multi_ligand_target</span>
        <span class="ow">and</span> <span class="n">cfg</span><span class="o">.</span><span class="n">rerank_multi_ligand_predictions</span>
    <span class="p">):</span>
        <span class="k">assert</span> <span class="p">(</span>
            <span class="n">cfg</span><span class="o">.</span><span class="n">preferred_multi_ligand_method</span> <span class="ow">in</span> <span class="n">PREFERRED_MULTI_LIGAND_METHODS</span>
        <span class="p">),</span> <span class="s2">&quot;Preferred multi-ligand method must be one of `neuralplexer` and `flowdock`.&quot;</span>
        <span class="n">ranked_predictions</span> <span class="o">=</span> <span class="n">rerank_ligand_predictions_according_to_method</span><span class="p">(</span>
            <span class="n">ranked_predictions</span><span class="p">,</span> <span class="n">cfg</span><span class="o">.</span><span class="n">preferred_multi_ligand_method</span>
        <span class="p">)</span>

    <span class="k">return</span> <span class="n">ranked_predictions</span></div>



<div class="viewcode-block" id="assign_reference_residue_b_factors">
<a class="viewcode-back" href="../../../modules/multicom_ligand.ensemble_generation.html#multicom_ligand.models.ensemble_generation.assign_reference_residue_b_factors">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">assign_reference_residue_b_factors</span><span class="p">(</span>
    <span class="n">protein_output_files</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">protein_reference_filepath</span><span class="p">:</span> <span class="nb">str</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;If `b_factor` column values are not already present, assign the</span>
<span class="sd">    reference protein or nucleic acid structure&#39;s per-residue confidence scores</span>
<span class="sd">    to each output protein. If `b_factor` column values are present but are not</span>
<span class="sd">    in the range [0.0, 100.0] (e.g., they are instead in the range [0.0, 1.0]),</span>
<span class="sd">    scale them to this range via multiplication by 100.0.</span>

<span class="sd">    :param protein_output_files: List of output protein structure PDB</span>
<span class="sd">        filepaths.</span>
<span class="sd">    :param protein_reference_filepath: Path to the input protein</span>
<span class="sd">        structure PDB file.</span>
<span class="sd">    :return: List of output protein structure PDB filepaths with the</span>
<span class="sd">        input protein&#39;s per-residue confidence scores.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">new_protein_output_files</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">set_residue_b_factors</span><span class="p">(</span>
        <span class="n">structure</span><span class="p">:</span> <span class="n">Structure</span><span class="p">,</span> <span class="n">residue_chain_b_factors</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">str</span><span class="p">],</span> <span class="nb">float</span><span class="p">]</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Set the per-residue confidence scores for each residue in the</span>
<span class="sd">        protein structure.</span>

<span class="sd">        :param structure: Structure object of the protein.</span>
<span class="sd">        :param: residue_chain_b_factors: Dictionary of residue keys to</span>
<span class="sd">            their confidence scores.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">model</span> <span class="ow">in</span> <span class="n">structure</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">chain</span> <span class="ow">in</span> <span class="n">model</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">residue</span> <span class="ow">in</span> <span class="n">chain</span><span class="p">:</span>
                    <span class="k">for</span> <span class="n">atom</span> <span class="ow">in</span> <span class="n">residue</span><span class="p">:</span>
                        <span class="n">residue_key</span> <span class="o">=</span> <span class="p">(</span><span class="n">residue</span><span class="o">.</span><span class="n">id</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">chain</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
                        <span class="n">atom</span><span class="o">.</span><span class="n">set_bfactor</span><span class="p">(</span><span class="n">residue_chain_b_factors</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">residue_key</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">))</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">set_atom_b_factors</span><span class="p">(</span><span class="n">structure</span><span class="p">:</span> <span class="n">Structure</span><span class="p">,</span> <span class="n">ref_atom_records</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">Any</span><span class="p">]):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Set the per-atom confidence scores for each atom in the protein</span>
<span class="sd">        structure.</span>

<span class="sd">        :param structure: Structure object of the protein.</span>
<span class="sd">        :param ref_atom_records: List of Atom objects from the reference</span>
<span class="sd">            protein structure.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">atom_index</span><span class="p">,</span> <span class="n">atom</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">structure</span><span class="o">.</span><span class="n">get_atoms</span><span class="p">()):</span>
            <span class="n">ref_atom</span> <span class="o">=</span> <span class="n">ref_atom_records</span><span class="p">[</span><span class="n">atom_index</span><span class="p">]</span>
            <span class="n">atom</span><span class="o">.</span><span class="n">set_bfactor</span><span class="p">(</span><span class="n">ref_atom</span><span class="o">.</span><span class="n">get_bfactor</span><span class="p">())</span>

    <span class="n">parser</span> <span class="o">=</span> <span class="n">PDBParser</span><span class="p">()</span>

    <span class="k">for</span> <span class="n">protein_output_file</span> <span class="ow">in</span> <span class="n">protein_output_files</span><span class="p">:</span>
        <span class="n">structure</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">get_structure</span><span class="p">(</span><span class="s2">&quot;protein&quot;</span><span class="p">,</span> <span class="n">protein_output_file</span><span class="p">)</span>
        <span class="k">assert</span> <span class="n">structure</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;Failed to parse structure from </span><span class="si">{</span><span class="n">protein_output_file</span><span class="si">}</span><span class="s2">.&quot;</span>

        <span class="n">atom_records</span> <span class="o">=</span> <span class="p">[</span><span class="n">atom</span> <span class="k">for</span> <span class="n">atom</span> <span class="ow">in</span> <span class="n">structure</span><span class="o">.</span><span class="n">get_atoms</span><span class="p">()]</span>
        <span class="n">b_factors</span> <span class="o">=</span> <span class="p">[</span><span class="n">atom</span><span class="o">.</span><span class="n">get_bfactor</span><span class="p">()</span> <span class="k">for</span> <span class="n">atom</span> <span class="ow">in</span> <span class="n">atom_records</span><span class="p">]</span>

        <span class="n">duplicate_b_factors</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">b_factors</span><span class="p">))</span> <span class="o">==</span> <span class="mi">1</span>
        <span class="n">decimal_b_factors</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">b_factors</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="mf">1.0</span>
        <span class="k">if</span> <span class="n">duplicate_b_factors</span> <span class="ow">or</span> <span class="n">decimal_b_factors</span><span class="p">:</span>
            <span class="k">with</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">NamedTemporaryFile</span><span class="p">(</span>
                <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">delete</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">suffix</span><span class="o">=</span><span class="s2">&quot;.pdb&quot;</span>
            <span class="p">)</span> <span class="k">as</span> <span class="n">temp_output_protein_file</span><span class="p">:</span>
                <span class="n">input_atom_records</span> <span class="o">=</span> <span class="p">[</span><span class="n">atom</span> <span class="k">for</span> <span class="n">atom</span> <span class="ow">in</span> <span class="n">structure</span><span class="o">.</span><span class="n">get_atoms</span><span class="p">()]</span>
                <span class="k">if</span> <span class="n">duplicate_b_factors</span><span class="p">:</span>
                    <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;All per-residue confidence scores in the input protein structure </span><span class="si">{</span><span class="n">protein_output_file</span><span class="si">}</span><span class="s2"> are identical. Replacing them with the confidence values stored in the predicted structure </span><span class="si">{</span><span class="n">protein_reference_filepath</span><span class="si">}</span><span class="s2">.&quot;</span>
                    <span class="p">)</span>
                    <span class="n">ref_structure</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">get_structure</span><span class="p">(</span>
                        <span class="s2">&quot;input_protein&quot;</span><span class="p">,</span> <span class="n">protein_reference_filepath</span>
                    <span class="p">)</span>
                    <span class="k">assert</span> <span class="p">(</span>
                        <span class="n">ref_structure</span>
                    <span class="p">),</span> <span class="sa">f</span><span class="s2">&quot;Failed to parse structure from </span><span class="si">{</span><span class="n">protein_reference_filepath</span><span class="si">}</span><span class="s2">.&quot;</span>
                    <span class="n">ref_atom_records</span> <span class="o">=</span> <span class="p">[</span><span class="n">atom</span> <span class="k">for</span> <span class="n">atom</span> <span class="ow">in</span> <span class="n">ref_structure</span><span class="o">.</span><span class="n">get_atoms</span><span class="p">()]</span>
                    <span class="n">same_num_atoms</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">input_atom_records</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">ref_atom_records</span><span class="p">)</span>
                    <span class="n">all_atom_records_match</span> <span class="o">=</span> <span class="n">same_num_atoms</span> <span class="ow">and</span> <span class="nb">all</span><span class="p">(</span>
                        <span class="n">input_atom</span><span class="o">.</span><span class="n">name</span> <span class="o">==</span> <span class="n">ref_atom</span><span class="o">.</span><span class="n">name</span>
                        <span class="k">for</span> <span class="n">input_atom</span><span class="p">,</span> <span class="n">ref_atom</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">input_atom_records</span><span class="p">,</span> <span class="n">ref_atom_records</span><span class="p">)</span>
                    <span class="p">)</span>
                    <span class="k">if</span> <span class="n">all_atom_records_match</span><span class="p">:</span>
                        <span class="c1"># assign per-atom confidence scores</span>
                        <span class="n">set_atom_b_factors</span><span class="p">(</span><span class="n">structure</span><span class="p">,</span> <span class="n">ref_atom_records</span><span class="p">)</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="c1"># assign per-residue confidence scores</span>
                        <span class="n">residue_chain_b_factors</span> <span class="o">=</span> <span class="p">{}</span>
                        <span class="k">for</span> <span class="n">atom</span> <span class="ow">in</span> <span class="n">ref_atom_records</span><span class="p">:</span>
                            <span class="n">residue_key</span> <span class="o">=</span> <span class="p">(</span>
                                <span class="n">atom</span><span class="o">.</span><span class="n">get_parent</span><span class="p">()</span><span class="o">.</span><span class="n">id</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>
                                <span class="n">atom</span><span class="o">.</span><span class="n">get_parent</span><span class="p">()</span><span class="o">.</span><span class="n">get_parent</span><span class="p">()</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
                            <span class="p">)</span>
                            <span class="k">if</span> <span class="n">residue_key</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">residue_chain_b_factors</span><span class="p">:</span>
                                <span class="n">residue_chain_b_factors</span><span class="p">[</span><span class="n">residue_key</span><span class="p">]</span> <span class="o">=</span> <span class="n">atom</span><span class="o">.</span><span class="n">get_bfactor</span><span class="p">()</span>
                        <span class="n">set_residue_b_factors</span><span class="p">(</span><span class="n">structure</span><span class="p">,</span> <span class="n">residue_chain_b_factors</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">assert</span> <span class="nb">min</span><span class="p">(</span><span class="n">b_factors</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mf">0.0</span> <span class="ow">and</span> <span class="nb">max</span><span class="p">(</span><span class="n">b_factors</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="mf">1.0</span><span class="p">,</span> <span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;Per-residue confidence scores in the input protein structure </span><span class="si">{</span><span class="n">protein_output_file</span><span class="si">}</span><span class="s2"> are in the range [</span><span class="si">{</span><span class="nb">min</span><span class="p">(</span><span class="n">b_factors</span><span class="p">)</span><span class="si">}</span><span class="s2">, </span><span class="si">{</span><span class="nb">max</span><span class="p">(</span><span class="n">b_factors</span><span class="p">)</span><span class="si">}</span><span class="s2">], not in the range [0.0, 1.0].&quot;</span>
                        <span class="s2">&quot; Please investigate the source of this discrepancy and then re-run this script after this issue has been addressed.&quot;</span>
                    <span class="p">)</span>
                    <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;Per-residue confidence scores in the input protein structure </span><span class="si">{</span><span class="n">protein_output_file</span><span class="si">}</span><span class="s2"> are in the range [</span><span class="si">{</span><span class="nb">min</span><span class="p">(</span><span class="n">b_factors</span><span class="p">)</span><span class="si">}</span><span class="s2">, </span><span class="si">{</span><span class="nb">max</span><span class="p">(</span><span class="n">b_factors</span><span class="p">)</span><span class="si">}</span><span class="s2">], not in the range [0.0, 100.0]. Scaling them to this range via multiplication by 100.0.&quot;</span>
                    <span class="p">)</span>
                    <span class="c1"># assign per-residue confidence scores</span>
                    <span class="n">residue_chain_b_factors</span> <span class="o">=</span> <span class="p">{}</span>
                    <span class="k">for</span> <span class="n">atom</span> <span class="ow">in</span> <span class="n">input_atom_records</span><span class="p">:</span>
                        <span class="n">residue_key</span> <span class="o">=</span> <span class="p">(</span><span class="n">atom</span><span class="o">.</span><span class="n">get_parent</span><span class="p">()</span><span class="o">.</span><span class="n">id</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">atom</span><span class="o">.</span><span class="n">get_parent</span><span class="p">()</span><span class="o">.</span><span class="n">get_parent</span><span class="p">()</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
                        <span class="k">if</span> <span class="n">residue_key</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">residue_chain_b_factors</span><span class="p">:</span>
                            <span class="n">residue_chain_b_factors</span><span class="p">[</span><span class="n">residue_key</span><span class="p">]</span> <span class="o">=</span> <span class="n">atom</span><span class="o">.</span><span class="n">get_bfactor</span><span class="p">()</span> <span class="o">*</span> <span class="mf">100.0</span>
                    <span class="n">set_residue_b_factors</span><span class="p">(</span><span class="n">structure</span><span class="p">,</span> <span class="n">residue_chain_b_factors</span><span class="p">)</span>

                <span class="n">io</span> <span class="o">=</span> <span class="n">PDBIO</span><span class="p">()</span>
                <span class="n">io</span><span class="o">.</span><span class="n">set_structure</span><span class="p">(</span><span class="n">structure</span><span class="p">)</span>
                <span class="n">io</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">temp_output_protein_file</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
                <span class="n">new_protein_output_files</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">temp_output_protein_file</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">new_protein_output_files</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">protein_output_file</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">new_protein_output_files</span></div>



<div class="viewcode-block" id="export_proteins_in_casp_format">
<a class="viewcode-back" href="../../../modules/multicom_ligand.ensemble_generation.html#multicom_ligand.models.ensemble_generation.export_proteins_in_casp_format">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">export_proteins_in_casp_format</span><span class="p">(</span>
    <span class="n">output_protein_filepaths</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
    <span class="n">output_protein_pdb_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">pdb_header</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">append</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">export_casp15_format</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">superligand_inputs</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">model_index</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">gap_insertion_point</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Export the predicted protein structures in CASP format.</span>

<span class="sd">    :param output_protein_filepaths: List of output protein structure</span>
<span class="sd">        PDB filepaths.</span>
<span class="sd">    :param output_protein_pdb_file: Path to the output protein structure</span>
<span class="sd">        PDB file.</span>
<span class="sd">    :param pdb_header: Header string for the PDB file.</span>
<span class="sd">    :param append: Whether to append the predicted protein structures to</span>
<span class="sd">        the output file.</span>
<span class="sd">    :param export_casp15_format: Whether to format the output file for</span>
<span class="sd">        CASP15 benchmarking.</span>
<span class="sd">    :param superligand_inputs: Whether to assume the inputs are for a</span>
<span class="sd">        CASP16 superligand target.</span>
<span class="sd">    :param model_index: Optional index of the model to write to the PDB</span>
<span class="sd">        file.</span>
<span class="sd">    :param gap_insertion_point: Optional `:`-separated string</span>
<span class="sd">        representing the chain-residue pair index of the residue at</span>
<span class="sd">        which to insert a single index gap.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">output_protein_pdb_file</span><span class="p">,</span> <span class="s2">&quot;a&quot;</span> <span class="k">if</span> <span class="n">append</span> <span class="k">else</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="k">if</span> <span class="p">(</span>
            <span class="n">export_casp15_format</span>
            <span class="ow">or</span> <span class="p">(</span><span class="n">append</span> <span class="ow">and</span> <span class="n">model_index</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">model_index</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)</span>
            <span class="ow">or</span> <span class="ow">not</span> <span class="n">append</span>
        <span class="p">):</span>
            <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">pdb_header</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">pdb_file</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">output_protein_filepaths</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
            <span class="n">structure</span> <span class="o">=</span> <span class="n">PDB</span><span class="o">.</span><span class="n">PDBParser</span><span class="p">()</span><span class="o">.</span><span class="n">get_structure</span><span class="p">(</span><span class="s2">&quot;pdb_structure&quot;</span><span class="p">,</span> <span class="n">pdb_file</span><span class="p">)</span>
            <span class="n">structure</span> <span class="o">=</span> <span class="n">renumber_biopython_structure_residues</span><span class="p">(</span>
                <span class="n">structure</span><span class="p">,</span> <span class="n">gap_insertion_point</span><span class="o">=</span><span class="n">gap_insertion_point</span>
            <span class="p">)</span>
            <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">structure</span><span class="o">.</span><span class="n">get_models</span><span class="p">()))</span> <span class="o">==</span> <span class="mi">1</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;Expected one model in </span><span class="si">{</span><span class="n">pdb_file</span><span class="si">}</span><span class="s2">.&quot;</span>
            <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;MODEL </span><span class="si">{</span><span class="n">model_index</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">model_index</span><span class="w"> </span><span class="ow">is</span><span class="w"> </span><span class="ow">not</span><span class="w"> </span><span class="kc">None</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="n">i</span><span class="si">}</span><span class="se">\n</span><span class="s2">REMARK N/A</span><span class="se">\n</span><span class="s2">PARENT N/A</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="p">)</span>
            <span class="n">io</span> <span class="o">=</span> <span class="n">PDB</span><span class="o">.</span><span class="n">PDBIO</span><span class="p">()</span>
            <span class="n">io</span><span class="o">.</span><span class="n">set_structure</span><span class="p">(</span><span class="n">structure</span><span class="p">)</span>
            <span class="n">io</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="n">write_end</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">append</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">superligand_inputs</span><span class="p">:</span>
                <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;END</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;CASP protein submission file saved to </span><span class="si">{</span><span class="n">output_protein_pdb_file</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span></div>



<div class="viewcode-block" id="export_ligands_in_casp15_format">
<a class="viewcode-back" href="../../../modules/multicom_ligand.ensemble_generation.html#multicom_ligand.models.ensemble_generation.export_ligands_in_casp15_format">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">export_ligands_in_casp15_format</span><span class="p">(</span>
    <span class="n">output_ligand_filepaths</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
    <span class="n">output_ligand_sdf_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">sdf_header</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">method</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">append</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">model_index</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">ligand_numbers</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">]]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">ligand_names</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Export the predicted ligand structures in CASP15 format.</span>

<span class="sd">    Note that for the sake of consistency when evaluating deep learning docking methods</span>
<span class="sd">    on the CASP15 benchmark, we only report a single pose per protein-ligand model</span>
<span class="sd">    (i.e., 5 submitted protein-ligand models with 1 pose per model vs. 5 poses per model).</span>

<span class="sd">    :param output_ligand_filepaths: List of output ligand structure SDF filepaths.</span>
<span class="sd">    :param ligand_output_filepath: Path to the output ligand structure SDF file.</span>
<span class="sd">    :param sdf_header: Header string for the SDF file.</span>
<span class="sd">    :param method: Method name.</span>
<span class="sd">    :param append: Whether to append the predicted ligand structures to the output file.</span>
<span class="sd">    :param model_index: Optional index of the model to write to the SDF file.</span>
<span class="sd">    :param ligand_numbers: Optional list of ligand numbers represented as a `_`-delimited string or a list of integers.</span>
<span class="sd">    :param ligand_names: Optional list of ligand names represented as a `_`-delimited string or a list of strings.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">sdf_content</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">()</span>
    <span class="n">sdf_content</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">sdf_header</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">sdf_file</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">output_ligand_filepaths</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
        <span class="n">mol</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolFromMolFile</span><span class="p">(</span><span class="n">sdf_file</span><span class="p">,</span> <span class="n">sanitize</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">mol_frags</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">GetMolFrags</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="n">asMols</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">sanitizeFrags</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="n">ligand_numbers_list</span> <span class="o">=</span> <span class="p">(</span>
            <span class="p">(</span>
                <span class="n">ligand_numbers</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)</span>
                <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">ligand_numbers</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span>
                <span class="k">else</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">num</span><span class="si">:</span><span class="s2">03</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">num</span> <span class="ow">in</span> <span class="n">ligand_numbers</span><span class="p">]</span>
            <span class="p">)</span>
            <span class="k">if</span> <span class="n">ligand_numbers</span>
            <span class="k">else</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">i</span><span class="si">:</span><span class="s2">03</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">mol_frags</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)]</span>
        <span class="p">)</span>
        <span class="n">ligand_names_list</span> <span class="o">=</span> <span class="p">(</span>
            <span class="p">(</span><span class="n">ligand_names</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">ligand_names</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span> <span class="k">else</span> <span class="n">ligand_names</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">ligand_names</span>
            <span class="k">else</span> <span class="p">[</span><span class="s2">&quot;LIG&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">mol_frags</span><span class="p">)</span>
        <span class="p">)</span>

        <span class="k">if</span> <span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;rfaa&quot;</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">ligand_numbers_list</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">ligand_names_list</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">mol_frags</span><span class="p">)):</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                    <span class="s2">&quot;Number of RFAA ligand numbers, names, and molecule fragments do not match. Note that this means it did not predict for all input ligands and that manual adjustments to the resulting CASP15 submission file may need to be made (e.g., to make sure ligand names are correctly aligned with listed molecular fragments).&quot;</span>
                <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">assert</span> <span class="p">(</span>
                <span class="nb">len</span><span class="p">(</span><span class="n">ligand_numbers_list</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">ligand_names_list</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">mol_frags</span><span class="p">)</span>
            <span class="p">),</span> <span class="s2">&quot;Number of ligand numbers, names, and molecule fragments must match.&quot;</span>

        <span class="n">sdf_content</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;MODEL </span><span class="si">{</span><span class="n">model_index</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">model_index</span><span class="w"> </span><span class="ow">is</span><span class="w"> </span><span class="ow">not</span><span class="w"> </span><span class="kc">None</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="n">i</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">mol_frag_index</span><span class="p">,</span> <span class="n">mol_frag</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">mol_frags</span><span class="p">):</span>
            <span class="n">sdf_content</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;LIGAND </span><span class="si">{</span><span class="n">ligand_numbers_list</span><span class="p">[</span><span class="n">mol_frag_index</span><span class="p">]</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">ligand_names_list</span><span class="p">[</span><span class="n">mol_frag_index</span><span class="p">]</span><span class="si">}</span><span class="se">\n</span><span class="s2">POSE 1</span><span class="se">\n</span><span class="si">{</span><span class="n">ligand_names_list</span><span class="p">[</span><span class="n">mol_frag_index</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="p">)</span>
            <span class="n">AllChem</span><span class="o">.</span><span class="n">ComputeGasteigerCharges</span><span class="p">(</span><span class="n">mol_frag</span><span class="p">)</span>
            <span class="n">sdf_content</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">Chem</span><span class="o">.</span><span class="n">MolToMolBlock</span><span class="p">(</span><span class="n">mol_frag</span><span class="p">))</span>

        <span class="n">sdf_content</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;END</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">output_ligand_sdf_file</span><span class="p">,</span> <span class="s2">&quot;a&quot;</span> <span class="k">if</span> <span class="n">append</span> <span class="k">else</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">sdf_content</span><span class="o">.</span><span class="n">getvalue</span><span class="p">())</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;CASP15 ligand submission file saved to </span><span class="si">{</span><span class="n">output_ligand_sdf_file</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span></div>



<div class="viewcode-block" id="export_ligands_in_casp16_format">
<a class="viewcode-back" href="../../../modules/multicom_ligand.ensemble_generation.html#multicom_ligand.models.ensemble_generation.export_ligands_in_casp16_format">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">export_ligands_in_casp16_format</span><span class="p">(</span>
    <span class="n">output_ligand_filepaths</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
    <span class="n">output_ligand_sdf_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">sdf_header</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">method</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">append</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">model_index</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">only_affinity_requested</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">ligand_numbers</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">]]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">ligand_names</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">ligand_tasks</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">aux_estimation_df</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Export the predicted ligand structures in CASP16 format.</span>

<span class="sd">    :param output_ligand_filepaths: List of output ligand structure SDF filepaths.</span>
<span class="sd">    :param ligand_output_filepath: Path to the output ligand structure SDF file.</span>
<span class="sd">    :param sdf_header: Header string for the SDF file.</span>
<span class="sd">    :param method: Method name.</span>
<span class="sd">    :param append: Whether to append the predicted ligand structures to the output file.</span>
<span class="sd">    :param model_index: Optional index of the model to write to the SDF file.</span>
<span class="sd">    :param only_affinity_requested: Whether to only include affinity scores in the output file.</span>
<span class="sd">    :param ligand_numbers: Optional list of ligand numbers represented as a `_`-delimited string or a list of integers.</span>
<span class="sd">    :param ligand_names: Optional list of ligand names represented as a `_`-delimited string or a list of strings.</span>
<span class="sd">    :param ligand_tasks: Optional ligand tasks specification.</span>
<span class="sd">    :param aux_estimation_df: Optional auxiliary DataFrame containing estimated ligand properties.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">sdf_content</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">()</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">append</span> <span class="ow">or</span> <span class="n">only_affinity_requested</span><span class="p">:</span>
        <span class="n">sdf_content</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">sdf_header</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">sdf_file</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">output_ligand_filepaths</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
        <span class="n">mol</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolFromMolFile</span><span class="p">(</span><span class="n">sdf_file</span><span class="p">,</span> <span class="n">sanitize</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">mol_frags</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">GetMolFrags</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="n">asMols</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">sanitizeFrags</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="n">ligand_numbers_list</span> <span class="o">=</span> <span class="p">(</span>
            <span class="p">(</span>
                <span class="n">ligand_numbers</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)</span>
                <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">ligand_numbers</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span>
                <span class="k">else</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">num</span><span class="si">:</span><span class="s2">03</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">num</span> <span class="ow">in</span> <span class="n">ligand_numbers</span><span class="p">]</span>
            <span class="p">)</span>
            <span class="k">if</span> <span class="n">ligand_numbers</span>
            <span class="k">else</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">i</span><span class="si">:</span><span class="s2">03</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">mol_frags</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)]</span>
        <span class="p">)</span>
        <span class="n">ligand_names_list</span> <span class="o">=</span> <span class="p">(</span>
            <span class="p">(</span><span class="n">ligand_names</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">ligand_names</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span> <span class="k">else</span> <span class="n">ligand_names</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">ligand_names</span>
            <span class="k">else</span> <span class="p">[</span><span class="s2">&quot;LIG&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">mol_frags</span><span class="p">)</span>
        <span class="p">)</span>
        <span class="n">ligand_plddt_match</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;plddt(\d+\.\d+)&quot;</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">sdf_file</span><span class="p">))</span>
        <span class="n">ligand_plddt_value</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">ligand_plddt_match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">))</span> <span class="k">if</span> <span class="n">ligand_plddt_match</span> <span class="k">else</span> <span class="kc">None</span>

        <span class="k">if</span> <span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;rfaa&quot;</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">ligand_numbers_list</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">ligand_names_list</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">mol_frags</span><span class="p">)):</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                    <span class="s2">&quot;Number of RFAA ligand numbers, names, and molecule fragments do not match. Note that this means it did not predict for all input ligands and that manual adjustments to the resulting CASP submission file may need to be made (e.g., to make sure ligand names are correctly aligned with listed molecular fragments).&quot;</span>
                <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">assert</span> <span class="p">(</span>
                <span class="nb">len</span><span class="p">(</span><span class="n">ligand_numbers_list</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">ligand_names_list</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">mol_frags</span><span class="p">)</span>
            <span class="p">),</span> <span class="s2">&quot;Number of ligand numbers, names, and molecule fragments must match.&quot;</span>

        <span class="k">if</span> <span class="n">only_affinity_requested</span><span class="p">:</span>
            <span class="n">sdf_content</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;MODEL </span><span class="si">{</span><span class="n">model_index</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">model_index</span><span class="w"> </span><span class="ow">is</span><span class="w"> </span><span class="ow">not</span><span class="w"> </span><span class="kc">None</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="n">i</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">aux_estimation_df</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">assert</span> <span class="p">(</span>
                    <span class="n">model_index</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
                <span class="p">),</span> <span class="s2">&quot;Model index must be provided when auxiliary estimation data is available and referenced.&quot;</span>
                <span class="n">rank_aux_estimation_row</span> <span class="o">=</span> <span class="n">aux_estimation_df</span><span class="p">[</span>
                    <span class="n">aux_estimation_df</span><span class="p">[</span><span class="s2">&quot;rank&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">model_index</span>
                <span class="p">]</span>
                <span class="n">ligand_plddt_values</span> <span class="o">=</span> <span class="n">rank_aux_estimation_row</span><span class="p">[</span><span class="s2">&quot;plddt_ligs&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">item</span><span class="p">()</span>
                <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">ligand_plddt_values</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span>
                    <span class="n">mol_frags</span>
                <span class="p">),</span> <span class="sa">f</span><span class="s2">&quot;Number of ligand plDDT values must match the number of ligand fragments for method </span><span class="si">{</span><span class="n">method</span><span class="si">}</span><span class="s2">.&quot;</span>

            <span class="k">for</span> <span class="n">mol_frag_index</span><span class="p">,</span> <span class="n">mol_frag</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">mol_frags</span><span class="p">):</span>
                <span class="n">ligand_plddt_value</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="n">ligand_plddt_values</span><span class="p">[</span><span class="n">mol_frag_index</span><span class="p">]</span>
                    <span class="k">if</span> <span class="n">aux_estimation_df</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
                    <span class="k">else</span> <span class="n">ligand_plddt_value</span>
                <span class="p">)</span>
                <span class="n">ligand_plddt_string</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;LSCORE </span><span class="si">{</span><span class="n">ligand_plddt_value</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
                    <span class="k">if</span> <span class="n">ligand_plddt_value</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">only_affinity_requested</span>
                    <span class="k">else</span> <span class="s2">&quot;&quot;</span>
                <span class="p">)</span>
                <span class="n">sdf_content</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;LIGAND </span><span class="si">{</span><span class="n">ligand_numbers_list</span><span class="p">[</span><span class="n">mol_frag_index</span><span class="p">]</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">ligand_names_list</span><span class="p">[</span><span class="n">mol_frag_index</span><span class="p">]</span><span class="si">}</span><span class="se">\n</span><span class="si">{</span><span class="n">ligand_plddt_string</span><span class="si">}{</span><span class="n">ligand_names_list</span><span class="p">[</span><span class="n">mol_frag_index</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="p">)</span>
                <span class="n">mol_frag</span><span class="o">.</span><span class="n">SetProp</span><span class="p">(</span>
                    <span class="s2">&quot;_Name&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span>
                <span class="p">)</span>  <span class="c1"># NOTE: CASP16 formatting requires null ligand conformer names</span>
                <span class="n">AllChem</span><span class="o">.</span><span class="n">ComputeGasteigerCharges</span><span class="p">(</span><span class="n">mol_frag</span><span class="p">)</span>
                <span class="n">sdf_content</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">Chem</span><span class="o">.</span><span class="n">MolToMolBlock</span><span class="p">(</span><span class="n">mol_frag</span><span class="p">))</span>

        <span class="n">ligand_affinity_match</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;affinity(\d+\.\d+)&quot;</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">sdf_file</span><span class="p">))</span>
        <span class="n">ligand_affinity_value</span> <span class="o">=</span> <span class="p">(</span>
            <span class="nb">float</span><span class="p">(</span><span class="n">ligand_affinity_match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">))</span> <span class="k">if</span> <span class="n">ligand_affinity_match</span> <span class="k">else</span> <span class="kc">None</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="n">aux_estimation_df</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">assert</span> <span class="p">(</span>
                <span class="n">model_index</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
            <span class="p">),</span> <span class="s2">&quot;Model index must be provided when auxiliary estimation data is available and referenced.&quot;</span>
            <span class="n">rank_aux_estimation_row</span> <span class="o">=</span> <span class="n">aux_estimation_df</span><span class="p">[</span><span class="n">aux_estimation_df</span><span class="p">[</span><span class="s2">&quot;rank&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">model_index</span><span class="p">]</span>
            <span class="n">ligand_aff_values</span> <span class="o">=</span> <span class="n">rank_aux_estimation_row</span><span class="p">[</span><span class="s2">&quot;affinity_ligs&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">item</span><span class="p">()</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">ligand_aff_values</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
                <span class="n">ligand_aff_values</span> <span class="o">=</span> <span class="n">ast</span><span class="o">.</span><span class="n">literal_eval</span><span class="p">(</span><span class="n">ligand_aff_values</span><span class="p">)</span>
            <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">ligand_aff_values</span><span class="p">,</span> <span class="nb">float</span><span class="p">)</span> <span class="ow">and</span> <span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">ligand_aff_values</span><span class="p">):</span>
                <span class="n">ligand_aff_values</span> <span class="o">=</span> <span class="p">[</span><span class="mf">0.0</span><span class="p">]</span>
            <span class="k">assert</span> <span class="p">(</span>
                <span class="nb">len</span><span class="p">(</span><span class="n">ligand_aff_values</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">1</span>
            <span class="p">),</span> <span class="s2">&quot;Number of ligand affinity values per complex must be at least 1.&quot;</span>
            <span class="n">ligand_affinity_value</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">ligand_aff_values</span><span class="p">)</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">ligand_aff_values</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">ligand_affinity_value</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="c1"># convert predicted pK to nanomoles/liter (nM)</span>
            <span class="n">ligand_affinity_value_in_moles</span> <span class="o">=</span> <span class="mi">10</span> <span class="o">**</span> <span class="p">(</span><span class="o">-</span><span class="n">ligand_affinity_value</span><span class="p">)</span>
            <span class="n">ligand_affinity_value_in_nanomoles</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">ligand_affinity_value_in_moles</span> <span class="o">/</span> <span class="n">AFFINITY_UNIT_CONVERSION_DICT</span><span class="p">[</span><span class="s2">&quot;nM&quot;</span><span class="p">]</span>
            <span class="p">)</span>
            <span class="n">ligand_affinity_value</span> <span class="o">=</span> <span class="n">ligand_affinity_value_in_nanomoles</span>
        <span class="n">ligand_affinity_string</span> <span class="o">=</span> <span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;AFFNTY </span><span class="si">{</span><span class="n">ligand_affinity_value</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2"> aa</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="k">if</span> <span class="n">ligand_affinity_value</span>
            <span class="ow">and</span> <span class="p">(</span><span class="n">only_affinity_requested</span> <span class="ow">or</span> <span class="p">(</span><span class="n">ligand_tasks</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="s2">&quot;A&quot;</span> <span class="ow">in</span> <span class="n">ligand_tasks</span><span class="p">))</span>
            <span class="k">else</span> <span class="s2">&quot;&quot;</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="n">ligand_affinity_string</span><span class="p">:</span>
            <span class="n">sdf_content</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">ligand_affinity_string</span><span class="p">)</span>

        <span class="n">sdf_content</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;END</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">output_ligand_sdf_file</span><span class="p">,</span> <span class="s2">&quot;a&quot;</span> <span class="k">if</span> <span class="n">append</span> <span class="k">else</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">sdf_content</span><span class="o">.</span><span class="n">getvalue</span><span class="p">())</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;CASP16 ligand submission file saved to </span><span class="si">{</span><span class="n">output_ligand_sdf_file</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span></div>



<div class="viewcode-block" id="save_ranked_predictions">
<a class="viewcode-back" href="../../../modules/multicom_ligand.ensemble_generation.html#multicom_ligand.models.ensemble_generation.save_ranked_predictions">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">save_ranked_predictions</span><span class="p">(</span>
    <span class="n">ranked_predictions</span><span class="p">:</span> <span class="n">RANKED_ENSEMBLE_PREDICTIONS</span><span class="p">,</span>
    <span class="n">protein_input_filepath</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">ligand_numbers</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">]]],</span>
    <span class="n">ligand_names</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]],</span>
    <span class="n">ligand_tasks</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
    <span class="n">cfg</span><span class="p">:</span> <span class="n">DictConfig</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Save the top-ranked predictions to the output directory.</span>

<span class="sd">    :param ranked_predictions: Dictionary of ranked predictions indexed by each</span>
<span class="sd">        prediction&#39;s ranking and valued as its method name, output protein filepath,</span>
<span class="sd">        output ligand filepath, and average pairwise RMSD or Vina energy score.</span>
<span class="sd">    :param protein_input_filepath: Path to the input protein structure PDB file.</span>
<span class="sd">    :param name: Name of the target protein-ligand pair.</span>
<span class="sd">    :param ligand_numbers: Optional list of ligand numbers represented as a `_`-delimited string or a list of integers.</span>
<span class="sd">    :param ligand_names: Optional list of ligand names represented as a `_`-delimited string or a list of strings.</span>
<span class="sd">    :param ligand_tasks: Optional ligand tasks specification.</span>
<span class="sd">    :param cfg: Configuration dictionary for runtime arguments.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">ranking_metric</span> <span class="o">=</span> <span class="s2">&quot;ff&quot;</span> <span class="k">if</span> <span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_ranking_method</span> <span class="o">==</span> <span class="s2">&quot;ff&quot;</span> <span class="k">else</span> <span class="s2">&quot;rmsd&quot;</span>
    <span class="n">relax_complex</span> <span class="o">=</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_method_ligands_pre_ranking</span> <span class="ow">or</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_method_ligands_post_ranking</span>
    <span class="n">ligand_relaxed_postfix</span> <span class="o">=</span> <span class="s2">&quot;_relaxed&quot;</span> <span class="k">if</span> <span class="n">relax_complex</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span>
    <span class="n">protein_relaxed_postfix</span> <span class="o">=</span> <span class="n">ligand_relaxed_postfix</span> <span class="k">if</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_protein</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span>

    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">cfg</span><span class="o">.</span><span class="n">output_dir</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cfg</span><span class="o">.</span><span class="n">output_dir</span><span class="p">,</span> <span class="n">name</span> <span class="o">+</span> <span class="n">ligand_relaxed_postfix</span><span class="p">),</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="n">relaxation_success_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">if</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_method_ligands_post_ranking</span><span class="p">:</span>
        <span class="c1"># relax ligand structures after ranking as desired;</span>
        <span class="c1"># NOTE: relaxation is practically necessary when making</span>
        <span class="c1"># multi-ligand predictions with DiffDock and DynamicBind,</span>
        <span class="c1"># and is optional (yet highly recommended) for all other predictions</span>
        <span class="n">pool</span> <span class="o">=</span> <span class="n">multiprocessing</span><span class="o">.</span><span class="n">Pool</span><span class="p">(</span><span class="n">processes</span><span class="o">=</span><span class="n">cfg</span><span class="o">.</span><span class="n">relax_num_processes</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">index</span><span class="p">,</span> <span class="p">(</span>
            <span class="n">rank</span><span class="p">,</span>
            <span class="p">(</span>
                <span class="n">method</span><span class="p">,</span>
                <span class="n">protein_filepath</span><span class="p">,</span>
                <span class="n">ligand_filepath</span><span class="p">,</span>
                <span class="n">_</span><span class="p">,</span>
            <span class="p">),</span>
        <span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">ranked_predictions</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
            <span class="c1"># only relax the top-N predictions as specified</span>
            <span class="k">if</span> <span class="n">cfg</span><span class="o">.</span><span class="n">export_top_n</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">if</span> <span class="p">(</span><span class="n">index</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">&gt;</span> <span class="n">cfg</span><span class="o">.</span><span class="n">export_top_n</span><span class="p">:</span>
                    <span class="k">break</span>
            <span class="k">if</span> <span class="p">(</span>
                <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">ligand_filepath</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;.sdf&quot;</span><span class="p">,</span> <span class="s2">&quot;_ensemble_relaxed.sdf&quot;</span><span class="p">))</span>
                <span class="ow">or</span> <span class="ow">not</span> <span class="n">cfg</span><span class="o">.</span><span class="n">skip_existing</span>
            <span class="p">):</span>
                <span class="k">with</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">TemporaryDirectory</span><span class="p">(</span>
                    <span class="n">suffix</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;_</span><span class="si">{</span><span class="n">method</span><span class="si">}</span><span class="s2">_ensemble_cache_dir&quot;</span>
                <span class="p">)</span> <span class="k">as</span> <span class="n">temp_directory</span><span class="p">:</span>
                    <span class="k">try</span><span class="p">:</span>
                        <span class="n">result</span> <span class="o">=</span> <span class="n">pool</span><span class="o">.</span><span class="n">apply_async</span><span class="p">(</span>
                            <span class="n">relax_single_filepair</span><span class="p">,</span>
                            <span class="n">args</span><span class="o">=</span><span class="p">(</span>
                                <span class="n">Path</span><span class="p">(</span><span class="n">protein_filepath</span><span class="p">),</span>
                                <span class="n">Path</span><span class="p">(</span><span class="n">ligand_filepath</span><span class="p">),</span>
                                <span class="n">Path</span><span class="p">(</span><span class="n">ligand_filepath</span><span class="p">)</span><span class="o">.</span><span class="n">parent</span><span class="p">,</span>
                                <span class="n">temp_directory</span><span class="p">,</span>
                                <span class="n">OmegaConf</span><span class="o">.</span><span class="n">create</span><span class="p">(</span>
                                    <span class="p">{</span>
                                        <span class="s2">&quot;method&quot;</span><span class="p">:</span> <span class="s2">&quot;ensemble&quot;</span><span class="p">,</span>
                                        <span class="s2">&quot;add_solvent&quot;</span><span class="p">:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_add_solvent</span><span class="p">,</span>
                                        <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="n">name</span><span class="p">,</span>
                                        <span class="s2">&quot;prep_only&quot;</span><span class="p">:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_prep_only</span><span class="p">,</span>
                                        <span class="s2">&quot;platform&quot;</span><span class="p">:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_platform</span><span class="p">,</span>
                                        <span class="s2">&quot;cuda_device_index&quot;</span><span class="p">:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">cuda_device_index</span><span class="p">,</span>
                                        <span class="s2">&quot;log_level&quot;</span><span class="p">:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_log_level</span><span class="p">,</span>
                                        <span class="s2">&quot;relax_protein&quot;</span><span class="p">:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_protein</span><span class="p">,</span>
                                        <span class="s2">&quot;remove_initial_protein_hydrogens&quot;</span><span class="p">:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_remove_initial_protein_hydrogens</span>
                                        <span class="ow">or</span> <span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;rfaa&quot;</span><span class="p">,</span>
                                        <span class="s2">&quot;assign_each_ligand_unique_force&quot;</span><span class="p">:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_assign_each_ligand_unique_force</span><span class="p">,</span>
                                        <span class="s2">&quot;model_ions&quot;</span><span class="p">:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_model_ions</span><span class="p">,</span>
                                        <span class="s2">&quot;cache_files&quot;</span><span class="p">:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_cache_files</span><span class="p">,</span>
                                        <span class="s2">&quot;assign_partial_charges_manually&quot;</span><span class="p">:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_assign_partial_charges_manually</span><span class="p">,</span>
                                        <span class="s2">&quot;report_initial_energy_only&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                        <span class="s2">&quot;max_final_e_value&quot;</span><span class="p">:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_max_final_e_value</span><span class="p">,</span>
                                        <span class="s2">&quot;max_num_attempts&quot;</span><span class="p">:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_max_num_attempts</span><span class="p">,</span>
                                        <span class="s2">&quot;skip_existing&quot;</span><span class="p">:</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_skip_existing</span><span class="p">,</span>
                                    <span class="p">}</span>
                                <span class="p">),</span>
                            <span class="p">),</span>
                        <span class="p">)</span>
                        <span class="n">result</span><span class="o">.</span><span class="n">get</span><span class="p">()</span>
                        <span class="n">relaxation_success_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
                    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                        <span class="n">relaxation_success_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                            <span class="sa">f</span><span class="s2">&quot;Failed to relax ligand structure </span><span class="si">{</span><span class="n">ligand_filepath</span><span class="si">}</span><span class="s2"> from method </span><span class="si">{</span><span class="n">method</span><span class="si">}</span><span class="s2"> due to: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">. Skipping relaxation...&quot;</span>
                        <span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">relaxation_success_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>

        <span class="n">pool</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
        <span class="n">pool</span><span class="o">.</span><span class="n">join</span><span class="p">()</span>

        <span class="k">for</span> <span class="n">index</span><span class="p">,</span> <span class="p">(</span>
            <span class="n">rank</span><span class="p">,</span>
            <span class="p">(</span>
                <span class="n">method</span><span class="p">,</span>
                <span class="n">protein_filepath</span><span class="p">,</span>
                <span class="n">ligand_filepath</span><span class="p">,</span>
                <span class="n">ranking_metric_value</span><span class="p">,</span>
            <span class="p">),</span>
        <span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">ranked_predictions</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
            <span class="c1"># only relax the top-N predictions as specified</span>
            <span class="k">if</span> <span class="n">cfg</span><span class="o">.</span><span class="n">export_top_n</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">if</span> <span class="p">(</span><span class="n">index</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">&gt;</span> <span class="n">cfg</span><span class="o">.</span><span class="n">export_top_n</span><span class="p">:</span>
                    <span class="k">break</span>
            <span class="k">if</span> <span class="p">(</span>
                <span class="mi">0</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">relaxation_success_list</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="nb">len</span><span class="p">(</span><span class="n">ranked_predictions</span><span class="p">)</span>
                <span class="ow">and</span> <span class="n">relaxation_success_list</span><span class="p">[</span><span class="n">index</span><span class="p">]</span>
            <span class="p">):</span>
                <span class="k">assert</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span>
                    <span class="n">ligand_filepath</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;.sdf&quot;</span><span class="p">,</span> <span class="s2">&quot;_ensemble_relaxed.sdf&quot;</span><span class="p">)</span>
                <span class="p">),</span> <span class="sa">f</span><span class="s2">&quot;Ligand relaxation failed for file: </span><span class="si">{</span><span class="n">ligand_filepath</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;.sdf&#39;</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;_ensemble_relaxed.sdf&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="k">if</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_protein</span><span class="p">:</span>
                    <span class="n">maybe_relaxed_protein_filepath</span> <span class="o">=</span> <span class="n">protein_filepath</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span>
                        <span class="s2">&quot;.pdb&quot;</span><span class="p">,</span> <span class="s2">&quot;_protein_ensemble_relaxed.pdb&quot;</span>
                    <span class="p">)</span>
                    <span class="k">assert</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span>
                        <span class="n">maybe_relaxed_protein_filepath</span>
                    <span class="p">),</span> <span class="sa">f</span><span class="s2">&quot;Protein relaxation failed for file: </span><span class="si">{</span><span class="n">maybe_relaxed_protein_filepath</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">maybe_relaxed_protein_filepath</span> <span class="o">=</span> <span class="n">protein_filepath</span>
                <span class="n">ranked_predictions</span><span class="p">[</span><span class="n">rank</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="n">method</span><span class="p">,</span>
                    <span class="n">maybe_relaxed_protein_filepath</span><span class="p">,</span>
                    <span class="n">ligand_filepath</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;.sdf&quot;</span><span class="p">,</span> <span class="s2">&quot;_ensemble_relaxed.sdf&quot;</span><span class="p">),</span>
                    <span class="n">ranking_metric_value</span><span class="p">,</span>
                <span class="p">)</span>

    <span class="k">if</span> <span class="n">cfg</span><span class="o">.</span><span class="n">superligand_inputs</span><span class="p">:</span>
        <span class="n">output_ligand_filepaths</span><span class="p">,</span> <span class="n">output_protein_filepaths</span> <span class="o">=</span> <span class="p">[],</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">rank</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">cfg</span><span class="o">.</span><span class="n">export_top_n</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
            <span class="n">ligand_output_filepath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                <span class="n">cfg</span><span class="o">.</span><span class="n">output_dir</span><span class="p">,</span>
                <span class="n">name</span> <span class="o">+</span> <span class="n">ligand_relaxed_postfix</span><span class="p">,</span>
                <span class="sa">f</span><span class="s2">&quot;*_rank</span><span class="si">{</span><span class="n">rank</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">ranking_metric</span><span class="si">}</span><span class="s2">*_pbvalid=*.sdf&quot;</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">protein_output_filepath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                <span class="n">cfg</span><span class="o">.</span><span class="n">output_dir</span><span class="p">,</span>
                <span class="n">name</span> <span class="o">+</span> <span class="n">ligand_relaxed_postfix</span><span class="p">,</span>
                <span class="sa">f</span><span class="s2">&quot;*_rank</span><span class="si">{</span><span class="n">rank</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">ranking_metric</span><span class="si">}</span><span class="s2">*.pdb&quot;</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">pb_validated_ligand_output_filepaths</span> <span class="o">=</span> <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="n">ligand_output_filepath</span><span class="p">)</span>
            <span class="n">protein_output_filepaths</span> <span class="o">=</span> <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="n">protein_output_filepath</span><span class="p">)</span>
            <span class="k">assert</span> <span class="p">(</span>
                <span class="nb">len</span><span class="p">(</span><span class="n">pb_validated_ligand_output_filepaths</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span>
            <span class="p">),</span> <span class="sa">f</span><span class="s2">&quot;Expected one PoseBusters-validated ligand structure file to match </span><span class="si">{</span><span class="n">ligand_output_filepath</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;.sdf&#39;</span><span class="p">,</span><span class="w"> </span><span class="sa">f</span><span class="s1">&#39;_pbvalid=*.sdf&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">, but found </span><span class="si">{</span><span class="n">pb_validated_ligand_output_filepaths</span><span class="si">}</span><span class="s2">.&quot;</span>
            <span class="k">assert</span> <span class="p">(</span>
                <span class="nb">len</span><span class="p">(</span><span class="n">protein_output_filepaths</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span>
            <span class="p">),</span> <span class="sa">f</span><span class="s2">&quot;Expected one protein structure file to match </span><span class="si">{</span><span class="n">protein_output_filepath</span><span class="si">}</span><span class="s2">, but found </span><span class="si">{</span><span class="n">protein_output_filepaths</span><span class="si">}</span><span class="s2">.&quot;</span>
            <span class="n">output_ligand_filepaths</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pb_validated_ligand_output_filepaths</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
            <span class="n">output_protein_filepaths</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">protein_output_filepaths</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># NOTE: we use the `dock` mode here since with each method we implicitly perform cognate (e.g., apo or ab initio) docking,</span>
        <span class="c1"># yet for arbitrary input data we do not have access to the ground-truth ligand structures in SDF format</span>
        <span class="n">buster</span> <span class="o">=</span> <span class="n">PoseBusters</span><span class="p">(</span><span class="n">config</span><span class="o">=</span><span class="s2">&quot;dock&quot;</span><span class="p">,</span> <span class="n">top_n</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
        <span class="n">output_ligand_filepaths</span><span class="p">,</span> <span class="n">output_protein_filepaths</span> <span class="o">=</span> <span class="p">[],</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">index</span><span class="p">,</span> <span class="p">(</span>
            <span class="n">rank</span><span class="p">,</span>
            <span class="p">(</span>
                <span class="n">method</span><span class="p">,</span>
                <span class="n">protein_filepath</span><span class="p">,</span>
                <span class="n">ligand_filepath</span><span class="p">,</span>
                <span class="n">ranking_metric_value</span><span class="p">,</span>
            <span class="p">),</span>
        <span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">ranked_predictions</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
            <span class="c1"># only save the top-N predictions as specified</span>
            <span class="k">if</span> <span class="n">cfg</span><span class="o">.</span><span class="n">export_top_n</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">if</span> <span class="p">(</span><span class="n">index</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">&gt;</span> <span class="n">cfg</span><span class="o">.</span><span class="n">export_top_n</span><span class="p">:</span>
                    <span class="k">break</span>
            <span class="n">ligand_plddt_match</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;plddt(\d+\.\d+)&quot;</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">ligand_filepath</span><span class="p">))</span>
            <span class="n">ligand_affinity_match</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span>
                <span class="sa">r</span><span class="s2">&quot;affinity(\d+\.\d+)&quot;</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">ligand_filepath</span><span class="p">)</span>
            <span class="p">)</span>
            <span class="n">ligand_plddt_value</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">ligand_plddt_match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">))</span> <span class="k">if</span> <span class="n">ligand_plddt_match</span> <span class="k">else</span> <span class="kc">None</span>
            <span class="n">ligand_affinity_value</span> <span class="o">=</span> <span class="p">(</span>
                <span class="nb">float</span><span class="p">(</span><span class="n">ligand_affinity_match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">))</span> <span class="k">if</span> <span class="n">ligand_affinity_match</span> <span class="k">else</span> <span class="kc">None</span>
            <span class="p">)</span>
            <span class="n">ligand_plddt_postfix</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;_plddt</span><span class="si">{</span><span class="n">ligand_plddt_value</span><span class="si">:</span><span class="s2">.7f</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">if</span> <span class="n">ligand_plddt_value</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span>
            <span class="n">ligand_affinity_postfix</span> <span class="o">=</span> <span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;_affinity</span><span class="si">{</span><span class="n">ligand_affinity_value</span><span class="si">:</span><span class="s2">.7f</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">if</span> <span class="n">ligand_affinity_value</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span>
            <span class="p">)</span>
            <span class="n">ligand_output_filepath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                <span class="n">cfg</span><span class="o">.</span><span class="n">output_dir</span><span class="p">,</span>
                <span class="n">name</span> <span class="o">+</span> <span class="n">ligand_relaxed_postfix</span><span class="p">,</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">method</span><span class="si">}</span><span class="s2">_rank</span><span class="si">{</span><span class="n">rank</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">ranking_metric</span><span class="si">}{</span><span class="n">ranking_metric_value</span><span class="si">:</span><span class="s2">.2e</span><span class="si">}{</span><span class="n">ligand_plddt_postfix</span><span class="si">}{</span><span class="n">ligand_affinity_postfix</span><span class="si">}{</span><span class="n">ligand_relaxed_postfix</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="mi">0</span><span class="w"> </span><span class="o">&lt;</span><span class="w"> </span><span class="nb">len</span><span class="p">(</span><span class="n">relaxation_success_list</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;=</span><span class="w"> </span><span class="nb">len</span><span class="p">(</span><span class="n">ranked_predictions</span><span class="p">)</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="n">relaxation_success_list</span><span class="p">[</span><span class="n">index</span><span class="p">]</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;&#39;</span><span class="si">}</span><span class="s2">.sdf&quot;</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">protein_output_filepath</span> <span class="o">=</span> <span class="n">protein_filepath</span>
            <span class="k">if</span> <span class="p">(</span>
                <span class="n">protein_filepath</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
                <span class="ow">and</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">protein_filepath</span><span class="p">)</span> <span class="o">!=</span> <span class="s2">&quot;ligand_only.pdb&quot;</span>
            <span class="p">):</span>
                <span class="n">protein_plddt_match</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span>
                    <span class="sa">r</span><span class="s2">&quot;plddt(\d+\.\d+)&quot;</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">protein_filepath</span><span class="p">)</span>
                <span class="p">)</span>
                <span class="n">protein_affinity_match</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span>
                    <span class="sa">r</span><span class="s2">&quot;affinity(\d+\.\d+)&quot;</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">protein_filepath</span><span class="p">)</span>
                <span class="p">)</span>
                <span class="n">protein_plddt_value</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="nb">float</span><span class="p">(</span><span class="n">protein_plddt_match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">))</span> <span class="k">if</span> <span class="n">protein_plddt_match</span> <span class="k">else</span> <span class="kc">None</span>
                <span class="p">)</span>
                <span class="n">protein_affinity_value</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="nb">float</span><span class="p">(</span><span class="n">protein_affinity_match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">))</span> <span class="k">if</span> <span class="n">protein_affinity_match</span> <span class="k">else</span> <span class="kc">None</span>
                <span class="p">)</span>
                <span class="n">protein_plddt_postfix</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;_plddt</span><span class="si">{</span><span class="n">protein_plddt_value</span><span class="si">:</span><span class="s2">.7f</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">if</span> <span class="n">protein_plddt_value</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span>
                <span class="p">)</span>
                <span class="n">protein_affinity_postfix</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;_affinity</span><span class="si">{</span><span class="n">protein_affinity_value</span><span class="si">:</span><span class="s2">.7f</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">if</span> <span class="n">protein_affinity_value</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span>
                <span class="p">)</span>
                <span class="n">protein_output_filepath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                    <span class="n">cfg</span><span class="o">.</span><span class="n">output_dir</span><span class="p">,</span>
                    <span class="n">name</span> <span class="o">+</span> <span class="n">ligand_relaxed_postfix</span><span class="p">,</span>
                    <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">method</span><span class="si">}</span><span class="s2">_rank</span><span class="si">{</span><span class="n">rank</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">ranking_metric</span><span class="si">}{</span><span class="n">ranking_metric_value</span><span class="si">:</span><span class="s2">.2e</span><span class="si">}{</span><span class="n">protein_plddt_postfix</span><span class="si">}{</span><span class="n">protein_affinity_postfix</span><span class="si">}{</span><span class="n">protein_relaxed_postfix</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="mi">0</span><span class="w"> </span><span class="o">&lt;</span><span class="w"> </span><span class="nb">len</span><span class="p">(</span><span class="n">relaxation_success_list</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;=</span><span class="w"> </span><span class="nb">len</span><span class="p">(</span><span class="n">ranked_predictions</span><span class="p">)</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="n">relaxation_success_list</span><span class="p">[</span><span class="n">index</span><span class="p">]</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;&#39;</span><span class="si">}</span><span class="s2">.pdb&quot;</span><span class="p">,</span>
                <span class="p">)</span>

            <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">ligand_output_filepath</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;.sdf&quot;</span><span class="p">,</span> <span class="s2">&quot;_bust_results.csv&quot;</span><span class="p">)):</span>
                <span class="c1"># skip PoseBusters validation if the results have previously been saved</span>
                <span class="n">mol_table</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
                    <span class="p">{</span>
                        <span class="s2">&quot;mol_pred&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">ligand_filepath</span><span class="p">],</span>
                        <span class="s2">&quot;mol_true&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
                        <span class="s2">&quot;mol_cond&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">protein_filepath</span><span class="p">],</span>
                    <span class="p">}</span>
                <span class="p">)</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">bust_results</span> <span class="o">=</span> <span class="n">buster</span><span class="o">.</span><span class="n">bust_table</span><span class="p">(</span><span class="n">mol_table</span><span class="p">,</span> <span class="n">full_report</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
                    <span class="n">bust_results</span><span class="p">[</span><span class="s2">&quot;valid&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
                        <span class="n">bust_results</span><span class="p">[</span><span class="s2">&quot;mol_pred_loaded&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span>
                        <span class="o">&amp;</span> <span class="n">bust_results</span><span class="p">[</span><span class="s2">&quot;mol_cond_loaded&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span>
                        <span class="o">&amp;</span> <span class="n">bust_results</span><span class="p">[</span><span class="s2">&quot;sanitization&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span>
                        <span class="o">&amp;</span> <span class="n">bust_results</span><span class="p">[</span><span class="s2">&quot;all_atoms_connected&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span>
                        <span class="o">&amp;</span> <span class="n">bust_results</span><span class="p">[</span><span class="s2">&quot;bond_lengths&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span>
                        <span class="o">&amp;</span> <span class="n">bust_results</span><span class="p">[</span><span class="s2">&quot;bond_angles&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span>
                        <span class="o">&amp;</span> <span class="n">bust_results</span><span class="p">[</span><span class="s2">&quot;internal_steric_clash&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span>
                        <span class="o">&amp;</span> <span class="n">bust_results</span><span class="p">[</span><span class="s2">&quot;aromatic_ring_flatness&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span>
                        <span class="o">&amp;</span> <span class="n">bust_results</span><span class="p">[</span><span class="s2">&quot;double_bond_flatness&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span>
                        <span class="o">&amp;</span> <span class="n">bust_results</span><span class="p">[</span><span class="s2">&quot;internal_energy&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span>
                        <span class="o">&amp;</span> <span class="n">bust_results</span><span class="p">[</span><span class="s2">&quot;protein-ligand_maximum_distance&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span>
                        <span class="o">&amp;</span> <span class="n">bust_results</span><span class="p">[</span><span class="s2">&quot;minimum_distance_to_protein&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span>
                        <span class="o">&amp;</span> <span class="n">bust_results</span><span class="p">[</span><span class="s2">&quot;minimum_distance_to_organic_cofactors&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span>
                        <span class="o">&amp;</span> <span class="n">bust_results</span><span class="p">[</span><span class="s2">&quot;minimum_distance_to_inorganic_cofactors&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span>
                        <span class="o">&amp;</span> <span class="n">bust_results</span><span class="p">[</span><span class="s2">&quot;minimum_distance_to_waters&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span>
                        <span class="o">&amp;</span> <span class="n">bust_results</span><span class="p">[</span><span class="s2">&quot;volume_overlap_with_protein&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span>
                        <span class="o">&amp;</span> <span class="n">bust_results</span><span class="p">[</span><span class="s2">&quot;volume_overlap_with_organic_cofactors&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span>
                        <span class="o">&amp;</span> <span class="n">bust_results</span><span class="p">[</span><span class="s2">&quot;volume_overlap_with_inorganic_cofactors&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span>
                        <span class="o">&amp;</span> <span class="n">bust_results</span><span class="p">[</span><span class="s2">&quot;volume_overlap_with_waters&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span>
                    <span class="p">)</span>
                    <span class="n">relaxed_mol_is_valid</span> <span class="o">=</span> <span class="n">bust_results</span><span class="p">[</span><span class="s2">&quot;valid&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">item</span><span class="p">()</span>
                    <span class="n">bust_results</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span>
                        <span class="n">ligand_output_filepath</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;.sdf&quot;</span><span class="p">,</span> <span class="s2">&quot;_bust_results.csv&quot;</span><span class="p">),</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span>
                    <span class="p">)</span>
                <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                    <span class="n">relaxed_mol_is_valid</span> <span class="o">=</span> <span class="kc">False</span>
                    <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;Failed to PoseBusters-validate ligand structure </span><span class="si">{</span><span class="n">ligand_filepath</span><span class="si">}</span><span class="s2"> from method </span><span class="si">{</span><span class="n">method</span><span class="si">}</span><span class="s2"> due to: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">. Skipping validation...&quot;</span>
                    <span class="p">)</span>
                <span class="k">if</span> <span class="n">cfg</span><span class="o">.</span><span class="n">superligand_inputs</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;Should not overwrite CASP16 top-5 predictions (i.e., `</span><span class="si">{</span><span class="n">ligand_output_filepath</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;.sdf&#39;</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;_bust_results.csv&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">`) once manually finalized.&quot;</span>
                    <span class="p">)</span>
                <span class="n">shutil</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span>
                    <span class="n">ligand_filepath</span><span class="p">,</span>
                    <span class="n">ligand_output_filepath</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;.sdf&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;_pbvalid=</span><span class="si">{</span><span class="n">relaxed_mol_is_valid</span><span class="si">}</span><span class="s2">.sdf&quot;</span><span class="p">),</span>
                <span class="p">)</span>
                <span class="n">shutil</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">protein_filepath</span><span class="p">,</span> <span class="n">protein_output_filepath</span><span class="p">)</span>

            <span class="k">if</span> <span class="p">(</span>
                <span class="n">relax_complex</span>
                <span class="ow">and</span> <span class="mi">0</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">relaxation_success_list</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="nb">len</span><span class="p">(</span><span class="n">ranked_predictions</span><span class="p">)</span>
                <span class="ow">and</span> <span class="n">relaxation_success_list</span><span class="p">[</span><span class="n">index</span><span class="p">]</span>
            <span class="p">):</span>
                <span class="c1"># if applicable, save a copy of the unrelaxed ligand structure as well</span>
                <span class="n">unrelaxed_ligand_filepath</span> <span class="o">=</span> <span class="n">ligand_filepath</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span>
                    <span class="s2">&quot;_ensemble_relaxed.sdf&quot;</span><span class="p">,</span> <span class="s2">&quot;.sdf&quot;</span>
                <span class="p">)</span>
                <span class="n">unrelaxed_protein_filepath</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="n">protein_filepath</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;_ensemble_protein_relaxed.sdf&quot;</span><span class="p">,</span> <span class="s2">&quot;.sdf&quot;</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_protein</span>
                    <span class="k">else</span> <span class="n">protein_filepath</span>
                <span class="p">)</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span>
                    <span class="n">ligand_output_filepath</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;_relaxed&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span>
                        <span class="s2">&quot;.sdf&quot;</span><span class="p">,</span> <span class="s2">&quot;_bust_results.csv&quot;</span>
                    <span class="p">)</span>
                <span class="p">):</span>
                    <span class="c1"># skip (relaxed) PoseBusters validation if the results have previously been saved</span>
                    <span class="n">mol_table</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
                        <span class="p">{</span>
                            <span class="s2">&quot;mol_pred&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">unrelaxed_ligand_filepath</span><span class="p">],</span>
                            <span class="s2">&quot;mol_true&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
                            <span class="s2">&quot;mol_cond&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">unrelaxed_protein_filepath</span><span class="p">],</span>
                        <span class="p">}</span>
                    <span class="p">)</span>
                    <span class="k">try</span><span class="p">:</span>
                        <span class="n">bust_results</span> <span class="o">=</span> <span class="n">buster</span><span class="o">.</span><span class="n">bust_table</span><span class="p">(</span><span class="n">mol_table</span><span class="p">,</span> <span class="n">full_report</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
                        <span class="n">bust_results</span><span class="p">[</span><span class="s2">&quot;valid&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
                            <span class="n">bust_results</span><span class="p">[</span><span class="s2">&quot;mol_pred_loaded&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span>
                            <span class="o">&amp;</span> <span class="n">bust_results</span><span class="p">[</span><span class="s2">&quot;mol_cond_loaded&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span>
                            <span class="o">&amp;</span> <span class="n">bust_results</span><span class="p">[</span><span class="s2">&quot;sanitization&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span>
                            <span class="o">&amp;</span> <span class="n">bust_results</span><span class="p">[</span><span class="s2">&quot;all_atoms_connected&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span>
                            <span class="o">&amp;</span> <span class="n">bust_results</span><span class="p">[</span><span class="s2">&quot;bond_lengths&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span>
                            <span class="o">&amp;</span> <span class="n">bust_results</span><span class="p">[</span><span class="s2">&quot;bond_angles&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span>
                            <span class="o">&amp;</span> <span class="n">bust_results</span><span class="p">[</span><span class="s2">&quot;internal_steric_clash&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span>
                            <span class="o">&amp;</span> <span class="n">bust_results</span><span class="p">[</span><span class="s2">&quot;aromatic_ring_flatness&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span>
                            <span class="o">&amp;</span> <span class="n">bust_results</span><span class="p">[</span><span class="s2">&quot;double_bond_flatness&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span>
                            <span class="o">&amp;</span> <span class="n">bust_results</span><span class="p">[</span><span class="s2">&quot;internal_energy&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span>
                            <span class="o">&amp;</span> <span class="n">bust_results</span><span class="p">[</span><span class="s2">&quot;protein-ligand_maximum_distance&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span>
                            <span class="o">&amp;</span> <span class="n">bust_results</span><span class="p">[</span><span class="s2">&quot;minimum_distance_to_protein&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span>
                            <span class="o">&amp;</span> <span class="n">bust_results</span><span class="p">[</span><span class="s2">&quot;minimum_distance_to_organic_cofactors&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span>
                            <span class="o">&amp;</span> <span class="n">bust_results</span><span class="p">[</span><span class="s2">&quot;minimum_distance_to_inorganic_cofactors&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span>
                            <span class="o">&amp;</span> <span class="n">bust_results</span><span class="p">[</span><span class="s2">&quot;minimum_distance_to_waters&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span>
                            <span class="o">&amp;</span> <span class="n">bust_results</span><span class="p">[</span><span class="s2">&quot;volume_overlap_with_protein&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span>
                            <span class="o">&amp;</span> <span class="n">bust_results</span><span class="p">[</span><span class="s2">&quot;volume_overlap_with_organic_cofactors&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span>
                            <span class="o">&amp;</span> <span class="n">bust_results</span><span class="p">[</span><span class="s2">&quot;volume_overlap_with_inorganic_cofactors&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span>
                            <span class="o">&amp;</span> <span class="n">bust_results</span><span class="p">[</span><span class="s2">&quot;volume_overlap_with_waters&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span>
                        <span class="p">)</span>
                        <span class="n">unrelaxed_mol_is_valid</span> <span class="o">=</span> <span class="n">bust_results</span><span class="p">[</span><span class="s2">&quot;valid&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">item</span><span class="p">()</span>
                        <span class="n">bust_results</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span>
                            <span class="n">ligand_output_filepath</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;_relaxed&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span>
                                <span class="s2">&quot;.sdf&quot;</span><span class="p">,</span> <span class="s2">&quot;_bust_results.csv&quot;</span>
                            <span class="p">),</span>
                            <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                        <span class="p">)</span>
                    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                        <span class="n">unrelaxed_mol_is_valid</span> <span class="o">=</span> <span class="kc">False</span>
                        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                            <span class="sa">f</span><span class="s2">&quot;Failed to PoseBusters-validate ligand structure </span><span class="si">{</span><span class="n">unrelaxed_ligand_filepath</span><span class="si">}</span><span class="s2"> from method </span><span class="si">{</span><span class="n">method</span><span class="si">}</span><span class="s2"> due to: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">. Skipping validation...&quot;</span>
                        <span class="p">)</span>
                    <span class="k">if</span> <span class="n">cfg</span><span class="o">.</span><span class="n">superligand_inputs</span><span class="p">:</span>
                        <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span>
                            <span class="s2">&quot;Should not overwrite CASP16 top-5 predictions once manually finalized.&quot;</span>
                        <span class="p">)</span>
                    <span class="n">shutil</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span>
                        <span class="n">unrelaxed_ligand_filepath</span><span class="p">,</span>
                        <span class="n">ligand_output_filepath</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;_relaxed&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span>
                            <span class="s2">&quot;.sdf&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;_pbvalid=</span><span class="si">{</span><span class="n">unrelaxed_mol_is_valid</span><span class="si">}</span><span class="s2">.sdf&quot;</span>
                        <span class="p">),</span>
                    <span class="p">)</span>
                <span class="k">if</span> <span class="p">(</span>
                    <span class="n">cfg</span><span class="o">.</span><span class="n">relax_protein</span>
                    <span class="ow">and</span> <span class="mi">0</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">relaxation_success_list</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="nb">len</span><span class="p">(</span><span class="n">ranked_predictions</span><span class="p">)</span>
                    <span class="ow">and</span> <span class="n">relaxation_success_list</span><span class="p">[</span><span class="n">index</span><span class="p">]</span>
                <span class="p">):</span>
                    <span class="c1"># if applicable, save a copy of the unrelaxed protein structure as well</span>
                    <span class="k">if</span> <span class="n">cfg</span><span class="o">.</span><span class="n">superligand_inputs</span><span class="p">:</span>
                        <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span>
                            <span class="s2">&quot;Should not overwrite CASP16 top-5 predictions once manually finalized.&quot;</span>
                        <span class="p">)</span>
                    <span class="n">shutil</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span>
                        <span class="n">unrelaxed_protein_filepath</span><span class="p">,</span>
                        <span class="n">protein_output_filepath</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;_relaxed&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">),</span>
                    <span class="p">)</span>

            <span class="n">pb_validated_ligand_output_filepaths</span> <span class="o">=</span> <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span>
                <span class="n">ligand_output_filepath</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;.sdf&quot;</span><span class="p">,</span> <span class="s2">&quot;_pbvalid=*.sdf&quot;</span><span class="p">)</span>
            <span class="p">)</span>
            <span class="k">assert</span> <span class="p">(</span>
                <span class="nb">len</span><span class="p">(</span><span class="n">pb_validated_ligand_output_filepaths</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span>
            <span class="p">),</span> <span class="sa">f</span><span class="s2">&quot;Expected one PoseBusters-validated ligand structure file to match </span><span class="si">{</span><span class="n">ligand_output_filepath</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;.sdf&#39;</span><span class="p">,</span><span class="w"> </span><span class="sa">f</span><span class="s1">&#39;_pbvalid=*.sdf&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">, but found </span><span class="si">{</span><span class="n">pb_validated_ligand_output_filepaths</span><span class="si">}</span><span class="s2">.&quot;</span>
            <span class="n">output_ligand_filepaths</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pb_validated_ligand_output_filepaths</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
            <span class="n">output_protein_filepaths</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">protein_output_filepath</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">cfg</span><span class="o">.</span><span class="n">export_file_format</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="c1"># NOTE: we rely on FlowDock&#39;s precomputed auxiliary estimation CSV files to assign plDDT and affinity scores</span>
        <span class="n">aux_estimation_csv_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
            <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">ligand_output_filepath</span><span class="p">),</span> <span class="s2">&quot;auxiliary_estimation.csv&quot;</span>
        <span class="p">)</span>
        <span class="n">aux_estimation_df</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">aux_estimation_csv_path</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">aux_estimation_csv_path</span><span class="p">)</span>
            <span class="k">else</span> <span class="kc">None</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="n">aux_estimation_df</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">assert</span> <span class="n">aux_estimation_df</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span> <span class="o">==</span> <span class="p">[</span>
                <span class="s2">&quot;sample_id&quot;</span><span class="p">,</span>
                <span class="s2">&quot;rank&quot;</span><span class="p">,</span>
                <span class="s2">&quot;plddt_ligs&quot;</span><span class="p">,</span>
                <span class="s2">&quot;affinity_ligs&quot;</span><span class="p">,</span>
            <span class="p">],</span> <span class="s2">&quot;The `sample_id`, `rank`, `plddt_ligs`, and `affinity_ligs` columns must be contained in the auxiliary estimation CSV file.&quot;</span>
        <span class="k">elif</span> <span class="n">cfg</span><span class="o">.</span><span class="n">export_file_format</span> <span class="o">==</span> <span class="s2">&quot;casp16&quot;</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;CASP16 ensemble generation is almost complete. However, the auxiliary estimation CSV file was not found at </span><span class="si">{</span><span class="n">aux_estimation_csv_path</span><span class="si">}</span><span class="s2">. Please re-run this script after finalizing (e.g., relaxed) ensemble predictions within </span><span class="si">{</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">ligand_output_filepath</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;_relaxed&#39;</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2"> and then (ensemble-)running FlowDock inference with `flowdock_auxiliary_estimation_only=true` and `flowdock_auxiliary_estimation_input_dir=</span><span class="si">{</span><span class="n">Path</span><span class="p">(</span><span class="n">ligand_output_filepath</span><span class="p">)</span><span class="o">.</span><span class="n">parent</span><span class="o">.</span><span class="n">parent</span><span class="si">}</span><span class="s2">`.&quot;</span>
            <span class="p">)</span>
            <span class="k">return</span>

        <span class="k">if</span> <span class="n">aux_estimation_df</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="c1"># NOTE: we assign to each ligand FlowDock&#39;s corresponding e.g., ligand plDDT score,</span>
            <span class="c1"># even if several different methods are among the top-N predictions</span>
            <span class="n">aux_estimation_df</span><span class="p">[</span><span class="s2">&quot;plddt_ligs&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">aux_estimation_df</span><span class="p">[</span><span class="s2">&quot;plddt_ligs&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
                <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">ast</span><span class="o">.</span><span class="n">literal_eval</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
            <span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">aux_estimation_df</span><span class="p">[</span><span class="s2">&quot;affinity_ligs&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isnull</span><span class="p">()</span><span class="o">.</span><span class="n">any</span><span class="p">():</span>
                <span class="n">aux_estimation_df</span><span class="p">[</span><span class="s2">&quot;affinity_ligs&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">aux_estimation_df</span><span class="p">[</span><span class="s2">&quot;affinity_ligs&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
                    <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">ast</span><span class="o">.</span><span class="n">literal_eval</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
                <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Auxiliary estimation CSV file not found at </span><span class="si">{</span><span class="n">aux_estimation_csv_path</span><span class="si">}</span><span class="s2">. Ligand confidence and affinity scores will not be included in the CASP submission file.&quot;</span>
            <span class="p">)</span>

        <span class="c1"># NOTE: relaxed ligand (and potentially protein) files are used for CASP submission when `relax_complex=True`</span>
        <span class="n">pdb_header</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;PFRMAT LG</span><span class="se">\n</span><span class="s2">TARGET </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="se">\n</span><span class="s2">AUTHOR </span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">casp_registration_code</span><span class="si">}</span><span class="se">\n</span><span class="s2">METHOD </span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">casp_method</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="n">sdf_header</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;PFRMAT LG</span><span class="se">\n</span><span class="s2">TARGET </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="se">\n</span><span class="s2">AUTHOR </span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">casp_registration_code</span><span class="si">}</span><span class="se">\n</span><span class="s2">METHOD </span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">casp_method</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>

        <span class="n">only_affinity_requested</span> <span class="o">=</span> <span class="n">ligand_tasks</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">ligand_tasks</span> <span class="o">==</span> <span class="s2">&quot;A&quot;</span>
        <span class="k">if</span> <span class="n">only_affinity_requested</span><span class="p">:</span>
            <span class="k">assert</span> <span class="p">(</span>
                <span class="n">cfg</span><span class="o">.</span><span class="n">export_file_format</span> <span class="o">==</span> <span class="s2">&quot;casp16&quot;</span>
            <span class="p">),</span> <span class="s2">&quot;Only CASP16 format is supported for affinity-only predictions.&quot;</span>
            <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">ligand_output_filepath</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">output_ligand_filepaths</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
                <span class="n">output_file</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                        <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">ligand_output_filepath</span><span class="p">),</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">LG</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">casp_author</span><span class="si">}</span><span class="s2">&quot;</span>
                    <span class="p">)</span>
                    <span class="k">if</span> <span class="n">cfg</span><span class="o">.</span><span class="n">combine_casp_output_files</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">cfg</span><span class="o">.</span><span class="n">superligand_inputs</span>
                    <span class="k">else</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                        <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">ligand_output_filepath</span><span class="p">),</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">LG</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">casp_author</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">&quot;</span>
                    <span class="p">)</span>
                <span class="p">)</span>
                <span class="n">method</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">ligand_output_filepath</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">export_ligands_in_casp16_format</span><span class="p">(</span>
                    <span class="p">[</span><span class="n">ligand_output_filepath</span><span class="p">],</span>
                    <span class="n">output_file</span><span class="p">,</span>
                    <span class="n">sdf_header</span><span class="p">,</span>
                    <span class="n">method</span><span class="p">,</span>
                    <span class="n">append</span><span class="o">=</span><span class="n">cfg</span><span class="o">.</span><span class="n">combine_casp_output_files</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">cfg</span><span class="o">.</span><span class="n">superligand_inputs</span><span class="p">,</span>
                    <span class="n">model_index</span><span class="o">=</span><span class="n">i</span><span class="p">,</span>
                    <span class="n">only_affinity_requested</span><span class="o">=</span><span class="n">only_affinity_requested</span><span class="p">,</span>
                    <span class="n">ligand_numbers</span><span class="o">=</span><span class="n">ligand_numbers</span><span class="p">,</span>
                    <span class="n">ligand_names</span><span class="o">=</span><span class="n">ligand_names</span><span class="p">,</span>
                    <span class="n">ligand_tasks</span><span class="o">=</span><span class="n">ligand_tasks</span><span class="p">,</span>
                    <span class="n">aux_estimation_df</span><span class="o">=</span><span class="n">aux_estimation_df</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="k">return</span>

        <span class="n">output_protein_filepaths</span> <span class="o">=</span> <span class="n">assign_reference_residue_b_factors</span><span class="p">(</span>
            <span class="n">output_protein_filepaths</span><span class="p">,</span> <span class="n">protein_input_filepath</span>
        <span class="p">)</span>

        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">protein_output_filepath</span><span class="p">,</span> <span class="n">ligand_output_filepath</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span>
            <span class="nb">zip</span><span class="p">(</span><span class="n">output_protein_filepaths</span><span class="p">,</span> <span class="n">output_ligand_filepaths</span><span class="p">),</span> <span class="n">start</span><span class="o">=</span><span class="mi">1</span>
        <span class="p">):</span>
            <span class="n">output_file</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                    <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">ligand_output_filepath</span><span class="p">),</span>
                    <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">LG</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">casp_author</span><span class="si">}{</span><span class="s1">&#39;&#39;</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">cfg</span><span class="o">.</span><span class="n">export_file_format</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s1">&#39;casp16&#39;</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="sa">f</span><span class="s1">&#39;_</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s1">&#39;</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
                <span class="p">)</span>
                <span class="k">if</span> <span class="n">cfg</span><span class="o">.</span><span class="n">combine_casp_output_files</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">cfg</span><span class="o">.</span><span class="n">superligand_inputs</span>
                <span class="k">else</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                    <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">ligand_output_filepath</span><span class="p">),</span>
                    <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">LG</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">casp_author</span><span class="si">}{</span><span class="s1">&#39;&#39;</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">cfg</span><span class="o">.</span><span class="n">export_file_format</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s1">&#39;casp16&#39;</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;_protein&#39;</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="p">)</span>
            <span class="n">method</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">ligand_output_filepath</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">gap_insertion_point</span> <span class="o">=</span> <span class="p">(</span>
                <span class="c1"># NOTE: for target `T1124` from CASP15, we have to insert a one-step gap starting at</span>
                <span class="c1"># residue `243` in chain `B` for methods that predict holo protein PDB files ab initio</span>
                <span class="c1"># to properly score these predictions</span>
                <span class="s2">&quot;B:243&quot;</span>
                <span class="k">if</span> <span class="n">name</span> <span class="o">==</span> <span class="s2">&quot;T1124&quot;</span> <span class="ow">and</span> <span class="n">method</span> <span class="ow">in</span> <span class="n">METHODS_PREDICTING_HOLO_PROTEIN_AB_INITIO</span>
                <span class="k">else</span> <span class="kc">None</span>
            <span class="p">)</span>
            <span class="n">export_proteins_in_casp_format</span><span class="p">(</span>
                <span class="p">[</span><span class="n">protein_output_filepath</span><span class="p">],</span>
                <span class="n">output_file</span><span class="p">,</span>
                <span class="n">pdb_header</span><span class="p">,</span>
                <span class="n">append</span><span class="o">=</span><span class="n">cfg</span><span class="o">.</span><span class="n">combine_casp_output_files</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">cfg</span><span class="o">.</span><span class="n">superligand_inputs</span><span class="p">,</span>
                <span class="n">export_casp15_format</span><span class="o">=</span><span class="n">cfg</span><span class="o">.</span><span class="n">export_file_format</span> <span class="o">==</span> <span class="s2">&quot;casp15&quot;</span><span class="p">,</span>
                <span class="n">superligand_inputs</span><span class="o">=</span><span class="n">cfg</span><span class="o">.</span><span class="n">superligand_inputs</span><span class="p">,</span>
                <span class="n">model_index</span><span class="o">=</span><span class="n">i</span><span class="p">,</span>
                <span class="n">gap_insertion_point</span><span class="o">=</span><span class="n">gap_insertion_point</span><span class="p">,</span>
            <span class="p">)</span>

            <span class="n">output_file</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                    <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">ligand_output_filepath</span><span class="p">),</span>
                    <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">LG</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">casp_author</span><span class="si">}{</span><span class="s1">&#39;&#39;</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">cfg</span><span class="o">.</span><span class="n">export_file_format</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s1">&#39;casp16&#39;</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="sa">f</span><span class="s1">&#39;_</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s1">&#39;</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
                <span class="p">)</span>
                <span class="k">if</span> <span class="n">cfg</span><span class="o">.</span><span class="n">combine_casp_output_files</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">cfg</span><span class="o">.</span><span class="n">superligand_inputs</span>
                <span class="k">else</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                    <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">ligand_output_filepath</span><span class="p">),</span>
                    <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">LG</span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">casp_author</span><span class="si">}{</span><span class="s1">&#39;&#39;</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">cfg</span><span class="o">.</span><span class="n">export_file_format</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s1">&#39;casp16&#39;</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;_ligand&#39;</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="p">)</span>
            <span class="k">if</span> <span class="n">cfg</span><span class="o">.</span><span class="n">export_file_format</span> <span class="o">==</span> <span class="s2">&quot;casp15&quot;</span><span class="p">:</span>
                <span class="n">export_ligands_in_casp15_format</span><span class="p">(</span>
                    <span class="p">[</span><span class="n">ligand_output_filepath</span><span class="p">],</span>
                    <span class="n">output_file</span><span class="p">,</span>
                    <span class="n">sdf_header</span><span class="p">,</span>
                    <span class="n">method</span><span class="p">,</span>
                    <span class="n">append</span><span class="o">=</span><span class="n">cfg</span><span class="o">.</span><span class="n">combine_casp_output_files</span> <span class="ow">or</span> <span class="n">cfg</span><span class="o">.</span><span class="n">superligand_inputs</span><span class="p">,</span>
                    <span class="n">model_index</span><span class="o">=</span><span class="n">i</span><span class="p">,</span>
                    <span class="n">ligand_numbers</span><span class="o">=</span><span class="n">ligand_numbers</span><span class="p">,</span>
                    <span class="n">ligand_names</span><span class="o">=</span><span class="n">ligand_names</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">export_ligands_in_casp16_format</span><span class="p">(</span>
                    <span class="p">[</span><span class="n">ligand_output_filepath</span><span class="p">],</span>
                    <span class="n">output_file</span><span class="p">,</span>
                    <span class="n">sdf_header</span><span class="p">,</span>
                    <span class="n">method</span><span class="p">,</span>
                    <span class="n">append</span><span class="o">=</span><span class="n">cfg</span><span class="o">.</span><span class="n">combine_casp_output_files</span> <span class="ow">or</span> <span class="n">cfg</span><span class="o">.</span><span class="n">superligand_inputs</span><span class="p">,</span>
                    <span class="n">model_index</span><span class="o">=</span><span class="n">i</span><span class="p">,</span>
                    <span class="n">ligand_numbers</span><span class="o">=</span><span class="n">ligand_numbers</span><span class="p">,</span>
                    <span class="n">ligand_names</span><span class="o">=</span><span class="n">ligand_names</span><span class="p">,</span>
                    <span class="n">ligand_tasks</span><span class="o">=</span><span class="n">ligand_tasks</span><span class="p">,</span>
                    <span class="n">aux_estimation_df</span><span class="o">=</span><span class="n">aux_estimation_df</span><span class="p">,</span>
                <span class="p">)</span></div>



<div class="viewcode-block" id="main">
<a class="viewcode-back" href="../../../modules/multicom_ligand.ensemble_generation.html#multicom_ligand.models.ensemble_generation.main">[docs]</a>
<span class="nd">@hydra</span><span class="o">.</span><span class="n">main</span><span class="p">(</span>
    <span class="n">version_base</span><span class="o">=</span><span class="s2">&quot;1.3&quot;</span><span class="p">,</span>
    <span class="n">config_path</span><span class="o">=</span><span class="s2">&quot;../../configs/model&quot;</span><span class="p">,</span>
    <span class="n">config_name</span><span class="o">=</span><span class="s2">&quot;ensemble_generation.yaml&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="k">def</span><span class="w"> </span><span class="nf">main</span><span class="p">(</span><span class="n">cfg</span><span class="p">:</span> <span class="n">DictConfig</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Generate predictions for a protein-ligand target pair using an ensemble</span>
<span class="sd">    of methods.&quot;&quot;&quot;</span>
    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">cfg</span><span class="o">.</span><span class="n">temp_protein_dir</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">input_csv_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">cfg</span><span class="o">.</span><span class="n">input_csv_filepath</span><span class="p">)</span>
    <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">input_csv_df</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span>
        <span class="n">input_csv_df</span>
    <span class="p">),</span> <span class="s2">&quot;Names in input CSV must all be unique.&quot;</span>
    <span class="k">assert</span> <span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_ranking_method</span> <span class="ow">in</span> <span class="p">[</span>
        <span class="s2">&quot;consensus&quot;</span><span class="p">,</span>
        <span class="s2">&quot;ff&quot;</span><span class="p">,</span>
    <span class="p">],</span> <span class="s2">&quot;`ensemble_ranking_method` must be either `consensus` or `ff`.&quot;</span>

    <span class="c1"># ensure the input CSV contains columns for ligand numbers and names</span>
    <span class="k">if</span> <span class="s2">&quot;ligand_numbers&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">input_csv_df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">input_csv_df</span><span class="p">[</span><span class="s2">&quot;ligand_numbers&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>
    <span class="k">if</span> <span class="s2">&quot;ligand_names&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">input_csv_df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">input_csv_df</span><span class="p">[</span><span class="s2">&quot;ligand_names&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>
    <span class="k">if</span> <span class="s2">&quot;ligand_tasks&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">input_csv_df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">input_csv_df</span><span class="p">[</span><span class="s2">&quot;ligand_tasks&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;PA&quot;</span>  <span class="c1"># NOTE: we predict poses and affinities by default</span>
    <span class="n">input_csv_df</span><span class="p">[</span><span class="s2">&quot;ligand_numbers&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">input_csv_df</span><span class="p">[</span><span class="s2">&quot;ligand_numbers&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
        <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">ast</span><span class="o">.</span><span class="n">literal_eval</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span> <span class="k">else</span> <span class="n">x</span>
    <span class="p">)</span>
    <span class="n">input_csv_df</span><span class="p">[</span><span class="s2">&quot;ligand_names&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">input_csv_df</span><span class="p">[</span><span class="s2">&quot;ligand_names&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
        <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">ast</span><span class="o">.</span><span class="n">literal_eval</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span> <span class="k">else</span> <span class="n">x</span>
    <span class="p">)</span>

    <span class="k">if</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_method_ligands_pre_ranking</span> <span class="ow">and</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_method_ligands_post_ranking</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="s2">&quot;Only one of `relax_method_ligands_pre_ranking` and `relax_method_ligands_post_ranking` can be set to `true`.&quot;</span>
        <span class="p">)</span>

    <span class="k">if</span> <span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_benchmarking</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">cfg</span><span class="o">.</span><span class="n">pocket_only_baseline</span><span class="p">:</span>
            <span class="c1"># NOTE: this is necessary to support protein pocket-based experiments</span>
            <span class="k">with</span> <span class="n">open_dict</span><span class="p">(</span><span class="n">cfg</span><span class="p">):</span>
                <span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_benchmarking_apo_protein_dir</span> <span class="o">+=</span> <span class="s2">&quot;_bs_cropped&quot;</span>
            <span class="k">assert</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span>
                <span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_benchmarking_apo_protein_dir</span>
            <span class="p">),</span> <span class="s2">&quot;Ensemble benchmarking for protein pocket-based experiments requires `ensemble_benchmarking_apo_protein_dir` to be set to a valid directory.&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_benchmarking_apo_protein_dir</span><span class="p">):</span>
            <span class="c1"># NOTE: this is necessary to support e.g., CASP15 ensemble benchmarking</span>
            <span class="k">with</span> <span class="n">open_dict</span><span class="p">(</span><span class="n">cfg</span><span class="p">):</span>
                <span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_benchmarking_apo_protein_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                    <span class="n">Path</span><span class="p">(</span><span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_benchmarking_apo_protein_dir</span><span class="p">)</span><span class="o">.</span><span class="n">parent</span><span class="p">,</span>
                    <span class="s2">&quot;predicted_structures&quot;</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="k">assert</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span>
                <span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_benchmarking_apo_protein_dir</span>
            <span class="p">),</span> <span class="s2">&quot;Ensemble benchmarking requires `ensemble_benchmarking_apo_protein_dir` to be set to a valid directory.&quot;</span>
        <span class="k">assert</span> <span class="n">cfg</span><span class="o">.</span><span class="n">resume</span> <span class="ow">is</span> <span class="kc">True</span><span class="p">,</span> <span class="s2">&quot;Ensemble benchmarking requires `resume=True`.&quot;</span>

    <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">input_csv_df</span><span class="o">.</span><span class="n">itertuples</span><span class="p">():</span>
        <span class="k">if</span> <span class="p">(</span>
            <span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_benchmarking</span>
            <span class="ow">and</span> <span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_benchmarking_dataset</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">row</span><span class="o">.</span><span class="n">protein_input</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">sep</span><span class="p">)[</span><span class="o">-</span><span class="mi">3</span><span class="p">]</span>
        <span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Row </span><span class="si">{</span><span class="n">row</span><span class="o">.</span><span class="n">Index</span><span class="si">}</span><span class="s2"> does not belong to the ensemble benchmarking dataset </span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_benchmarking_dataset</span><span class="si">}</span><span class="s2">.&quot;</span>
            <span class="p">)</span>
        <span class="k">if</span> <span class="n">row</span><span class="o">.</span><span class="n">ligand_tasks</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;P&quot;</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;PA&quot;</span><span class="p">]:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Row </span><span class="si">{</span><span class="n">row</span><span class="o">.</span><span class="n">Index</span><span class="si">}</span><span class="s2"> has an invalid `ligand_tasks` value of </span><span class="si">{</span><span class="n">row</span><span class="o">.</span><span class="n">ligand_tasks</span><span class="si">}</span><span class="s2">. Should be one of `P`, `A`, and `PA`.&quot;</span>
            <span class="p">)</span>

        <span class="n">config</span> <span class="o">=</span> <span class="p">(</span>
            <span class="s2">&quot;_relaxed&quot;</span>
            <span class="k">if</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_method_ligands_pre_ranking</span> <span class="ow">or</span> <span class="n">cfg</span><span class="o">.</span><span class="n">relax_method_ligands_post_ranking</span>
            <span class="k">else</span> <span class="s2">&quot;&quot;</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="p">(</span>
            <span class="n">cfg</span><span class="o">.</span><span class="n">skip_existing</span>
            <span class="ow">and</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cfg</span><span class="o">.</span><span class="n">output_dir</span><span class="p">,</span> <span class="n">row</span><span class="o">.</span><span class="n">name</span> <span class="o">+</span> <span class="n">config</span><span class="p">))</span>
            <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cfg</span><span class="o">.</span><span class="n">output_dir</span><span class="p">,</span> <span class="n">row</span><span class="o">.</span><span class="n">name</span> <span class="o">+</span> <span class="n">config</span><span class="p">,</span> <span class="s2">&quot;*.sdf&quot;</span><span class="p">))))</span>
        <span class="p">):</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Skipping target </span><span class="si">{</span><span class="n">row</span><span class="o">.</span><span class="n">name</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">config</span><span class="si">}</span><span class="s2"> as its predictions already exist in </span><span class="si">{</span><span class="n">cfg</span><span class="o">.</span><span class="n">output_dir</span><span class="si">}</span><span class="s2">.&quot;</span>
            <span class="p">)</span>
            <span class="k">continue</span>

        <span class="c1"># ensure an input protein structure is available</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">row</span><span class="o">.</span><span class="n">protein_input</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span> <span class="ow">and</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">row</span><span class="o">.</span><span class="n">protein_input</span><span class="p">):</span>
            <span class="n">temp_protein_filepath</span> <span class="o">=</span> <span class="n">row</span><span class="o">.</span><span class="n">protein_input</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">cfg</span><span class="o">.</span><span class="n">ensemble_benchmarking</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span>
                    <span class="s2">&quot;An input (e.g., predicted) protein structure must be available for ensemble benchmarking.&quot;</span>
                <span class="p">)</span>
            <span class="c1"># NOTE: a placeholder protein sequence is used when making ligand-only predictions</span>
            <span class="n">row_protein_input</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">row</span><span class="o">.</span><span class="n">protein_input</span>
                <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">row</span><span class="o">.</span><span class="n">protein_input</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">row</span><span class="o">.</span><span class="n">protein_input</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span>
                <span class="k">else</span> <span class="n">LIGAND_ONLY_RECEPTOR_PLACEHOLDER_SEQUENCE</span>
            <span class="p">)</span>
            <span class="n">row_name</span> <span class="o">=</span> <span class="p">(</span>
                <span class="s2">&quot;ligand_only&quot;</span>
                <span class="k">if</span> <span class="n">row_protein_input</span> <span class="o">==</span> <span class="n">LIGAND_ONLY_RECEPTOR_PLACEHOLDER_SEQUENCE</span>
                <span class="k">else</span> <span class="n">row</span><span class="o">.</span><span class="n">name</span>
            <span class="p">)</span>
            <span class="n">temp_protein_filepath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cfg</span><span class="o">.</span><span class="n">temp_protein_dir</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">row_name</span><span class="si">}</span><span class="s2">.pdb&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">temp_protein_filepath</span><span class="p">):</span>
                <span class="n">temp_fasta_filepath</span> <span class="o">=</span> <span class="n">create_temporary_fasta_file</span><span class="p">(</span><span class="n">row_protein_input</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="n">row_name</span><span class="p">)</span>
                <span class="n">predict_protein_structure_from_sequence</span><span class="p">(</span>
                    <span class="n">cfg</span><span class="o">.</span><span class="n">diffdock_python_exec_path</span><span class="p">,</span>
                    <span class="n">cfg</span><span class="o">.</span><span class="n">structure_prediction_script_path</span><span class="p">,</span>
                    <span class="n">temp_fasta_filepath</span><span class="p">,</span>
                    <span class="n">cfg</span><span class="o">.</span><span class="n">temp_protein_dir</span><span class="p">,</span>
                    <span class="n">chunk_size</span><span class="o">=</span><span class="n">cfg</span><span class="o">.</span><span class="n">structure_prediction_chunk_size</span><span class="p">,</span>
                    <span class="n">cpu_only</span><span class="o">=</span><span class="n">cfg</span><span class="o">.</span><span class="n">structure_prediction_cpu_only</span><span class="p">,</span>
                    <span class="n">cpu_offload</span><span class="o">=</span><span class="n">cfg</span><span class="o">.</span><span class="n">structure_prediction_cpu_offload</span><span class="p">,</span>
                    <span class="n">cuda_device_index</span><span class="o">=</span><span class="n">cfg</span><span class="o">.</span><span class="n">cuda_device_index</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">temp_protein_filepath</span><span class="p">):</span>
                <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Predicted protein structure not found at </span><span class="si">{</span><span class="n">temp_protein_filepath</span><span class="si">}</span><span class="s2">.&quot;</span>
                <span class="p">)</span>

        <span class="c1"># generate bound protein-ligand complex predictions using all selected methods</span>
        <span class="n">method_filepaths_mapping</span> <span class="o">=</span> <span class="p">(</span>
            <span class="p">{</span>
                <span class="n">method</span><span class="p">:</span> <span class="n">get_method_predictions</span><span class="p">(</span><span class="n">method</span><span class="p">,</span> <span class="n">row</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">cfg</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">method</span> <span class="ow">in</span> <span class="n">cfg</span><span class="o">.</span><span class="n">vina_binding_site_methods</span>
            <span class="p">}</span>
            <span class="k">if</span> <span class="n">cfg</span><span class="o">.</span><span class="n">generate_vina_scripts</span> <span class="ow">and</span> <span class="n">cfg</span><span class="o">.</span><span class="n">resume</span>
            <span class="k">else</span> <span class="kc">None</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="n">cfg</span><span class="o">.</span><span class="n">flowdock_auxiliary_estimation_input_dir</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">auxiliary_estimation_input_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                <span class="n">cfg</span><span class="o">.</span><span class="n">flowdock_auxiliary_estimation_input_dir</span><span class="p">,</span>
                <span class="n">row</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="k">assert</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span>
                <span class="n">auxiliary_estimation_input_dir</span>
            <span class="p">),</span> <span class="sa">f</span><span class="s2">&quot;Auxiliary estimation input directory not found at </span><span class="si">{</span><span class="n">auxiliary_estimation_input_dir</span><span class="si">}</span><span class="s2">. &quot;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">auxiliary_estimation_input_dir</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="n">ensemble_predictions_dict</span><span class="p">,</span> <span class="n">generating_scripts</span> <span class="o">=</span> <span class="n">generate_ensemble_predictions</span><span class="p">(</span>
            <span class="n">temp_protein_filepath</span><span class="p">,</span>
            <span class="n">row</span><span class="o">.</span><span class="n">ligand_smiles</span><span class="p">,</span>
            <span class="n">row</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
            <span class="n">cfg</span><span class="p">,</span>
            <span class="n">method_filepaths_mapping</span><span class="o">=</span><span class="n">method_filepaths_mapping</span><span class="p">,</span>
            <span class="n">generate_hpc_scripts</span><span class="o">=</span><span class="n">cfg</span><span class="o">.</span><span class="n">generate_hpc_scripts</span><span class="p">,</span>
            <span class="n">auxiliary_estimation_input_dir</span><span class="o">=</span><span class="n">auxiliary_estimation_input_dir</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="c1"># inform the user of next steps</span>
        <span class="k">if</span> <span class="n">generating_scripts</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;All method prediction scripts for target </span><span class="si">{</span><span class="n">row</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2"> have been generated. Please run these scripts and continue this ensemble generation pipeline with `resume=True`.&quot;</span>
            <span class="p">)</span>
            <span class="k">continue</span>

        <span class="c1"># skip to the next target if no predictions from any method were found</span>
        <span class="n">predictions_found</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">for</span> <span class="n">method</span> <span class="ow">in</span> <span class="n">ensemble_predictions_dict</span><span class="p">:</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">ensemble_predictions_dict</span><span class="p">[</span><span class="n">method</span><span class="p">]):</span>
                <span class="n">predictions_found</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="k">break</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">predictions_found</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;No predictions from any method found for target </span><span class="si">{</span><span class="n">row</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">. Skipping...&quot;</span>
            <span class="p">)</span>
            <span class="k">continue</span>

        <span class="c1"># rank the predictions to select the top prediction(s)</span>
        <span class="n">is_multi_ligand_target</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">row</span><span class="o">.</span><span class="n">ligand_names</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span>
        <span class="n">ranked_predictions</span> <span class="o">=</span> <span class="n">rank_ensemble_predictions</span><span class="p">(</span>
            <span class="n">ensemble_predictions_dict</span><span class="p">,</span> <span class="n">row</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">is_multi_ligand_target</span><span class="p">,</span> <span class="n">cfg</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">len</span><span class="p">(</span><span class="n">ranked_predictions</span><span class="p">):</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No ranked predictions found for target </span><span class="si">{</span><span class="n">row</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">. Skipping...&quot;</span><span class="p">)</span>
            <span class="k">continue</span>

        <span class="c1"># save the top-ranked predictions to the output directory</span>
        <span class="n">save_ranked_predictions</span><span class="p">(</span>
            <span class="n">ranked_predictions</span><span class="p">,</span>
            <span class="n">temp_protein_filepath</span><span class="p">,</span>
            <span class="n">row</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
            <span class="p">(</span>
                <span class="kc">None</span>
                <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">row</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">)</span> <span class="ow">and</span> <span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">row</span><span class="o">.</span><span class="n">ligand_numbers</span><span class="p">)</span><span class="o">.</span><span class="n">any</span><span class="p">()</span>
                <span class="k">else</span> <span class="n">row</span><span class="o">.</span><span class="n">ligand_numbers</span>
            <span class="p">),</span>
            <span class="p">(</span>
                <span class="kc">None</span>
                <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">row</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">)</span> <span class="ow">and</span> <span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">row</span><span class="o">.</span><span class="n">ligand_names</span><span class="p">)</span><span class="o">.</span><span class="n">any</span><span class="p">()</span>
                <span class="k">else</span> <span class="n">row</span><span class="o">.</span><span class="n">ligand_names</span>
            <span class="p">),</span>
            <span class="p">(</span>
                <span class="kc">None</span>
                <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">row</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">)</span> <span class="ow">and</span> <span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">row</span><span class="o">.</span><span class="n">ligand_tasks</span><span class="p">)</span><span class="o">.</span><span class="n">any</span><span class="p">()</span>
                <span class="k">else</span> <span class="n">row</span><span class="o">.</span><span class="n">ligand_tasks</span>
            <span class="p">),</span>
            <span class="n">cfg</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Ensemble generation for target </span><span class="si">{</span><span class="n">row</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2"> has been completed.&quot;</span><span class="p">)</span></div>



<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="n">register_custom_omegaconf_resolvers</span><span class="p">()</span>
    <span class="n">main</span><span class="p">()</span>
</pre></div>
        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          
          
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2025, Alex Morehead
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer no-toc">
      
      
      
    </aside>
  </div>
</div><script src="../../../_static/jquery.js?v=5d32c60e"></script>
    <script src="../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
    <script src="../../../_static/documentation_options.js?v=6c02275b"></script>
    <script src="../../../_static/doctools.js?v=9a2dae69"></script>
    <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../_static/scripts/furo.js?v=46bd48cc"></script>
    <script src="../../../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../../../_static/copybutton.js?v=f281be69"></script>
    <script src="../../../_static/tabs.js?v=3ee01567"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.1.10/require.min.js"></script>
    <script src="http://ajax.googleapis.com/ajax/libs/jquery/1.9.1/jquery.min.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    </body>
</html>